{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis and visualization of Lasso model results\n",
    "Author: Robert Ietswaart  \n",
    "Date: 20220621  \n",
    "License: BSD2. \n",
    "ij: interactive job with 4 cores and 128GB RAM.\n",
    "Load modules j3dl and activate virtual environment using j4RNAdecay on O2.  \n",
    "Python v3.7.4\n",
    "\n",
    "Source: `ML_20220616_subcell_feat_select2.ipynb`  \n",
    "For RNA flow project.\n",
    "This code dev for py script for a batch run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import copy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option(\"display.max_columns\", None)\n",
    "pd.set_option(\"display.max_rows\", None)\n",
    "import logging\n",
    "import argparse\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "import seaborn as sns\n",
    "from scipy.stats import ttest_1samp, ttest_rel #ttest_ind,\n",
    "\n",
    "from sklearn.model_selection import train_test_split, KFold\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn import linear_model\n",
    "\n",
    "from __init__ import default_logger_format, default_date_format "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def main():\n",
    "np.random.seed(12345)\n",
    "\n",
    "parser = argparse.ArgumentParser(\n",
    "    description='Analysis and visualization of Lasso model results')\n",
    "\n",
    "num2feat_files = {1: 'gene_structure_v2.txt gene_structure_agarwal.txt',\n",
    "                  2: 'gene_sequence_v2.txt gene_sequence_agarwal.txt',\n",
    "                  3: 'Codon_features.tsv.gz', 4: 'kmer_ORF_features.tsv.gz',\n",
    "                  5: 'kmer_5UTR_features.tsv.gz', 6: 'kmer_3UTR_features.tsv.gz',\n",
    "                  7: 'polyA_tail_length_data.txt', 8: 'RNA_modifications_v2.txt',\n",
    "                  9: 'gene_location_v2.txt', 10: 'histone_modifications_v1.txt',\n",
    "                  11: 'microRNA_targets_v2.txt', 12: 'RBP_targets_filter_v2.txt',\n",
    "                  13: 'TF_scRNAseq_v1.csv', 14: 'TF_MSigDB_targets_v2.tsv'}\n",
    "\n",
    "args = parser.parse_args(\"\")#EDIT: added \"\" as argument to run in ipynb instead of .py script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "path = os.path.join('/n','groups','churchman','ri23','bseq','RF20220426')\n",
    "\n",
    "# Add a logger specific to the project and processing stage\n",
    "logger = logging.getLogger('ML')\n",
    "log_file = os.path.join(path,'LogErr', 'LASSO_20220621_subcell_analysis.log')\n",
    "formatter = logging.Formatter(default_logger_format,\n",
    "                              datefmt=default_date_format)\n",
    "log_handler = logging.FileHandler(log_file)\n",
    "log_handler.setFormatter(formatter)\n",
    "logger.addHandler(log_handler)\n",
    "\n",
    "organisms = ['m','h']\n",
    "org_map = {'m': 'mouse', 'h': 'human'}\n",
    "org_red_reps = {'m': ['G_R','H_S'], 'h': ['T', 'U']}\n",
    "\n",
    "k_bound_lo = 1e-4 #unit: min^-1: 1 per 7 days\n",
    "k_bound_hi = 1e4 #unit: min^-1: 1 per 6 ms\n",
    "\n",
    "RATE_TYPE = ['half_life_','k_','T_']\n",
    "\n",
    "rt = RATE_TYPE[1] #ML model dependent variable: rates\n",
    "\n",
    "Timescales = ['chr',\n",
    "              'nuc',\n",
    "              'nucexp_from_chr',\n",
    "              'nucdeg',\n",
    "              'cyto',\n",
    "              'poly_entry',\n",
    "              'whole_cell',\n",
    "              'nucexp_from_nucdeg']\n",
    "Timescales = [rt + ts for ts in Timescales]\n",
    "\n",
    "OUT_TYPES = ['.Mean', '.MAP', '.0.975.quantile', '.0.025.quantile']\n",
    "\n",
    "B = dict()          #Bayes fits file\n",
    "K = dict()          #Bayes Factor\n",
    "\n",
    "o = 'h'\n",
    "# for o in organisms:    \n",
    "if o == 'm':\n",
    "    path_b = os.path.join('/n','groups','churchman','ri23','bseq','Bayes20220222')\n",
    "    filename_b = 'Bayes20220228.tsv'   \n",
    "elif o == 'h':\n",
    "    path_b = os.path.join('/n','groups','churchman','ri23','bseq','Bayes20220228_human')\n",
    "    filename_b = 'Bayes20220228_human.tsv'\n",
    "path_k = os.path.join('/n','groups','churchman','ri23','bseq','BayesFactor20220307')\n",
    "filename_k = 'Bayes_factor_20220315_' + org_map[o] + '.tsv'\n",
    "\n",
    "B[o] = pd.read_csv(os.path.join(path_b, filename_b), sep='\\t')\n",
    "K[o] = pd.read_csv(os.path.join(path_k, filename_k), sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocess rates: dependent variables in model\n",
    "- preprocess nucexp: nucexp_from_chr or nucexp_from_nucdeg depending on nucdeg no or yes \n",
    "- clip to domain bounds\n",
    "- log transform\n",
    "- standardize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-07-20 16:32:20] ML - human all genes: 18155\n"
     ]
    }
   ],
   "source": [
    "T_bf = 100\n",
    "\n",
    "C = copy.deepcopy(B)\n",
    "# for o in organisms:\n",
    "C[o] = C[o].merge(K[o], on='Gene', how='outer', suffixes=('', '_bf'))\n",
    "logger.info('%s all genes: %d' % (org_map[o], len(C[o])))\n",
    "\n",
    "for ot in OUT_TYPES:\n",
    "    for rr in org_red_reps[o]:\n",
    "        ts = rt+'nucexp_from_chr'\n",
    "        C[o][rr+'.'+ts+ot].where(((C[o][rr+'.bayes_factor'] <= T_bf) | (C[o][rr+'.bayes_factor'].isna())), \n",
    "                                 C[o][rr+'.'+rt+'nucexp_from_nucdeg'+ot], inplace=True)                        \n",
    "        ts = rt+'nucdeg'\n",
    "        C[o][rr+'.'+ts+ot].where(C[o][rr+'.bayes_factor'] > T_bf, np.nan, inplace=True)\n",
    "\n",
    "\n",
    "        for ts in Timescales:\n",
    "\n",
    "            ###Clip range of values beyond numerical integration domain bounds                  \n",
    "            C[o][rr+'.'+ts+ot].where(((C[o][rr+'.'+ts+ot] > k_bound_lo) | (C[o][rr+'.'+ts+ot].isna())), \n",
    "                                     k_bound_lo, inplace=True) \n",
    "            C[o][rr+'.'+ts+ot].where(((C[o][rr+'.'+ts+ot] < k_bound_hi) | (C[o][rr+'.'+ts+ot].isna())), \n",
    "                                     k_bound_hi, inplace=True)  \n",
    "\n",
    "            #standardize rates (to Z-score) to enhance learning\n",
    "            rates = np.log(C[o][rr+'.'+ts+ot])\n",
    "            C[o][rr+'z'+ts+ot] = (rates - rates.mean()) / rates.std()\n",
    "            \n",
    "Timescales.remove(rt+'nucdeg')            \n",
    "Timescales.remove(rt+'nucexp_from_nucdeg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-07-20 16:32:34] ML - Load r2 train/cv/test and features for k_chr\n",
      "INFO: [2022-07-20 16:32:34] ML - Load r2 train/cv/test and features for k_nuc\n",
      "INFO: [2022-07-20 16:32:35] ML - Load r2 train/cv/test and features for k_nucexp_from_chr\n",
      "INFO: [2022-07-20 16:32:35] ML - Load r2 train/cv/test and features for k_cyto\n",
      "INFO: [2022-07-20 16:32:36] ML - Load r2 train/cv/test and features for k_poly_entry\n",
      "INFO: [2022-07-20 16:32:36] ML - Load r2 train/cv/test and features for k_whole_cell\n",
      "INFO: [2022-07-20 16:32:36] ML - Load r2 train/cv/test and features for Agarwal whole cell half lives\n"
     ]
    }
   ],
   "source": [
    "subfolder = os.path.join('select_features')#'old','20220620')#,'20220626'#20220625 #,\n",
    "# L1_alpha = [0.0001 * 10**(i) for i in range(4)] #20220620:\n",
    "L1_alpha = [0.0001, 0.00033, 0.00066, 0.001, 0.0033, 0.0066, 0.01, 0.033, 0.066, 0.1]\n",
    "rand_seed = 42\n",
    "shuffle = True\n",
    "test_size = 0.1\n",
    "\n",
    "RND2 = dict()\n",
    "for ts in Timescales:\n",
    "    logger.info('Load r2 train/cv/test and features for %s' % ts)\n",
    "    filename = 'LASSO2_'+ts+'_r2_' \n",
    "    RND2[ts+'r2_train'] = pd.read_csv(os.path.join(path,subfolder,filename+'train.tsv'),sep='\\t')#'old','20220620',\n",
    "    RND2[ts+'r2_cv'] = pd.read_csv(os.path.join(path,subfolder,filename+'cv.tsv'),sep='\\t') \n",
    "    RND2[ts+'r2_test'] = pd.read_csv(os.path.join(path,subfolder,filename+'test_seed'+str(rand_seed)+'.tsv'),sep='\\t')     \n",
    "\n",
    "    for a_ in L1_alpha:\n",
    "        a_ = str(a_)\n",
    "        filename = 'LASSO2_'+ts+'_alpha'+a_+'_nonzero_feats.tsv'\n",
    "        RND2[ts+a_] = pd.read_csv(os.path.join(path,subfolder,filename),sep='\\t')\n",
    "        \n",
    "        \n",
    "ts = 'Agarwal whole cell half lives'\n",
    "logger.info('Load r2 train/cv/test and features for %s' % ts)\n",
    "filename = 'LASSO2_Agarwal_tcvtsplit_r2_'\n",
    "RND2[ts+'r2_train'] = pd.read_csv(os.path.join(path,subfolder,filename+'train.tsv'),sep='\\t')\n",
    "RND2[ts+'r2_cv'] = pd.read_csv(os.path.join(path,subfolder,filename+'cv.tsv'),sep='\\t') \n",
    "RND2[ts+'r2_test'] = pd.read_csv(os.path.join(path,subfolder,filename+'test_seed'+str(rand_seed)+'.tsv'),sep='\\t')     \n",
    "\n",
    "for a_ in L1_alpha:\n",
    "    a_ = str(a_)\n",
    "    filename = 'LASSO2_Agarwal_tcvtsplit_alpha'+a_+'_nonzero_feats.tsv'\n",
    "    RND2[ts+a_] = pd.read_csv(os.path.join(path,subfolder,filename),sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.0001 ratio 1.257607, r2_train 0.366529 r2_cv 0.291449\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.00033 ratio 1.240510, r2_train 0.362909 r2_cv 0.292548\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.00066 ratio 1.222767, r2_train 0.353313 r2_cv 0.288945\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.001 ratio 1.205979, r2_train 0.340716 r2_cv 0.282523\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.0033 ratio 1.112169, r2_train 0.280829 r2_cv 0.252506\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.0066 ratio 1.079094, r2_train 0.250234 r2_cv 0.231893\n",
      "INFO: [2022-07-20 16:33:09] ML - 0.0066 r2_opt 0.231893\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.01 ratio 1.057662, r2_train 0.230381 r2_cv 0.217821\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.033 ratio 1.039616, r2_train 0.146078 r2_cv 0.140512\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.066 ratio 1.032707, r2_train 0.083579 r2_cv 0.080932\n",
      "INFO: [2022-07-20 16:33:09] ML - k_chr 0.1 ratio 1.030829, r2_train 0.039044 r2_cv 0.037877\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.0001 ratio 1.121935, r2_train 0.313743 r2_cv 0.279645\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.00033 ratio 1.117829, r2_train 0.312504 r2_cv 0.279563\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.00066 ratio 1.113089, r2_train 0.309256 r2_cv 0.277836\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.001 ratio 1.109239, r2_train 0.305232 r2_cv 0.275173\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.0033 ratio 1.086802, r2_train 0.285203 r2_cv 0.262424\n",
      "INFO: [2022-07-20 16:33:09] ML - 0.0033 r2_opt 0.262424\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.0066 ratio 1.063047, r2_train 0.269804 r2_cv 0.253803\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.01 ratio 1.053945, r2_train 0.257717 r2_cv 0.244526\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.033 ratio 1.029932, r2_train 0.198741 r2_cv 0.192965\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.066 ratio 1.020854, r2_train 0.126242 r2_cv 0.123663\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nuc 0.1 ratio 1.021399, r2_train 0.092726 r2_cv 0.090783\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.0001 ratio 1.202766, r2_train 0.425429 r2_cv 0.353709\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.00033 ratio 1.186315, r2_train 0.422012 r2_cv 0.355734\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.00066 ratio 1.166965, r2_train 0.413580 r2_cv 0.354407\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.001 ratio 1.150964, r2_train 0.402677 r2_cv 0.349861\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.0033 ratio 1.074956, r2_train 0.354540 r2_cv 0.329818\n",
      "INFO: [2022-07-20 16:33:09] ML - 0.0033 r2_opt 0.329818\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.0066 ratio 1.050212, r2_train 0.332379 r2_cv 0.316487\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.01 ratio 1.040019, r2_train 0.319510 r2_cv 0.307216\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.033 ratio 1.023118, r2_train 0.262888 r2_cv 0.256948\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.066 ratio 1.018634, r2_train 0.198068 r2_cv 0.194444\n",
      "INFO: [2022-07-20 16:33:09] ML - k_nucexp_from_chr 0.1 ratio 1.015429, r2_train 0.159747 r2_cv 0.157320\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.0001 ratio 1.137147, r2_train 0.500119 r2_cv 0.439802\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.00033 ratio 1.124017, r2_train 0.496033 r2_cv 0.441304\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.00066 ratio 1.109021, r2_train 0.486891 r2_cv 0.439028\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.001 ratio 1.096612, r2_train 0.476220 r2_cv 0.434265\n",
      "INFO: [2022-07-20 16:33:09] ML - 0.001 r2_opt 0.434265\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.0033 ratio 1.046863, r2_train 0.430197 r2_cv 0.410939\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.0066 ratio 1.029387, r2_train 0.406743 r2_cv 0.395132\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.01 ratio 1.022557, r2_train 0.393421 r2_cv 0.384742\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.033 ratio 1.015042, r2_train 0.321248 r2_cv 0.316488\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.066 ratio 1.005938, r2_train 0.260253 r2_cv 0.258717\n",
      "INFO: [2022-07-20 16:33:09] ML - k_cyto 0.1 ratio 1.004208, r2_train 0.233759 r2_cv 0.232779\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.0001 ratio 1.522129, r2_train 0.185851 r2_cv 0.122099\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.00033 ratio 1.459937, r2_train 0.184047 r2_cv 0.126065\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.00066 ratio 1.401566, r2_train 0.179604 r2_cv 0.128145\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.001 ratio 1.355421, r2_train 0.173616 r2_cv 0.128090\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.0033 ratio 1.182308, r2_train 0.149037 r2_cv 0.126056\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.0066 ratio 1.122400, r2_train 0.136180 r2_cv 0.121330\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.01 ratio 1.096930, r2_train 0.126753 r2_cv 0.115552\n",
      "INFO: [2022-07-20 16:33:09] ML - 0.01 r2_opt 0.115552\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.033 ratio 1.051019, r2_train 0.081083 r2_cv 0.077147\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.066 ratio 1.067652, r2_train 0.035804 r2_cv 0.033536\n",
      "INFO: [2022-07-20 16:33:09] ML - k_poly_entry 0.1 ratio 1.153486, r2_train 0.011722 r2_cv 0.010163\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.0001 ratio 1.214323, r2_train 0.553800 r2_cv 0.456057\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.00033 ratio 1.172324, r2_train 0.524747 r2_cv 0.447613\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.00066 ratio 1.143363, r2_train 0.494857 r2_cv 0.432808\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.001 ratio 1.122715, r2_train 0.471627 r2_cv 0.420078\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.0033 ratio 1.057951, r2_train 0.410384 r2_cv 0.387905\n",
      "INFO: [2022-07-20 16:33:09] ML - 0.0033 r2_opt 0.387905\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.0066 ratio 1.038999, r2_train 0.385875 r2_cv 0.371391\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.01 ratio 1.033803, r2_train 0.367385 r2_cv 0.355372\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.033 ratio 1.024164, r2_train 0.276878 r2_cv 0.270345\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.066 ratio 1.012717, r2_train 0.185475 r2_cv 0.183146\n",
      "INFO: [2022-07-20 16:33:09] ML - k_whole_cell 0.1 ratio 1.009604, r2_train 0.156522 r2_cv 0.155033\n",
      "INFO: [2022-07-20 16:33:09] ML -                rate   alpha  type        r2\n",
      "0               chr  0.0066    CV  0.260385\n",
      "1               chr  0.0066    CV  0.250939\n",
      "2               chr  0.0066    CV  0.234706\n",
      "3               chr  0.0066    CV  0.221080\n",
      "4               chr  0.0066    CV  0.222139\n",
      "5               chr  0.0066    CV  0.172542\n",
      "6               chr  0.0066    CV  0.235476\n",
      "7               chr  0.0066    CV  0.263855\n",
      "8               chr  0.0066    CV  0.229881\n",
      "9               chr  0.0066    CV  0.227930\n",
      "10              chr  0.0066  Test  0.203088\n",
      "11              nuc  0.0033    CV  0.247473\n",
      "12              nuc  0.0033    CV  0.281090\n",
      "13              nuc  0.0033    CV  0.280896\n",
      "14              nuc  0.0033    CV  0.319324\n",
      "15              nuc  0.0033    CV  0.246808\n",
      "16              nuc  0.0033    CV  0.241376\n",
      "17              nuc  0.0033    CV  0.236936\n",
      "18              nuc  0.0033    CV  0.232408\n",
      "19              nuc  0.0033    CV  0.269989\n",
      "20              nuc  0.0033    CV  0.267939\n",
      "21              nuc  0.0033  Test  0.217234\n",
      "22  nucexp_from_chr  0.0033    CV  0.316356\n",
      "23  nucexp_from_chr  0.0033    CV  0.285599\n",
      "24  nucexp_from_chr  0.0033    CV  0.312149\n",
      "25  nucexp_from_chr  0.0033    CV  0.311969\n",
      "26  nucexp_from_chr  0.0033    CV  0.324467\n",
      "27  nucexp_from_chr  0.0033    CV  0.349743\n",
      "28  nucexp_from_chr  0.0033    CV  0.330944\n",
      "29  nucexp_from_chr  0.0033    CV  0.324521\n",
      "30  nucexp_from_chr  0.0033    CV  0.357252\n",
      "31  nucexp_from_chr  0.0033    CV  0.385180\n",
      "32  nucexp_from_chr  0.0033  Test  0.319116\n",
      "33             cyto   0.001    CV  0.411044\n",
      "34             cyto   0.001    CV  0.442738\n",
      "35             cyto   0.001    CV  0.461682\n",
      "36             cyto   0.001    CV  0.405346\n",
      "37             cyto   0.001    CV  0.431630\n",
      "38             cyto   0.001    CV  0.445575\n",
      "39             cyto   0.001    CV  0.432007\n",
      "40             cyto   0.001    CV  0.453370\n",
      "41             cyto   0.001    CV  0.451245\n",
      "42             cyto   0.001    CV  0.408010\n",
      "43             cyto   0.001  Test  0.402409\n",
      "44       poly_entry    0.01    CV  0.102592\n",
      "45       poly_entry    0.01    CV  0.124560\n",
      "46       poly_entry    0.01    CV  0.093397\n",
      "47       poly_entry    0.01    CV  0.092487\n",
      "48       poly_entry    0.01    CV  0.131988\n",
      "49       poly_entry    0.01    CV  0.128567\n",
      "50       poly_entry    0.01    CV  0.139452\n",
      "51       poly_entry    0.01    CV  0.102526\n",
      "52       poly_entry    0.01    CV  0.110679\n",
      "53       poly_entry    0.01    CV  0.129274\n",
      "54       poly_entry    0.01  Test  0.131596\n",
      "55       whole_cell  0.0033    CV  0.357544\n",
      "56       whole_cell  0.0033    CV  0.430740\n",
      "57       whole_cell  0.0033    CV  0.392660\n",
      "58       whole_cell  0.0033    CV  0.374558\n",
      "59       whole_cell  0.0033    CV  0.418138\n",
      "60       whole_cell  0.0033    CV  0.384341\n",
      "61       whole_cell  0.0033    CV  0.368345\n",
      "62       whole_cell  0.0033    CV  0.378282\n",
      "63       whole_cell  0.0033    CV  0.388390\n",
      "64       whole_cell  0.0033    CV  0.386049\n",
      "65       whole_cell  0.0033  Test  0.378828\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-07-20 16:33:09] ML - Write final params_opt to file LASSOfinal_params_opt.tsv\n"
     ]
    }
   ],
   "source": [
    "cv_folds = 10\n",
    "\n",
    "def get_cv_alpha_opt(ts):    \n",
    "    #simple max average R2CV + max on overfitting train vs CV\n",
    "    max_train_cv_ratio = 1.1 #require at most 10% training overfitting compared to CV to identify robust features\n",
    "    r2_opt = 0\n",
    "    alpha_opt = np.nan\n",
    "    for a_ in L1_alpha:\n",
    "        a_ = str(a_)\n",
    "        train_samples = RND2[ts+'r2_train'][a_].values\n",
    "        cv_samples = RND2[ts+'r2_cv'][a_].values        \n",
    "        r2_cv = np.mean(cv_samples)\n",
    "        train_cv_ratio = np.mean(train_samples) / r2_cv\n",
    "        logger.info('%s %s ratio %f, r2_train %f r2_cv %f' % \n",
    "                    (ts, a_, train_cv_ratio, np.mean(train_samples), r2_cv))\n",
    "        if train_cv_ratio < max_train_cv_ratio:#if not in overfitting regime\n",
    "            if r2_opt < r2_cv:#better than best hypothesis so far   \n",
    "                r2_opt = r2_cv\n",
    "                alpha_opt = a_\n",
    "                r2_cv_opt = cv_samples\n",
    "                logger.info('%s r2_opt %f' % (alpha_opt, r2_opt))\n",
    "\n",
    "    return alpha_opt, r2_cv_opt\n",
    "\n",
    "\n",
    "params_opt = dict()\n",
    "params_opt['rate'] = []\n",
    "params_opt['alpha'] = []\n",
    "params_opt['type'] = []\n",
    "params_opt['r2'] = []\n",
    "\n",
    "for ts in Timescales: \n",
    "    alpha_opt, r2_cv_opt = get_cv_alpha_opt(ts)\n",
    "\n",
    "    #add data to dataframe used for boxplot\n",
    "    params_opt['rate'].extend([ts.replace(rt,'') for _ in range(cv_folds)])\n",
    "    params_opt['alpha'].extend([alpha_opt for _ in range(cv_folds)])\n",
    "    params_opt['type'].extend(['CV'  for _ in range(cv_folds)])\n",
    "    params_opt['r2'].extend(r2_cv_opt)\n",
    "    \n",
    "    params_opt['rate'].append(ts.replace(rt,''))\n",
    "    params_opt['alpha'].append(alpha_opt)\n",
    "    params_opt['type'].append('Test')\n",
    "    params_opt['r2'].extend(RND2[ts+'r2_test'][alpha_opt].values)    \n",
    "       \n",
    "params_opt = pd.DataFrame.from_dict(params_opt)\n",
    "params_opt['rate'] = params_opt['rate'].astype('category')\n",
    "Timescales_lab = [ts.replace(rt,'') for ts in Timescales]\n",
    "params_opt['rate'].cat.reorder_categories(Timescales_lab, inplace=True)\n",
    "params_opt['type'] = params_opt['type'].astype('category')\n",
    "params_opt['type'].cat.reorder_categories(['CV','Test'], inplace=True)\n",
    "\n",
    "logger.info(params_opt)   \n",
    "\n",
    "filename = 'LASSOfinal_params_opt.tsv'\n",
    "params_opt.to_csv(os.path.join(path,subfolder,filename), sep='\\t', index=False) \n",
    "logger.info('Write final params_opt to file %s' % filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subcellular rate: bar plot of CV and test set performances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcQAAAC+CAYAAACrkX8MAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAc5UlEQVR4nO3df3hU1b3v8feeSSYTCARCILEqPy1LK1aOQaWtgVNv29Mjra33saC9ilqt0tvec/tIpdXaa1PRPvqc03qPTxVrzymJLd56a4+Wey1WpSo/5KGm5XDgygJEI0aChpCEMMnkx8z9Yw84hPzYk8xkZyaf1/PkmZm919r7ux2TL2vttdZ24vE4IiIiY13A7wBERERGAyVEERERlBBFREQAJUQRERFACVFERARQQhQREQGUEEVERADI81LIGBMAVgM3A+OBzcBt1tq6fsrHgQ6gJ2nzJ6y1/zG8cEVERDLDU0IEVgHXAouAeuAnwHpjzHxrbayfOn9vrX15+CGKiIhkntcu0xXAg9bVhpsgDXBZxiITEREZQYO2EI0xxcAM4PUT26y1zcaY/cB84NV+qj5pjAkBdcCj1trHvQZVW1tbAFwMHOLUblcREelfEDgD+HNFRUXU72CyjZcu04mJ1+Ze25uT9vX2GWArbjK7HFhnjMmz1j7qMa6LgU0ey4qIyKkqccd6SAq8JMTWxGtxr+2Tkvadwlr7UtLHDcaYnwDXA14T4iGAuXPnEgqFPFYRERnbOjs72bt3LyT+hkpqBk2I1toWY0wdsIBEt2miG3UOsMPjeWKAk0JcPQChUIiCgoIUqomICLrVNCReR5muAe4wxmzEHWX6ALCXPprkxpiLcAfr7MRNhJ8Gbgd+lI6ARUREMsFrQnwQt8t0Mx/OQ7zSWhszxlQCfwA+Zq19BzgzUf5soBt3UM3d1to16Q5eREQkXZzR+IDg2tramcBb8+bNU5epiIhH0WiUXbt2AcyqqKh42+dwso6WbhMREUEJUUREBFBCFBERAZQQRUREACVEkRG3d+9e9u/f73cYItKLEqLICHrllVe48sor+eIXv8jWrVv9DkdEkighioyg/fv3E4/HicViaiWKjDJeJ+aLSBosW7aMuro62traeO+999i2bRsLFy70OywRQQlRZEQdPHiQnTt3Ul9fT2trK+vWrWPr1q0UFRX5HZrImKeEKDKCampqeOONN05+LiwsJC9Pv4Yio4HuIYqMoMWLFxMIBMjPzyccDhOLxThw4IDfYYkISogiIyoajRKLxejq6qKjo4PW1lZeeOEFv8MSEdRlKjKiXnrpw2dnh0IhysrKWLJkiY8RicgJaiGKjKAbb7yRcDhMYWEhF154IS+++CLnnHOO32GJCEqIIiNi586drF27lpkzZ3LhhRcSCoWor6+nra3N79AkBdFolAMHDjAaH5snw6cuU5E0iXZ2URDKP217c3Mzy5cvp729nS1btnD06FFaWlpoaWnhscceY+XKlUM6roys7u5uli1bxhtvvME111xDVVWV3yFJmikhiqRJQSif86++67Tt8e4oPR1RADb9ZQ+Tx3+Y3H7x+9f419dOr5Ns92/vT2+gMiTHjh07OWVm+/btPkcjmaCEKJJhTl4BwVmVxI834kyeSShimTp1Kl3xIMdKP+p3eOLR5MmTWblyJRs3buQb3/iG3+FIBighiowAZ3wpzvjSk58LCgoIBsK0OY6PUUmqbr31Vm699VYA2tvbCYfDOPoOc4YG1YiMsLgTOOVVss9vfvMbLrroIr785S8TiUT8DkfSRL+RIiOsrfAsonkTaCs8y+9QZIg2bNhALBZjz549Wmkoh6jLVCTD4vE4sUM7obONwBkfJ1owmWj+ZL/DkmG48cYbqaur4+Mf/zjnnnuu3+FImighimRY/PgHxI+4zz6MBUMEz76YWOsh4o37cCadRaBkts8RSqoWL17Mxo0b/Q5D0sxTQjTGBIDVwM3AeGAzcJu1tm6QehXANmCLtfZvhxeqSHZyQkUQyIdYF07hZHre20H8yAEgTvz4BziTZuAEgn6HKTLmeb2HuAq4FlgElAPvAOsTibJPxpgwsBZ4ZZgximQ1JzSOoPkcwY9+BmfSWcSPvAkkVjoJTwINrhEZFbz+Jq4AHrSuNtwEaYDLBqhzH/ASbmtSZExz8sI44WIIFuCMn+punDyT4Jy/Hfaw/aqqKmbPnq2VU0bAtm3buP3229m2bZvfoUgGDNplaowpBmYAr5/YZq1tNsbsB+YDr/ZRZxHwBeBvcJPnkOzatWuoVUVGXEVFxaBlHMchMKsSYt04QW/LsfV0dBAMh/vdX11dTTQapaamhnvuucdzvB3HjrF7717P5QUeeeQR6uvraWxsJD9fy+nlGi/3ECcmXpt7bW9O2neSMaYI+Ffga9baiDFmyMHNmzePgoKCIdcXGY0cxwGPyRAgGA5TU17e7/6LHYdtjsOCSGTAcr0tb2jwlMTFZa1lz549xGIxzjzzzFH53y4ajaohMQxeEmJr4rW41/ZJSfuS/SPwnLX2tJajiKTfknicJXr6QkZ1dnZy55130tLSAqCnlOSoQe8hWmtbgDpgwYltiW7UOcCOPqp8HlhujGk0xjTidpl+KvF5VnrCllzX3d3NK6+8Qn19vd+hiHDLLbewe/duwG3hq+cqN3mdh7gGuMMYsxGoBx4A9tL3gJmFvY57O/AJ4CvAoaGHKmPJfffdx7p165g4cSLPP/88JSUlfockOWygR2w1NTWd8nSLQCCQ0kAoPb4re3hNiA/idplu5sN5iFdaa2PGmErgD8DHrLXvWGsbkisaY1qBqLX23TTGLTnu4MGDALS2ttLS0qKEKBnV36O7AOI9XcSdPIh3AdDT08Ob77zXb/ne9Piu7OEpIVprY8CdiZ/e+zYBRQPU/eFQg5Ox6wc/+AFr1qxh/vz5zJqlnnbxjxPMJ3jO5cSOvk28cS/E4xxr+gBKozh56e06raqqorq6mhtuuCGlEcOSHpoRLKPSjBkz+PGPf8yyZcv8DkUEp6CIYPk8woXjAejp6aZn30vEWgbv+Orp6PB8nuQpNOk8rnijtUxFRDwqnDiZrmg7PT090N1OrGE3geKBn1oy2LSZZKlMoVne0DDgfkmdEqKIiAex9/fQ0nyU8ZPLOH78OLH2FpyJZ6T1HJpC4y8lRBGRQcR7uogddqddHIvFCJrPE+yO4uT3v4KQZB/dQxQRGUwgD8ZNAcCZUIbjOEqGOUgtRBkV2tvbqaurY+7cuQQC+neajDKxLpypc3Hyx+OET1uxUnKE/vKI73p6eli2bBlf+tKXuPPO02b2iPiu58Am4nWvEav/y7CfTiKjlxKi+C4SibA38dSFHTv6Wg1QxGedbae+Sk5SQhTfTZgwgbvvvptLL72Uu+++m7a2Nr7zne/w7W9/m+bm3g9ZERl5gemX4hSfTWD6JX6HIhmke4gyKlx33XVcd911ADzxxBOsX78egHPPPZcVK1b4GZoIgQnlMMH7o7UkOykhyqgTj8fdUXyOw/nnn+93OCIyRighyqizdetWysrKmDlzJpWVlX6HIyJjhO4hyqgTiUTIy8sjrhU7RGQEKSGKiIighCijVFdXFwcOHOC5557zOxQRGSOUEGVUOnr0KB988AErV67k/fff9zscERkDlBBlxEQ7uwbcv2HDBh555BG6u7sJBoMAhEIhtm/fPqzjioh4oVGmMmIKQvmcf/Vdfe6Ltx+lZ/9GAAqLJlJSUsIHR5ro6Ghn5cqVrHr0OZzCSX3W3f3b+zMWs4iMHWohZpFIJMK3vvUtrrnmGt5++22/w0kvJ/jh28QcxLy8UNJ+rR8pIpmlhJhF/vSnP/HCCy/w17/+lZqaGr/DSSsnPJHgrEU4U88j2h6hsbGR8cWTCXxkPoFZlTjh4rSer6qqitmzZ1NVVZXW44pI9lJCzCIXXHABkyZNIi8vj09+8pN+h5N2TtFU4p1t9HR30dHRQVtzE86k6QSKpg1Yr6ejI+VzVVdXE41GB/2HxVCOLSLZSfcQs8j06dN56aWX6OzspKSkxO9wMiPpaQLtx1tx3n2d4IxPDFglGA5TU57aOpMXOw7bHIcFkciAdZc3NKR0XBHJXp4SojEmAKwGbgbGA5uB26y1dX2UnQGsAwwQAj4A1gKrrbVaemSYioqK/A4ho5z8ccTbj364IdaTkfMsicdZopVwRCSJ1y7TVcC1wCKgHHgHWJ9IlL0dwU2cZdbaicBnga8C3xh+uJLrAmddROCMjxMovwBn2nkEzqrwOyQRGSO8dpmuAB6w1loAY8wq4DBwGfBqckFrbRuwJ2lTHIjhthhFBuQEQzilH/U7DBEZgwZtIRpjioEZwOsntllrm4H9wPwB6m0yxrQDB4CJwKPDjlZERCRDvLQQJyZeez+6vDlp32mstZXGmCBwKbAESHn9rV27dqVaRUaxiors7P6sra31VC4br8/rteW6bPzuQN9funlJiK2J194TwSYl7euTtbYH2GqMuQy3hbgsleDmzZtHQUFBKlVE0i5b/1h6kcvXNhb0/v6i0agaEsMwaJeptbYFqAMWnNiW6EadA+zweJ58dA9RRERGMa+DatYAdxhjNgL1wAPAXtzpF6cwxnwGaAdqgW6gEvjvwC/SEbDkrnhPJ7GDfyYe6yZ41gKc0Hi/QxKRMcRrQnwQt8t0Mx/OQ7zSWhszxlQCfwA+Zq19J7H/n4DZuKNL64GHcJOoDNGRI0d4+umnOe+886isrPQ7nIyIt9QTP+ZOhI81vUWwfJ7PEYnIWOIpIVprY8CdiZ/e+zYBRUmfnwWeTVeA4vrKV75CfX09AD/96U+54oorfI4o/ZxxUyCQB/EYzvipfocjImOMlm7LEk1NTSffb9myhRdffJGlS5eycOFCH6NKLyc8keC5V7gJMU+DqURkZCkhZonHHnuM+++/n+nTp3P48GHeeustIpFITiVEACeY73cIIjJG6WkXWeLSSy/l2Wef5eGHH6azs5Pjx49TW1vL888/73doIiI5QQkxC0UiEY4ePUprayvf+973/A5HRCQn5GxCfPTRR7nllltycpJqa+uH6yHk7GOgRERGWE7eQ3zzzTd56KGHANi3bx8333wzy5cv9zmq9CkpKeHQoUMEAgEef/xxv8MREckJOdlCnDZtGmVlZQA0NDRw33338dprr/kc1eCinV2eyoVCIaZNm8aFF17I7Nmz03ZcEZGxLCdbiBMmTOD3v/89Dz30EE8++SR5eXlMmTLF77AGVRDK5/yr7xq0XGnLEfKAukNHPJXf/dv70xCdiEhuy7mEuGPHDl5++WWuuuoq7rnnHhYtWkR5eTlz5871OzQRERnFciohdnZ2ctNNNxGJRHj55Zd55plnuPzyy/0OKyN6enro7ukk1vwOTmi8u8qLiIgMWU4lRMdxCIVCRCIRwuGw3+FkTHdXJ0caGojH48A7gEPwo5/BCff7eEoRERlETg2qyc/PZ926dVx//fWUlJSwbds2v0PKiK6uzkQyPCEOMQ2cEREZjpxKiABz5szh3XffxVrL2rVr/Q4nI7pKDHmhwpOfnSlz1GUqIjJMOZUQ4/E4O3fuPDlxPRKJ8Mwzz/DUU08Ri8V8ji59OgtK6Sk558MNsR7/ghERyRE5dQ9x9erV/OpXvyIUClFaWkpjYyPf/e53Aeju7uarX/2qzxGmjzOhjHjDfwAQP/6Bz9GIiGS/nGoh7t69G3BHm3Z1dXH48OGT+xzH8SusjHAKJkB+otu08zjxSNPAFUREZEBZlRAHW3HlrrvuYvHixZx99tkcOXKEtrY2AG666SaWLVs2rGOPNo4TwCk+y+8wRERyRlZ1mXpbyaWU4q4Dp9wzrNlSzxNL7x6wVjau5hIoO594/jgIjcMZp0W+RUSGI6sS4mDiHS30vLeD1q62k9ucaee53Ys5yAkEcUrPGbygiIgMKqcSYuz9N+B4I8mdn2o5iYiIF1l1D3EwzrhS99UJkJ+fT3jcBJyiaT5HlX2qqqqYPXs2VVVVfociIjJiciohBkrPIWj+jtKzZlFWVkbx1DNwnJy6xCHp6ehIqXx1dTXRaJSampq0H1tEZLTKqS5TACdURMSZAR2HiITP8DucUSEYDlNTXu65/MWOwzbHYUEkMmi95Q0Nww1PRGRU8JQQjTEBYDVwMzAe2AzcZq2t66PsQuAHwAKgENgPrLbW/i5dQQ8mmj+ZaP7kkTpdzlkSj7PklLVSRURyn9f+xFXAtcAioBz3EQvrE4mytxLgN8D5wCTgPmCdMeaS4YcrIiKSGV4T4grgQetqw02QBrisd0Fr7XPW2hprbaO1NmatfRrY1VfZbKVBJyIiuWfQhGiMKQZmAK+f2GatbcbtCp3vof5HgPOAfx96mJmXyuAQDToREck9Xu4hnnjqbHOv7c1J+/pkjCkCfgest9a+lGpwu3btOuVzRUVFqofwLJWBJ6kOOqmtrfV03ExeXybp+lzZeH1ery3XZeN3B/r+0s1LQmxNvBb32j4pad9pEi3L/wscApYPJbh58+ZRUFAwlKoZleqgk2z9ZfNK15e9cvnaxoLe3180Gj2tIdGfjs6uhnAovywTcSWOfzgcyvc+vB0wxnwVuB04FzgG7AC2ALcAs6y18aSyecB7wNestf8nHTEPmhCttS3GmDrcUaOvJwIpBuYkgj2NMWYK8DywD7jeWtudjmBFRCQ9wqH8ssHXhh663b+9P6Vka4y5Hfge7piV54FO4PO4gzknAYuBl5OqfB6IAxvSEC7gfR7iGuAOY8xGoB54ANiLO/3iFMaYcuBFYDtwi7U2d57MKyIiaZdoZP0IuKnXFL31uDMainF7Gl9O2rccWJfOBpfXUaYPAk/hJsD3gVnAldbamDGm0hjTZoyZnih7G+6Ui2VAa2JfmzFmTbqCFhGRnPIJIAz8Wz/7q4GrjTGFcDKBfjGxPW08tRATrbw7Ez+9920CipI+VwGajyAiIl5NARr7a+1Za7cYYw4DVwHrgKXAXmttn7fthkoLfYqIiN+OAKWJgTL9qeHDAZrXJz6nVU4mxHhHK917/kD33ueJdx73OxwRERnYa0AU+PIAZZ4A/pMx5hPAQuDX6Q4i5xb3Boi31ENXxH3f+h5O6Ud9jkhERPqTmM3wP4CfGWO6gT8CXcBngE9ba1dZa982xmwGngResNam/ckCOdlCdCZ+BPIKIK8QZ0JK02BERMQH1tp/wp2DeDfwAXAQ+BbwTFKxatyV09LeXQo52kJ0CovJO+8LfochIjJqdXR2HU51rmCqxw+H8lOqY639NQN0hVpr1wJrhxXYAHIyIYqIyMBSXUVmCMfP5OEzIie7TEVERFKlhCgiIoISooiICKCEKCIiAighioiIAEqIIpJm9fX1HDt2zO8wRFKmhCgiafPUU09x+eWX87nPfY7Dhw/7HY5ISpQQRSRttm/fDkBTUxP79u3zORqR1GhivoikzW233UZdXR2NjY1s3ryZT33qUziO43dY0ofujo6GvHA4YyvVdHd0HM4Lhwed/G+MaUv6OA53ke+exOfbEqvXeGaMeRn4lbX2F6nUAyVEEUmDo0eP8sgjj3DmmWcyY8YMdu7cyS9/+Uu2bt3Kz3/+c8rL07MoSlVVFdXV1dxwww3cc889aTnmWJUXDpfVpOl76cvyhgZPydZae/J5usaYt4FbrLUvZiisASkhisiwPfzww/z61+4/5JcuXXpyu7WWp59+mm9+85v91u3p6CAYDns6T3V1NdFolJqamkETYirHldHHGBMAVgFfByYBLwErrLVNxpgw8Avg74EgsA/4AvAPQCWw0BjzELDWWvstr+dUQhSRYZswYcLJ92+99Rbz58/njTfewHEcFi5cOGDdYDiM15bKxY7DNsdhQSQyaJ3lDWl/OpCMrP+G+3zExbhPv/hn4GfAtcANQDFwNm4X63yg3Vr7fWPMpxhil6kG1YjIsF111VUn3+/fv5/GxkYuueQSNm3aREVFRdrOsyQe595YjCXxeNqOKaPWCuD71tp3rbVR4IfA1caYPNxnJU4BzrHW9lhra621rcM9oVqIIjJsM2fO5N577+Uvf/kLb775Jk1NTXR2djJx4kS/Q5PsNQP4N2NMLGlbD1AGPIHbOvxfxphJwK9wk2fXcE6ohCgiabF06VKWLl3K8uXL/Q5FcsNB4GvW2i397K8CqowxM4HnAAv8CzDk7gN1mYrIoKKd3v/hHYvFaGpqYs+ePRw8eDCDUUmOWwPcZ4yZAWCMmWqM+VLi/aeNMRcYY4JAK24X6omW5GFg9lBOqBaiiAyqIJTP+Vff5als0aGDRCIRAD679OsEz7xowPK7f3v/sOOTnPQ/AQf4ozHmI8D7wG+AZ4Fy3IR5FtCW2P5EUr1qY8w3gCestf/g9YSeEmJi+Otq4GZgPLAZd8JkXR9lC4Ea3FE/c4AfWWt/6DUgEclOsegxcALk5RfgOA7xeBxnXKnfYUk/ujs6DnudKzjU4+elOO3FWjsz6X0M+Enip3e5J4En+znGa8DclE6c4LXLdBXuUNdFuJn5HWB9IlH2Fge2ArcC24cSlIhkl57D/4/Y3j8Ssxvo7IpSXl5O6UdmEZg83e/QpB+JVWScTP14WaVmtPHaZboCeMBaawGMMatw+2kvA15NLmit7QB+mijXkb5QRWS0irceOvn+eDRGQdEU2sNn+BiRSOoGTYjGmGLc4a+vn9hmrW02xuzH7RZ9tb+6w7Vr165TPqdzPtNIqq2t9VRO1zc65fL1pevaAmdcQKxuGzgBYlPOpamodFSsYZrL3x14vz7xxksL8cREouZe25uT9mXEvHnzKCgoyOQpRkS2/rJ5pevLXum6tkDRNALnX0m33UD87VfpKSwh75xPp+XYw5HL3x2cfn3RaPS0hoR45+Ue4onZ/8W9tk9K2iciY1ysuws6j7sf2pv8DUZkCAZNiNbaFqAOWHBiW6IbdQ6wI3OhiUg2cYJ5EBrvfhg/1d9gRIbA66CaNcAdxpiNQD3wALAXd/rFaYwxBbgjjQJAXmJl8pi1tnP4IYvIaOQ4DsG5fwfdHTj5hX6HI5Iyr9MuHgSewk2A7wOzgCuttTFjTKUxps0Ykzy+2gLtuI/h+H7i/R/TF7aIjEaO4ygZStby1EJMTJC8M/HTe98moKjXtpnpCE5ERGSkaC1TERERlBBFREQAJUQRERFACVFERARQQhQREQGUEEVERAAlRBEREUAJUUREBFBCFBERAZQQRUREACVEERERQAlRREQEUEIUEREBlBBFREQAJUQRERFACVFERARQQhQREQGUEEVERAAlRBEREUAJUUREBFBCFBERAZQQRUREAMjzUsgYEwBWAzcD44HNwG3W2rp+yv8N8DNgPtAI/KO19p/TErGIiEgGeG0hrgKuBRYB5cA7wPpEojyFMWYisAF4HigBlgI/NMZcnZaIRUREMsBrQlwBPGhdbbgJ0gCX9VH2PwM9wL3W2g5r7TbgceC/piNgERGRTBi0y9QYUwzMAF4/sc1a22yM2Y/bJfpqryoXAn+11saStr0OfD2FuIIAnZ2dp+0oLR6XwmG8i0aj5E+blpHjpiIT15epaztx7FTo+oYWQ67+v3kiDl3f0I7bW9LfzGDaTzgGOPF4fMACxpizcbtI51pr9yVt3wL8wVq7ulf5fwHC1tr/krTts4mynu5Z1tbWXgZs8nwVIiKSrLKiomKz30FkGy8JqjXxWtxr+6Skfb3Ll3ks258/A5XAIdzuVxERGVwQOAP3b6ikaNCEaK1tMcbUAQtIdJsmulHnADv6qPLvwFJjTCCp27Sin7J9qqioiOKOZBURkdS86XcA2crroJo1wB3GmLnGmPHAA8Be+k5av8NNtN83xhQYYy7BvX/4aDoCFhERyQSvCfFB4CncBPg+MAu40lobM8ZUGmPajDHTAay1rcDngSuAZuBpoMpa+7/THr2IiEiaDDqoRkREZCzQ0m0iIiIoIYqIiABKiCIiIoASooiICKCEKCIiAighioiIAEqIIiIigBKiiIgIoIQoIiICwP8HWmX5g7vbHDYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "type_col = {'CV': '#1D5386', 'Test': '#B80000'}\n",
    "\n",
    "fig2 = plt.figure(figsize=(6,3))\n",
    "sns.set(style=\"whitegrid\", color_codes=True,font_scale=1.2)\n",
    "# ax = sns.boxplot(x='rate', y='r2', hue='type', data=params_opt,\n",
    "#                  whis=[10, 90], palette=hist_type_col,flierprops=dict(alpha=0))\n",
    "ax = sns.barplot(x='rate', y='r2', hue='type', data=params_opt,palette=type_col)\n",
    "ax = sns.stripplot(x='rate', y='r2', hue='type', data=params_opt,\n",
    "                   jitter=True,dodge=True, color='k', alpha=1,size=3,linewidth=0)      \n",
    "handles, labels = ax.get_legend_handles_labels()# get legend information from the plot object\n",
    "fs=12\n",
    "l = plt.legend(handles[2:4], labels[2:4],loc='center left', bbox_to_anchor=(1, 0.3),fontsize=fs,frameon=False)\n",
    "plt.ylim([0, 0.5])\n",
    "\n",
    "plt.xticks(rotation = 60)\n",
    "plt.xlabel('', fontsize=fs)\n",
    "plt.ylabel('',fontsize=fs)#'R^2'\n",
    "ax.set(xticklabels=[])#Timescales_lab)#\n",
    "\n",
    "\n",
    "filename = 'LASSO2_subcell_cv_test_bar.'#old: box.\n",
    "# plt.savefig(os.path.join(path,'figures',filename+'pdf'), bbox_inches=\"tight\", transparent=True)\n",
    "# plt.savefig(os.path.join(path,'figures',filename+'png'), bbox_inches=\"tight\", transparent=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Bar chart of performance comparison with Agarwal et al\n",
    "Results on Agarwal obtained in: `ML_20220619_subcell_feat_select2.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.0001 ratio 1.214323, r2_train 0.553800 r2_cv 0.456057\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.00033 ratio 1.172324, r2_train 0.524747 r2_cv 0.447613\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.00066 ratio 1.143363, r2_train 0.494857 r2_cv 0.432808\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.001 ratio 1.122715, r2_train 0.471627 r2_cv 0.420078\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.0033 ratio 1.057951, r2_train 0.410384 r2_cv 0.387905\n",
      "INFO: [2022-06-28 23:00:42] ML - 0.0033 r2_opt 0.387905\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.0066 ratio 1.038999, r2_train 0.385875 r2_cv 0.371391\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.01 ratio 1.033803, r2_train 0.367385 r2_cv 0.355372\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.033 ratio 1.024164, r2_train 0.276878 r2_cv 0.270345\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.066 ratio 1.012717, r2_train 0.185475 r2_cv 0.183146\n",
      "INFO: [2022-06-28 23:00:42] ML - k_whole_cell 0.1 ratio 1.009604, r2_train 0.156522 r2_cv 0.155033\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.0001 ratio 1.338998, r2_train 0.704373 r2_cv 0.526045\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.00033 ratio 1.292327, r2_train 0.701852 r2_cv 0.543092\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.00066 ratio 1.246088, r2_train 0.696185 r2_cv 0.558696\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.001 ratio 1.211383, r2_train 0.689778 r2_cv 0.569414\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.0033 ratio 1.119138, r2_train 0.664773 r2_cv 0.594004\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.0066 ratio 1.078329, r2_train 0.647541 r2_cv 0.600504\n",
      "INFO: [2022-06-28 23:00:42] ML - 0.0066 r2_opt 0.600504\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.01 ratio 1.059827, r2_train 0.635911 r2_cv 0.600014\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.033 ratio 1.030264, r2_train 0.600780 r2_cv 0.583132\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.066 ratio 1.021849, r2_train 0.573472 r2_cv 0.561210\n",
      "INFO: [2022-06-28 23:00:42] ML - Agarwal whole cell half lives 0.1 ratio 1.020290, r2_train 0.548061 r2_cv 0.537162\n",
      "INFO: [2022-06-28 23:00:42] ML -                              rate   alpha  type        r2\n",
      "0                    k_whole_cell  0.0033    CV  0.357544\n",
      "1                    k_whole_cell  0.0033    CV  0.430740\n",
      "2                    k_whole_cell  0.0033    CV  0.392660\n",
      "3                    k_whole_cell  0.0033    CV  0.374558\n",
      "4                    k_whole_cell  0.0033    CV  0.418138\n",
      "5                    k_whole_cell  0.0033    CV  0.384341\n",
      "6                    k_whole_cell  0.0033    CV  0.368345\n",
      "7                    k_whole_cell  0.0033    CV  0.378282\n",
      "8                    k_whole_cell  0.0033    CV  0.388390\n",
      "9                    k_whole_cell  0.0033    CV  0.386049\n",
      "10                   k_whole_cell  0.0033  Test  0.378828\n",
      "11  Agarwal whole cell half lives  0.0066    CV  0.594379\n",
      "12  Agarwal whole cell half lives  0.0066    CV  0.595792\n",
      "13  Agarwal whole cell half lives  0.0066    CV  0.641116\n",
      "14  Agarwal whole cell half lives  0.0066    CV  0.585839\n",
      "15  Agarwal whole cell half lives  0.0066    CV  0.597028\n",
      "16  Agarwal whole cell half lives  0.0066    CV  0.565812\n",
      "17  Agarwal whole cell half lives  0.0066    CV  0.583507\n",
      "18  Agarwal whole cell half lives  0.0066    CV  0.629594\n",
      "19  Agarwal whole cell half lives  0.0066    CV  0.625688\n",
      "20  Agarwal whole cell half lives  0.0066    CV  0.586286\n",
      "21  Agarwal whole cell half lives  0.0066  Test  0.569193\n",
      "22   Agarwal model and half lives     NaN    CV  0.608400\n",
      "INFO: [2022-06-28 23:00:42] ML - Write final params_opt to file LASSOagarwal_params_opt.tsv\n"
     ]
    }
   ],
   "source": [
    "agar_compare = dict()\n",
    "agar_compare['rate'] = []\n",
    "agar_compare['alpha'] = []\n",
    "agar_compare['type'] = []\n",
    "agar_compare['r2'] = []\n",
    "\n",
    "agar_types = [Timescales[-1], 'Agarwal whole cell half lives', 'Agarwal model and half lives']\n",
    "for ts in agar_types[:2]:\n",
    "    alpha_opt, r2_cv_opt = get_cv_alpha_opt(ts)\n",
    "\n",
    "    #add data to dataframe used for boxplot\n",
    "    agar_compare['rate'].extend([ts for _ in range(cv_folds)])\n",
    "    agar_compare['alpha'].extend([alpha_opt for _ in range(cv_folds)])\n",
    "    agar_compare['type'].extend(['CV'  for _ in range(cv_folds)])\n",
    "    agar_compare['r2'].extend(r2_cv_opt)\n",
    "\n",
    "    agar_compare['rate'].append(ts)\n",
    "    agar_compare['alpha'].append(alpha_opt)\n",
    "    agar_compare['type'].append('Test')\n",
    "    agar_compare['r2'].extend(RND2[ts+'r2_test'][alpha_opt].values) \n",
    "\n",
    "\n",
    "\n",
    "#add data to dataframe used for boxplot\n",
    "agar_compare['rate'].append('Agarwal model and half lives')\n",
    "agar_compare['alpha'].append(np.nan)\n",
    "agar_compare['type'].append('CV')\n",
    "agar_compare['r2'].append(0.78**2)#Pearson correlation CV r=0.78 taken from biorXiv: BC3MS+BeEM\n",
    "\n",
    "agar_compare = pd.DataFrame.from_dict(agar_compare)\n",
    "agar_compare['rate'] = agar_compare['rate'].astype('category')\n",
    "agar_compare['rate'].cat.reorder_categories(agar_types, inplace=True)\n",
    "agar_compare['type'] = agar_compare['type'].astype('category')\n",
    "agar_compare['type'].cat.reorder_categories(['CV','Test'], inplace=True)\n",
    "\n",
    "logger.info(agar_compare)\n",
    "\n",
    "filename = 'LASSOagarwal_params_opt.tsv'\n",
    "# agar_compare.to_csv(os.path.join(path,subfolder,filename), sep='\\t', index=False) \n",
    "logger.info('Write final params_opt to file %s' % filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATUAAAC+CAYAAABUH5irAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAX9UlEQVR4nO3dfVBTd74G8CdBCNyQKPX6FqBwsVWwKStERBGLrVprrY7eqR2ka7stwcVWd2dqWat3py06OiiKunbBvYVqXbRa2REs7G471h1HrC9A6Wgsqa3NCo3VWqV6wkugSe4fXFMQMAdIIByfz4wzzfl9zznfU+OTk/OSI3M4HA4QEUmEfKAbICJyJ4YaEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSRIVaWVkZUlJSEBsbi/Hjx7usr6urQ2pqKmJiYpCQkIBt27aBl8MRUX8QFWpqtRopKSlYu3aty1qbzYb09HRoNBqUl5ejsLAQH330Ed57770+N0tE5IqoUJs+fTqeeeYZhIaGuqytrKzE5cuXkZGRAaVSiYiICOj1euzfv7/PzRIRuTLE3Qs0Go0ICwuDWq12TtNqtfjuu+9gsVgQGBjochl2ux0NDQ3w9fWFTCZzd4tEveZwONDa2gqlUgm5nIekvZHbQ81isUClUnWYdifgxIZaQ0MDLl686O7WiNxm3Lhxnd7n5B3cHmqBgYGwWCwdpt2+fds5Joavry+AtjeOn5+fexsk6oOWlhZcvHjR+R4l7+P2UIuMjMTly5chCILzk+zChQsICQkRHWp3vnL6+flBoVC4u0WiPuNhEe8l6qCAzWaD1WpFa2srAMBqtcJqtcJut3eqnTRpEh588EFkZ2ejsbERJpMJ+fn5WLJkiXs7JyLqgqhQKykpQXR0NFJTUwEA0dHRiI6ORkVFBa5cuYKYmBhUVlYCAHx8fLBr1y6YzWYkJCQgJSUFzzzzjHNeIiJPknnjj0RarVYYDAZotVp+/SSvwvem9+M5aSKSFIYaEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSGGpEJCkMNSKSFIYaEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSGGpEJCkMNSKSFIYaEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSRIWa3W5HTk4OEhISEBMTg9TUVJjN5m7rjxw5gvnz5yM2NhaPPfYYNmzYgJaWFrc1TUTUHVGhlp+fj9LSUhQWFqK8vBwajQbp6emw2+2dao1GI1avXo1XX30VVVVV+OCDD1BeXo7c3Fy3N09EdDdRoXbgwAHo9XpERERAqVQiIyMDJpMJVVVVnWrr6uowdOhQPPXUU5DJZAgODsaMGTNgNBrd3jwR0d2GuCoQBAFmsxlardY5Ta1WIywsDDU1NYiLi+tQn5iYiJCQEJSVleGpp56C2WzGsWPHsGzZsh43ZzAYejwPEd3fXIaaxWIB0BZk7alUKudYewEBAXj22Wfx1ltvISMjAzabDYsWLcLChQt73JxWq4VCoejxfESeYrVa+WHr5Vx+/QwMDATQtsfWniAIzrH2Dh8+jJycHOzatQsGgwEnTpxAfX09Vq9e7aaWiYi65zLUVCoVgoODO3w6CYKA2tpaREVFdao3GAyIj4/HpEmTIJfLMXLkSDz33HP49NNP3ds5EVEXRJ0oSE5ORkFBAUwmExobG5GdnY3w8HDodLpOtTqdDmfPnkV1dTUcDgdu3LiBDz/8sMMxOSIiT3F5TA0A9Ho9BEFASkoKmpqaoNPpkJeXB7lcjsrKSqSlpaGsrAwajQZPP/00rl+/jjVr1uDatWsICAjA5MmT8fbbb3t4U4iIAJnD4XAMdBN3u3MwlicKyNvwven9eJsUEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSGGpEJCkMNSKSFIYaEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSGGpEJCkMNSKSFIYaEUkKQ20Qu3nzJs6dOwcv/PFiogHDUBukbt++jQULFmDx4sXYvHnzQLdD5DUYaoPU9evXcf36dQDAuXPnBrgbIu/BUBukjh07Bh8fHwCATCYb4G6IvIeoR+SRd6mrq8OWLVucrxsbGwewGyLvImpPzW63IycnBwkJCYiJiUFqairMZnO39c3NzcjKysJjjz2GiRMnYvbs2Th+/Ljbmr7fffPNN/D39wcAjBs3jsfUiNoRtaeWn5+P0tJSFBYWYtSoUcjKykJ6ejpKSkogl3fMRYfDgVdffRUAsG/fPoSGhuLq1av4+eef3d/9fergwYMICgrCmDFjsH//fn79JGpH1J7agQMHoNfrERERAaVSiYyMDJhMJlRVVXWqPXnyJCoqKpCdnY3Q0FAAwOjRoxESEuLezu9jjY2NkMvl8PX1lVSgZWVl4dFHH0VWVtZAt0KDmMtQEwQBZrMZWq3WOU2tViMsLAw1NTWd6k+fPo2QkBDk5eVh2rRpeOKJJ7B+/Xo0NDS4t/P70GuvvYbJkyfj4sWLMJvNMBgMsFqtA92WaNaW1nuO79u3Dy0tLdi/f79bl0v3F5dfPy0WC4C2IGtPpVI5x9qrr6/HpUuXMG3aNBw9ehT19fVYsWIFNm3ahHXr1vWoOYPB0KP6wW585AQEKgO6HKutrUVZWVmHaQ0NDTCZTIiMjLznci0NTfjK+KXb+uwtnU6HR55d2+24TR0G3LiEFtWD96y724WijV1+a6D7k8tQCwwMBNC2x9aeIAjOsfaUSiV8fHzw+uuvQ6FQICAgAGlpaVi/fn2PQ02r1UKhUPRonsGuu3/MDltLp2mKACX++3/ed/kV9ELRRuh0Orf050k+Y6KBMdG9mre/ts9qtd53H7aDjcuvnyqVCsHBwR3+IgVBQG1tLaKiojrVT5gwAUDHa6ekdNxnoMh8/IDA0R2mKYcO5/9boruIOlGQnJyMgoICmEwmNDY2Ijs7G+Hh4V1+Os6ePRvDhw/Htm3b0NLSgmvXriE/Px9z5sxxe/P3G3loHODTtucq8xmCZlX4wDZE5IVEhZper8fcuXORkpKChIQEmM1m5OXlQS6Xo7KyEjExMbhy5QqAtq+f7733HgwGA+Lj47F48WLExsbiD3/4g0c35H4gH+IHn7FJgG8AHD4KWOX/4ZblZmZmIiIiApmZmW5ZHtFAkjm88Cce7hy34DG1zuw/GGG/dgEAIB/9KOQjxrlc5rnCN+Hz/xfrdiUiIgJWqxX+/v64dOlSj/q1NTffc9l368kJALEuFG10+zK7cz+/NwcL3iY1yMjUY4Ab3wCQtf23CD7+/tg7enS343EyGU7LZJjU2HjPuq68cPVqj+qJPI2hNsjI/IdiSNQzbl3mPIcD87xvh52oV/grHUQkKQw1IpIUhhoRSQpDjYgkhaFGRJLCUCMiSWGoEZGkMNSISFIYakQkKQw1IpIUhhoRSQpDjYgkhaFGRJLCUCMiSWGoEZGkMNSISFIYakQkKQw1IpIUyYZacXExpk2bhjfeeANe+GwZIvIQyYba7t278eOPP+Lw4cO4yoeDEN03JBNqn332GebOnYvVq1fDZrNhwYIFkMlkmDp1KkaOHDnQ7RFRPxEVana7HTk5OUhISEBMTAxSU1NhNptdzmcwGPDII49g6dKlfW7UlXfffRfffvstiouLYTQakZqaivPnz2PPnj3w8fHx+PqJyDuICrX8/HyUlpaisLAQ5eXl0Gg0SE9Ph91u73Yeq9WKNWvWIC4uzm3N3susWbMgk8nw0EMPITw8HADg6+vbL+smul9UV1fj97//PRITE6HVahEfH4+XXnoJhw8fxvr16zFhwgT88MMPXc7rcDjw+OOPe3wnR1SoHThwAHq9HhEREVAqlcjIyIDJZEJVVVW382zbtg1TpkyBTqdzW7P38vzzz+Ps2bMoKSnB+fPn8dprr+H06dP9sm4iMawtrYN6/Xv27MGSJUtw69YtvP7669i9ezc2bNiA8PBwvP322wgKCoLNZkNpaWmX8585cwZXrlzBwoUL+9SHKy4fZiwIAsxmM7RarXOaWq1GWFgYampqutwTq6iowL/+9S8UFxcjPz/fvR3fg1qtBtB2kqCmpgZnzpzBiy++iGXLlvVbD0TdUfj54pFn1w7Y+i8Ubez1vBUVFcjKysKvf/1r/PGPf+wwNmvWLLz00ktobGzExx9/jOLiYrz88sudllFSUoKAgADMmTOn132I4TLULBYLgF8C4w6VSuUca6+hoQFr167Fxo0bERAQ0KfmDAZDj+odDge2b9+OiooKyGQyOBwObN26FTKZDLGxsX3qpT/0116tu91rj709T26f2B6od959910MHToUGRkZXY4/+OCDAIBFixZh06ZNMBqNiIyMdI43NTXh448/xqxZsxAYGOjRXl2G2p0GBEHoMF0QhC6b27RpE5KSktxyLE2r1UKhUIiur62tRUVFBQB0uDYtNzcXn3zyCUaMGNHnnqgzbwjj/urBarX2+MN2sLPZbDhz5gxmzZrl8t/j/PnzsWXLFpSUlHQItaNHj6KhoQGLFi3ydLuuj6mpVCoEBwd3+IsUBAG1tbWIiorqVF9eXo7i4mLEx8cjPj4e+fn5+PzzzxEfH4+6uro+NevqmMCaNWu6nN7U1ISGhoY+LRsAMjMzERERgczMTJe1RFJRX1+P5uZmaDQal7UjRoxAYmIiSktLYbPZnNOLi4sxatQoTJ061ZOtAhCxpwYAycnJKCgowJQpUzBq1ChkZ2cjPDy8y0/HgwcPdtiY3bt344svvsCOHTv6vKfk6pjEzxeNHV77+PjC5uMHDH8Y817/33su+1zhmwDufbb0/fffh9Vqxd69e/HWW2+J6tnW3Awff39RtURSsHDhQhw/fhynTp1CYmIifvjhB5w6dQovv/wy5HLPXxorKtT0ej0EQUBKSgqampqg0+mQl5cHuVyOyspKpKWloaysDBqNplNwBQYGws/PD6NHj/bIBrTnExoP27UvAeF7AIDN1grYWgGb1fW8/v7Y66LHOJkMp2UyTGpsdFl7xwu8m4EGuWHDhsHf3x9XrlwRVT9z5kyo1WoUFxcjMTERH330EWw2W7989QREhppcLseqVauwatWqTmOTJk1CdXV1t/OuXLmy9931kCxgGIaEJ8DR0gDblXOAcMU53R3mORyYx/tI6T4zZMgQTJ48GSdPnkRLSwv8/PzuWa9QKDB37lwcOXIEDQ0NKCkpgVarxdixY/ulX8ncJtWezE8JNPxyAaBMNWYAuyEa/JYtW4affvoJmzdv7nK8rq4ORuMvh38WLVqEpqYm7Ny5E1999VW/7aUBIvfUBiPZ8LFw/PgNZA/8F2Qy2UC3QzSoxcXF4Y033kBWVhYuXbqERYsWQaPR4NatWzh16hSKioqwZcsW5xnPmJgYhIeHY8+ePfD19cW8efP6rVfJhprPaC0wWuu6kIhE+c1vfoPo6Gjs2bMHmzdvRn19PZRKJbRaLTIzM/HEE090qF+4cCG2b9+OpKQkBAUF9Vufkg01Im9jbWnt01X97li/wq9v90PHxsaKvpB9+fLlWL58eZ/W1xuSPKZG5I36GiiDff39haFGRJLCUCMiSWGoEZGkMNSISFIYakQkKQw1IpIUhhoRSQpDjYgkhaFGRJLC26SIqFvjx493WRMcHIxjx471aT01NTU4evQoli5dimHD+vZTYQw1on4y0L+C3Jv1Hzx4sMPrFStWYPz48R1+J9HV76uJUVNTg3feeQcLFixgqBENFmJ+XdmTevMrzBMnTuzw2s/PD0FBQZ2mexOGGhH1SV1dHbZv346TJ0/CYrFg7NixWLFiBWbPnu2sMZlM2LJlCz7//HNYLBYMHz4c0dHRyMnJwZEjR5wPTXryySed83z66acICQnpcT8MNSLqte+//x7PPfcchg8fjjVr1uCBBx7A3//+d6xcuRJ//vOfMXPmTADAb3/7W6jVaueT3K9du4bjx4/DbrdjxowZWL58OfLy8rBjxw7n80xGjhzZq54YakTUazt37oTD4cBf//pX5w9BTp8+HVevXsWf/vQnzJw5Ezdv3sTly5eRm5vrDDmg7RmhAPDAAw84H4YcFRWFsLCwPvXESzqIqNdOnDiBpKQkqFQq/Pzzz84/iYmJMBqNsFgsCAoKQmhoKLZu3YoPP/wQ//73vz3aE/fUiKjXbt68ieLiYhQXF3c5Xl9fj8DAQOzevRs7d+7E1q1b8dNPPyEkJASpqalISUlxe08MNSLqtWHDhkGn0yEtLa3L8VGjRgEAQkNDsXnzZjgcDhiNRhQWFiIzMxPBwcFISkpya0+iQs1ut2P79u0oKipCU1MTYmNjsW7dOgQHB3eq/eKLL5CbmwuDwYDm5maEhYVh+fLlHc5qEJE0TJ8+HdXV1Xj44YfhL+IaOJlMhqioKKxZswZFRUX4+uuvkZSU5LzWrbm5uc89iTqmlp+fj9LSUhQWFqK8vBwajQbp6emw2+2dam/duoWnn34apaWlqKysRHp6OlatWoVz5871uVki8i6/+93vYLFY8Pzzz+Pw4cM4e/Ysjh49itzcXOdlGkajEUuXLsUHH3yAzz77DCdOnMCbb76JIUOGYMqUKQCAhx56CACwb98+VFdX4/z582hpaelVT6L21A4cOAC9Xo+IiAgAQEZGBhISElBVVYW4uLgOtXfvSs6ZMwd/+ctfUFVVhejo6F41SUTeSaPR4G9/+xt27tyJnJwc1NfXY9iwYXj44YexcOFCAMCIESOg0WiwZ88eXL16FQqFAuPGjcOuXbug1bY9xjIyMhIrV67EwYMHcejQIdjtds9dpyYIAsxms3PlAKBWqxEWFoaamppOoXa3a9eu4dtvv3U+5LQnDAZDh9c6na7Hy/AGVVVVouq4fZ7vYSDZmpt7dVW/O9ff19u0urrHc/To0diwYUO38wwfPhybNm1yuewVK1ZgxYoVfeoPEBFqFosFQFuQtadSqZxj3WloaMDKlSvx+OOPY+rUqT1uTqvVQqFQ9Hg+bzNYw0osb9i+/urBarV2+rAVayDv+/SG9fcXl8fUAgMDAbTtsbUnCIJzrCuCIECv12PEiBGiUpqIyB1chppKpUJwcHCHTydBEFBbW4uoqKgu56mvr8eLL76IMWPGYMeOHW65i5+ISAxRZz+Tk5NRUFAAk8mExsZGZGdnIzw8vMtd/uvXr2Pp0qUYP348tmzZgiFDeCkcEfUfUYmj1+shCAJSUlLQ1NQEnU6HvLw8yOVyVFZWIi0tDWVlZdBoNDh48CC+/vprfPfdd/jnP//pXMb8+fOxbt06j20IEREgMtTkcjlWrVqFVatWdRqbNGkSqqurna/ddQaDiKg3eEM7EUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSGGpEJCkMNSKSFIYaEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSGGpEJCkMNSKSFIYaEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSRIWa3W5HTk4OEhISEBMTg9TUVJjN5m7rv/zySyQnJ+NXv/oVZsyYgb1797qtYSKiexEVavn5+SgtLUVhYSHKy8uh0WiQnp4Ou93eqdZisUCv1yMxMRFnz57F9u3b8c4773R4sDERkaeICrUDBw5Ar9cjIiICSqUSGRkZMJlMqKqq6lT7ySefQC6X45VXXoFCocDEiROxePFi7N+/3+3NExHdzeUT2gVBgNlshlardU5Tq9UICwtDTU0N4uLiOtQbjUZMmDABcvkveanVanHo0CHRTTkcDgBAS0tLp7H/HPofopfTE1arFb4jR3pkuT3hie3z1LbdWXZPeGr7+sud9+Sd9yh5H5ehZrFYALQFWXsqlco5dne9SqXqME2tVndZ253W1lYAwMWLFzuN5bwyV/RyesJgMGDc7t0eWW5PeGL7PLVtd5bdE57avv7W2toKf3//fl8vueYy1AIDAwG07bG1JwiCc+zu+hs3bnSYdvv27S5ru6NUKjFu3Dj4+vpCJpOJno/I0xwOB1pbW6FUKge6FeqGy1BTqVQIDg6GwWDAo48+CqAt0GpraxEVFdWpPjIyEv/4xz9gt9udX0EvXLiAyMhI0U3J5fJOe3tE3oJ7aN5N1ImC5ORkFBQUwGQyobGxEdnZ2QgPD4dOp+tU++STT8JmsyEvLw8tLS04d+4cDh06hCVLlri9eSKiu8kcIo542u12bNu2DUVFRWhqaoJOp0NmZiZCQkJQWVmJtLQ0lJWVQaPRAGi7Ti0zMxM1NTUICgpCamoqXnjhBY9vDBGRqFAjIhoseJsUEUkKQ42IJIWhRkSSwlAjIklhqBGRpDDUiEhSGGpEJCkMNSKSFIYaEUnK/wHyCq9shEvpkgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 216x216 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "type_col = {'CV': '#1D5386', 'Test': '#B80000'}\n",
    "    \n",
    "fig2 = plt.figure(figsize=(3,3))\n",
    "sns.set(style=\"whitegrid\", color_codes=True,font_scale=1.2)\n",
    "ax = sns.barplot(x='rate', y='r2', hue='type', data=agar_compare,palette=type_col)\n",
    "ax = sns.stripplot(x='rate', y='r2', hue='type', data=agar_compare,\n",
    "                   jitter=True,dodge=True, color='k', alpha=1,size=3,linewidth=0)      \n",
    "handles, labels = ax.get_legend_handles_labels()# get legend information from the plot object\n",
    "fs=16\n",
    "l = plt.legend(handles[2:4], labels[2:4],loc='center left', bbox_to_anchor=(1, 0.3),fontsize=fs,frameon=False)\n",
    "plt.ylim([0, 1])\n",
    "\n",
    "plt.xticks(rotation = 75)\n",
    "plt.xlabel('', fontsize=fs)\n",
    "plt.ylabel('',fontsize=fs)#'R2'\n",
    "ax.set(xticklabels=[])#['Our model, our rates', 'Our model, Agarwal rates', 'Agarwal model and rates'])#\n",
    "\n",
    "\n",
    "filename = 'LASSO2_Agarwal_comparison_bar.'\n",
    "# plt.savefig(os.path.join(path,'figures',filename+'pdf'), bbox_inches=\"tight\", transparent=True)\n",
    "# plt.savefig(os.path.join(path,'figures',filename+'png'), bbox_inches=\"tight\", transparent=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature selection analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rate</th>\n",
       "      <th>alpha</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr</td>\n",
       "      <td>0.0066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nuc</td>\n",
       "      <td>0.0033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>nucexp_from_chr</td>\n",
       "      <td>0.0033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>cyto</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>poly_entry</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>whole_cell</td>\n",
       "      <td>0.0033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               rate   alpha\n",
       "0               chr  0.0066\n",
       "11              nuc  0.0033\n",
       "22  nucexp_from_chr  0.0033\n",
       "33             cyto   0.001\n",
       "44       poly_entry    0.01\n",
       "55       whole_cell  0.0033"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "params_opt[['rate', 'alpha']].drop_duplicates() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "rel_feat = dict()\n",
    "for idx in params_opt[['rate', 'alpha']].drop_duplicates().index:\n",
    "    ts = params_opt['rate'][idx]\n",
    "    a_ = params_opt['alpha'][idx]\n",
    "    rel_feat[ts] = RND2[rt+ts+a_]   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 6C: heatmap of feature class contributions \n",
    "\n",
    "- Load feature files: (takes a while)\n",
    "- Load MSigDB to split according to TF vs miRNA \n",
    "- Process data and write to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-06-28 23:01:18] ML - Load relevant feature files whole_cell:\n",
      "INFO: [2022-06-28 23:01:18] ML - ['gene_structure_v2.txt', 'gene_structure_agarwal.txt', 'gene_sequence_v2.txt', 'gene_sequence_agarwal.txt', 'Codon_features.tsv.gz', 'kmer_ORF_features.tsv.gz', 'kmer_5UTR_features.tsv.gz', 'kmer_3UTR_features.tsv.gz', 'polyA_tail_length_data.txt', 'RNA_modifications_v2.txt', 'gene_location_v2.txt', 'histone_modifications_v1.txt', 'microRNA_targets_v2.txt', 'RBP_targets_filter_v2.txt', 'TF_scRNAseq_v1.csv', 'TF_MSigDB_targets_v2.tsv']\n",
      "INFO: [2022-06-28 23:01:18] ML - gene_structure_v2.txt 19955\n",
      "INFO: [2022-06-28 23:01:18] ML - Index(['Gene', 'Symbol', 'pre_mRNA_length_max', 'n_exons_max',\n",
      "       'mRNA_length_max', '3UTR_proportion_max', '5UTR_proportion_max'],\n",
      "      dtype='object')\n",
      "INFO: [2022-06-28 23:01:18] ML -               Gene    Symbol  pre_mRNA_length_max  n_exons_max  \\\n",
      "0  ENSG00000000003    TSPAN6               9995.0            8   \n",
      "1  ENSG00000000005      TNMD              14949.0            7   \n",
      "2  ENSG00000000419      DPM1              23678.0           10   \n",
      "3  ENSG00000000457     SCYL3              44254.0           14   \n",
      "4  ENSG00000000460  C1orf112             191062.0           29   \n",
      "\n",
      "   mRNA_length_max  3UTR_proportion_max  5UTR_proportion_max  \n",
      "0             3789             0.765426             0.348264  \n",
      "1             1198             0.139399             0.068447  \n",
      "2             1149             0.248120             0.026980  \n",
      "3             6292             0.648284             0.125969  \n",
      "4             4333             0.348894             0.190424  \n",
      "INFO: [2022-06-28 23:01:18] ML - gene_structure_agarwal.txt 12981\n",
      "INFO: [2022-06-28 23:01:18] ML - Index(['Gene', 'UTR5LEN', 'CDSLEN', 'INTRONLEN', 'UTR3LEN', 'ORFEXONDENSITY'], dtype='object')\n",
      "INFO: [2022-06-28 23:01:18] ML -               Gene   UTR5LEN    CDSLEN  INTRONLEN   UTR3LEN  ORFEXONDENSITY\n",
      "0  ENSG00000000003  2.045714  2.862191   3.774013  3.131651        0.987666\n",
      "1  ENSG00000000457  2.342620  3.345001   4.577654  2.804889        0.775974\n",
      "2  ENSG00000000460  2.843295  3.404338   4.735160  3.038262        0.943000\n",
      "3  ENSG00000000938  2.456518  3.197584   4.311798  2.928959        0.850033\n",
      "4  ENSG00000000971  2.378580  3.564796   4.961545  2.278982        0.784617\n",
      "INFO: [2022-06-28 23:01:18] ML - gene_sequence_v2.txt 19908\n",
      "INFO: [2022-06-28 23:01:18] ML - Index(['Gene', 'Symbol', 'ARE_motif_n_intronic', 'ARE_motif_max_intronic',\n",
      "       'ARE_motif_n_3UTR', 'ARE_motif_n_total', 'TOP_motif_predicted',\n",
      "       'TOP_motif_strong', 'codon_CG3_frequency', 'sequence_CG_content',\n",
      "       'gene_GC_content', 'median_5SS', 'max_5SS', 'min_5SS', 'median_3SS',\n",
      "       'max_3SS', 'min_3SS'],\n",
      "      dtype='object')\n",
      "INFO: [2022-06-28 23:01:18] ML -               Gene    Symbol  ARE_motif_n_intronic  ARE_motif_max_intronic  \\\n",
      "0  ENSG00000000003    TSPAN6                     3                       1   \n",
      "1  ENSG00000000005      TNMD                     2                       1   \n",
      "2  ENSG00000000419      DPM1                     7                       5   \n",
      "3  ENSG00000000457     SCYL3                     4                       1   \n",
      "4  ENSG00000000460  C1orf112                    21                       5   \n",
      "\n",
      "   ARE_motif_n_3UTR  ARE_motif_n_total  TOP_motif_predicted  TOP_motif_strong  \\\n",
      "0                 1                  4                  0.0               0.0   \n",
      "1                 0                  2                  0.0               0.0   \n",
      "2                 0                  7                  0.0               0.0   \n",
      "3                 0                  4                  0.0               0.0   \n",
      "4                 0                 21                  0.0               0.0   \n",
      "\n",
      "   codon_CG3_frequency  sequence_CG_content  gene_GC_content  median_5SS  \\\n",
      "0             42.85714             41.19241            40.40       9.850   \n",
      "1                  NaN                  NaN            40.78       9.590   \n",
      "2             40.06969             42.59259            40.20       8.540   \n",
      "3             42.99191             45.04262            40.14       9.250   \n",
      "4                  NaN                  NaN            39.22       8.855   \n",
      "\n",
      "   max_5SS  min_5SS  median_3SS  max_3SS  min_3SS  \n",
      "0    11.81     6.04       9.320    12.07     6.84  \n",
      "1    10.74     5.56       9.235    10.10     6.62  \n",
      "2    10.28     0.85       7.420    10.19     5.24  \n",
      "3    10.57     6.38       8.410    12.75     3.22  \n",
      "4    11.08   -13.36       8.310    11.35    -3.91  \n",
      "INFO: [2022-06-28 23:01:18] ML - gene_sequence_agarwal.txt 12981\n",
      "INFO: [2022-06-28 23:01:18] ML - Index(['Gene', 'UTR5GC', 'CDSGC', 'UTR3GC'], dtype='object')\n",
      "INFO: [2022-06-28 23:01:18] ML -               Gene  UTR5GC  CDSGC  UTR3GC\n",
      "0  ENSG00000000003   0.625  0.414   0.368\n",
      "1  ENSG00000000457   0.613  0.451   0.369\n",
      "2  ENSG00000000460   0.411  0.421   0.355\n",
      "3  ENSG00000000938   0.702  0.587   0.508\n",
      "4  ENSG00000000971   0.396  0.398   0.277\n",
      "INFO: [2022-06-28 23:01:19] ML - Codon_features.tsv.gz 12981\n",
      "INFO: [2022-06-28 23:04:57] ML - kmer_ORF_features.tsv.gz 12981\n",
      "INFO: [2022-06-28 23:08:33] ML - kmer_5UTR_features.tsv.gz 12981\n",
      "INFO: [2022-06-28 23:12:13] ML - kmer_3UTR_features.tsv.gz 12981\n",
      "INFO: [2022-06-28 23:12:13] ML - polyA_tail_length_data.txt 19908\n",
      "INFO: [2022-06-28 23:12:13] ML - Index(['Gene', 'Symbol', 'tail_length_median_chromatin',\n",
      "       'tail_length_Q25_chromatin', 'tail_length_Q75_chromatin',\n",
      "       'tail_length_median_cytoplasm', 'tail_length_Q25_cytoplasm',\n",
      "       'tail_length_Q75_cytoplasm', 'tail_length_median_polysome',\n",
      "       'tail_length_Q25_polysome', 'tail_length_Q75_polysome',\n",
      "       'tail_length_median_total', 'tail_length_Q25_total',\n",
      "       'tail_length_Q75_total'],\n",
      "      dtype='object')\n",
      "INFO: [2022-06-28 23:12:13] ML -               Gene    Symbol  tail_length_median_chromatin  \\\n",
      "0  ENSG00000000003    TSPAN6                           NaN   \n",
      "1  ENSG00000000005      TNMD                           NaN   \n",
      "2  ENSG00000000419      DPM1                     171.60860   \n",
      "3  ENSG00000000457     SCYL3                     182.75910   \n",
      "4  ENSG00000000460  C1orf112                     202.80385   \n",
      "\n",
      "   tail_length_Q25_chromatin  tail_length_Q75_chromatin  \\\n",
      "0                        NaN                        NaN   \n",
      "1                        NaN                        NaN   \n",
      "2                  92.632325                   201.9753   \n",
      "3                 157.243900                   199.9289   \n",
      "4                 111.361925                   257.7499   \n",
      "\n",
      "   tail_length_median_cytoplasm  tail_length_Q25_cytoplasm  \\\n",
      "0                           NaN                        NaN   \n",
      "1                           NaN                        NaN   \n",
      "2                     84.017651                  47.249752   \n",
      "3                    160.667475                  96.201533   \n",
      "4                     98.142712                  54.541255   \n",
      "\n",
      "   tail_length_Q75_cytoplasm  tail_length_median_polysome  \\\n",
      "0                        NaN                          NaN   \n",
      "1                        NaN                          NaN   \n",
      "2                 138.207815                     66.56940   \n",
      "3                 187.449628                     79.81413   \n",
      "4                 159.705791                     69.59281   \n",
      "\n",
      "   tail_length_Q25_polysome  tail_length_Q75_polysome  \\\n",
      "0                       NaN                       NaN   \n",
      "1                       NaN                       NaN   \n",
      "2                 40.075025                117.122650   \n",
      "3                 26.467205                133.482175   \n",
      "4                 39.193197                 97.463035   \n",
      "\n",
      "   tail_length_median_total  tail_length_Q25_total  tail_length_Q75_total  \n",
      "0                       NaN                    NaN                    NaN  \n",
      "1                       NaN                    NaN                    NaN  \n",
      "2                 97.901155              50.202600             137.053525  \n",
      "3                192.355450             171.947975             212.762925  \n",
      "4                 87.357800              53.385880             157.308900  \n",
      "INFO: [2022-06-28 23:12:13] ML - RNA_modifications_v2.txt 19908\n",
      "INFO: [2022-06-28 23:12:13] ML - Index(['Gene', 'Symbol', 'm6A_n_peaks_mean'], dtype='object')\n",
      "INFO: [2022-06-28 23:12:13] ML -               Gene    Symbol  m6A_n_peaks_mean\n",
      "0  ENSG00000000003    TSPAN6               0.0\n",
      "1  ENSG00000000005      TNMD               0.5\n",
      "2  ENSG00000000419      DPM1               0.0\n",
      "3  ENSG00000000457     SCYL3               4.5\n",
      "4  ENSG00000000460  C1orf112               0.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-06-28 23:12:13] ML - gene_location_v2.txt 19908\n",
      "INFO: [2022-06-28 23:12:13] ML - Index(['Gene', 'Symbol', 'chr_band'], dtype='object')\n",
      "INFO: [2022-06-28 23:12:13] ML -               Gene    Symbol  chr_band\n",
      "0  ENSG00000000003    TSPAN6    Xq22.1\n",
      "1  ENSG00000000005      TNMD    Xq22.1\n",
      "2  ENSG00000000419      DPM1  20q13.13\n",
      "3  ENSG00000000457     SCYL3    1q24.2\n",
      "4  ENSG00000000460  C1orf112    1q24.2\n",
      "INFO: [2022-06-28 23:12:13] ML - histone_modifications_v1.txt 88219\n",
      "INFO: [2022-06-28 23:12:13] ML - Index(['Gene', 'Symbol', 'Histone_mod'], dtype='object')\n",
      "INFO: [2022-06-28 23:12:13] ML -               Gene    Symbol     Histone_mod\n",
      "0  ENSG00000000419      DPM1  H3K79me2-human\n",
      "1  ENSG00000000419      DPM1  H3K36me3-human\n",
      "2  ENSG00000000457     SCYL3  H3K36me3-human\n",
      "3  ENSG00000000460  C1orf112   H3K9me3-human\n",
      "4  ENSG00000000460  C1orf112    H3K9ac-human\n",
      "INFO: [2022-06-28 23:12:13] ML - microRNA_targets_v2.txt 100763\n",
      "INFO: [2022-06-28 23:12:13] ML - Index(['Gene', 'Symbol', 'miRNA_family'], dtype='object')\n",
      "INFO: [2022-06-28 23:12:13] ML -               Gene Symbol miRNA_family\n",
      "0  ENSG00000120088  CRHR1      GGCAGUG\n",
      "1  ENSG00000164733   CTSB      UUGGCAC\n",
      "2  ENSG00000164733   CTSB      UGGCACU\n",
      "3  ENSG00000164733   CTSB      GGAAGAC\n",
      "4  ENSG00000164733   CTSB      UUAUUGA\n",
      "INFO: [2022-06-28 23:12:13] ML - RBP_targets_filter_v2.txt 105583\n",
      "INFO: [2022-06-28 23:12:13] ML - Index(['Gene', 'Symbol', 'RBP'], dtype='object')\n",
      "INFO: [2022-06-28 23:12:13] ML -               Gene Symbol     RBP\n",
      "0  ENSG00000000419   DPM1   PRPF8\n",
      "1  ENSG00000000419   DPM1    TIA1\n",
      "2  ENSG00000000419   DPM1  LIN28B\n",
      "3  ENSG00000000419   DPM1   DDX3X\n",
      "4  ENSG00000000419   DPM1   KHSRP\n",
      "INFO: [2022-06-28 23:12:13] ML - TF_scRNAseq_v1.csv 28171\n",
      "INFO: [2022-06-28 23:12:13] ML - Index(['Gene', 'Symbol', 'TF_scRNAseq'], dtype='object')\n",
      "INFO: [2022-06-28 23:12:13] ML -               Gene  Symbol TF_scRNAseq\n",
      "0  ENSG00000165997   ARL5B        ADNP\n",
      "1  ENSG00000128989  ARPP19        ADNP\n",
      "2  ENSG00000171456   ASXL1        ADNP\n",
      "3  ENSG00000145734    BDP1        ADNP\n",
      "4  ENSG00000140299   BNIP2        ADNP\n",
      "INFO: [2022-06-28 23:12:13] ML - TF_MSigDB_targets_v2.tsv 596788\n",
      "INFO: [2022-06-28 23:12:13] ML - Index(['Gene', 'Symbol', 'TF_MSigDB'], dtype='object')\n",
      "INFO: [2022-06-28 23:12:13] ML -               Gene    Symbol          TF_MSigDB\n",
      "0  ENSG00000115977      AAK1  ADA2_TARGET_GENES\n",
      "1  ENSG00000131043      AAR2  ADA2_TARGET_GENES\n",
      "2  ENSG00000150967     ABCB9  ADA2_TARGET_GENES\n",
      "3  ENSG00000139826    ABHD13  ADA2_TARGET_GENES\n",
      "4  ENSG00000163322  ABRAXAS1  ADA2_TARGET_GENES\n"
     ]
    }
   ],
   "source": [
    "logger.info('Load relevant feature files %s:' % ts)\n",
    "\n",
    "F_raw = dict()\n",
    "\n",
    "feature_files = []\n",
    "for fk in num2feat_files.keys():\n",
    "    feature_files.extend(num2feat_files[fk].split())       \n",
    "logger.info(feature_files)\n",
    "                         \n",
    "for f in feature_files:\n",
    "    F_raw[f] = pd.read_csv(os.path.join(path, 'features', f), sep='\\t')\n",
    "\n",
    "    logger.info('%s %d' % (f, len(F_raw[f])))\n",
    "    \n",
    "    if not (f in {'Codon_features.tsv.gz',#too many features to list\n",
    "                 'kmer_ORF_features.tsv.gz',\n",
    "                 'kmer_5UTR_features.tsv.gz',\n",
    "                 'kmer_3UTR_features.tsv.gz'}):\n",
    "        logger.info(F_raw[f].columns)\n",
    "        logger.info(F_raw[f].head())   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load original MSigDB files to split TFs from miRNAs\n",
    "#see MSigDB_20220426.ipynb\n",
    "mpath = os.path.join('/n','groups','churchman','ri23','bseq','MSigDB')\n",
    "\n",
    "MSDB = ['h.all','c3.mir.mirdb','c3.tft.gtrd']\n",
    "cols = ['TF_MSigDB', 'Source']\n",
    "cols.extend(['g'+str(i) for i in range(10000)])\n",
    "\n",
    "M = dict()\n",
    "for m in MSDB:\n",
    "    filename = m + '.v7.5.1.symbols.gmt'\n",
    "    if os.path.exists(os.path.join(mpath, filename)):\n",
    "        M[m] = pd.read_csv(os.path.join(mpath, filename),\n",
    "                           names=cols, sep='\\t', header=None, dtype=str)\n",
    "        M[m] = M[m]['TF_MSigDB']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process data and write to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-06-29 00:06:24] ML - Write final params_opt to file LASSOfinal_feat_class_elements_chr.tsv\n",
      "INFO: [2022-06-29 00:06:25] ML - Write final params_opt to file LASSOfinal_feat_class_elements_nuc.tsv\n",
      "INFO: [2022-06-29 00:06:25] ML - Write final params_opt to file LASSOfinal_feat_class_elements_nucexp_from_chr.tsv\n",
      "INFO: [2022-06-29 00:06:27] ML - Write final params_opt to file LASSOfinal_feat_class_elements_cyto.tsv\n",
      "INFO: [2022-06-29 00:06:27] ML - Write final params_opt to file LASSOfinal_feat_class_elements_poly_entry.tsv\n",
      "INFO: [2022-06-29 00:06:27] ML - Write final params_opt to file LASSOfinal_feat_class_elements_whole_cell.tsv\n",
      "INFO: [2022-06-29 00:06:27] ML - Write final summed importances for heatmap to file LASSOfinal_feat_class_heatmap.tsv\n",
      "INFO: [2022-06-29 00:06:27] ML - Write final number of rel feat per class for heatmap to file LASSOfinal_feat_class_n_heatmap.tsv\n"
     ]
    }
   ],
   "source": [
    "feat_heat = dict() #generate one tsv with all summed importances per class (row) and rates (columns)\n",
    "feat_heat_n = dict() #generate one tsv with number of relevant features per class (row) and rates (columns)\n",
    "feat_heat_names = dict()#generate a tsv file per each rate\n",
    "feat_heat['class'] = ['Gene structure','Histone modifications','Gene location',\n",
    "                      'Sequence','Codon frequencies',\n",
    "                      'Poly(A) tail length', 'RNA modifications',\n",
    "                      'microRNA targets','RBP targets','TF targets']\n",
    "feat_heat_n['class'] = feat_heat['class']\n",
    "\n",
    "for ts in rel_feat.keys():\n",
    "    feat_heat[ts] = np.zeros(len(feat_heat['class']))\n",
    "    feat_heat_n[ts] = np.zeros(len(feat_heat_n['class']))\n",
    "    feat_heat_names[ts] = dict()\n",
    "    for cl in feat_heat['class']:\n",
    "        feat_heat_names[ts][cl] = []\n",
    "    for idx in rel_feat[ts].index:\n",
    "        fn = rel_feat[ts]['features'][idx]\n",
    "        imp = rel_feat[ts]['importance'][idx]\n",
    "        if (fn in F_raw['gene_structure_v2.txt'].columns) or \\\n",
    "           (fn in F_raw['gene_structure_agarwal.txt'].columns):\n",
    "            cl = 'Gene structure'\n",
    "        elif fn.replace('Histone_mod_','') in F_raw['histone_modifications_v1.txt']['Histone_mod'].values:\n",
    "            cl = 'Histone modifications'            \n",
    "        elif fn.replace('chr_band_','') in F_raw['gene_location_v2.txt']['chr_band'].values:\n",
    "            cl = 'Gene location'  \n",
    "        elif (fn in F_raw['gene_sequence_v2.txt'].columns) or \\\n",
    "            (fn in F_raw['gene_sequence_agarwal.txt'].columns) or \\\n",
    "            (fn in F_raw['kmer_ORF_features.tsv.gz'].columns) or \\\n",
    "            (fn in F_raw['kmer_5UTR_features.tsv.gz'].columns) or \\\n",
    "            (fn in F_raw['kmer_3UTR_features.tsv.gz'].columns):\n",
    "            cl = 'Sequence'    \n",
    "        elif (fn in F_raw['Codon_features.tsv.gz'].columns):\n",
    "            cl = 'Codon frequencies'\n",
    "        elif fn in F_raw['polyA_tail_length_data.txt'].columns:\n",
    "            cl = 'Poly(A) tail length'                       \n",
    "        elif fn in F_raw['RNA_modifications_v2.txt'].columns:\n",
    "            cl = 'RNA modifications'  \n",
    "        elif (fn.replace('TF_MSigDB_','') in M['c3.mir.mirdb'].values) or \\\n",
    "             (fn.replace('miRNA_family_','') in F_raw['microRNA_targets_v2.txt']['miRNA_family'].values):\n",
    "            cl = 'microRNA targets'            \n",
    "        elif fn.replace('RBP_','') in F_raw['RBP_targets_filter_v2.txt']['RBP'].values:\n",
    "            cl = 'RBP targets' \n",
    "        elif (fn.replace('TF_MSigDB_','') in M['h.all'].values) or \\\n",
    "             (fn.replace('TF_MSigDB_','') in M['c3.tft.gtrd'].values) or \\\n",
    "             (fn.replace('TF_scRNAseq_','') in F_raw['TF_scRNAseq_v1.csv']['TF_scRNAseq'].values):\n",
    "            cl = 'TF targets'\n",
    "        else:\n",
    "            logger.info('class not found for: %s' % fn)\n",
    "        \n",
    "            \n",
    "        feat_heat_names[ts][cl].append(fn)\n",
    "        cl_idx = feat_heat['class'].index(cl)\n",
    "        feat_heat[ts][cl_idx] += imp\n",
    "        feat_heat_n[ts][cl_idx] += 1\n",
    "        \n",
    "    feat_heat_names[ts] = pd.DataFrame.from_dict(feat_heat_names[ts],orient='index')\n",
    "    filename = 'LASSOfinal_feat_class_elements_'+ts+'.tsv'\n",
    "    feat_heat_names[ts].to_csv(os.path.join(path,subfolder,filename), sep='\\t', index=True, header=False) \n",
    "    logger.info('Write final params_opt to file %s' % filename)\n",
    "\n",
    "feat_heat = pd.DataFrame.from_dict(feat_heat) \n",
    "filename = 'LASSOfinal_feat_class_heatmap.tsv'\n",
    "feat_heat.to_csv(os.path.join(path,subfolder,filename), sep='\\t', index=False) \n",
    "logger.info('Write final summed importances for heatmap to file %s' % filename)\n",
    "\n",
    "feat_heat_n = pd.DataFrame.from_dict(feat_heat_n) \n",
    "filename = 'LASSOfinal_feat_class_n_heatmap.tsv'\n",
    "feat_heat_n.to_csv(os.path.join(path,subfolder,filename), sep='\\t', index=False) \n",
    "logger.info('Write final number of rel feat per class for heatmap to file %s' % filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>chr</th>\n",
       "      <th>nuc</th>\n",
       "      <th>nucexp_from_chr</th>\n",
       "      <th>cyto</th>\n",
       "      <th>poly_entry</th>\n",
       "      <th>whole_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gene structure</td>\n",
       "      <td>0.418890</td>\n",
       "      <td>0.436481</td>\n",
       "      <td>0.527294</td>\n",
       "      <td>0.811643</td>\n",
       "      <td>0.208134</td>\n",
       "      <td>0.555656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Histone modifications</td>\n",
       "      <td>0.282092</td>\n",
       "      <td>0.429672</td>\n",
       "      <td>0.432815</td>\n",
       "      <td>0.279175</td>\n",
       "      <td>0.123018</td>\n",
       "      <td>0.354870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gene location</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034203</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.349090</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.186136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sequence</td>\n",
       "      <td>0.244348</td>\n",
       "      <td>0.179599</td>\n",
       "      <td>0.221814</td>\n",
       "      <td>0.365274</td>\n",
       "      <td>0.034131</td>\n",
       "      <td>0.263344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Codon frequencies</td>\n",
       "      <td>0.608907</td>\n",
       "      <td>0.903850</td>\n",
       "      <td>0.657559</td>\n",
       "      <td>0.665511</td>\n",
       "      <td>0.285600</td>\n",
       "      <td>0.699022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Poly(A) tail length</td>\n",
       "      <td>0.569194</td>\n",
       "      <td>0.558944</td>\n",
       "      <td>0.337198</td>\n",
       "      <td>0.548230</td>\n",
       "      <td>0.115533</td>\n",
       "      <td>0.300212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RNA modifications</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.213313</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>microRNA targets</td>\n",
       "      <td>0.090218</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.006080</td>\n",
       "      <td>6.605497</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.373089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RBP targets</td>\n",
       "      <td>1.520364</td>\n",
       "      <td>2.736346</td>\n",
       "      <td>1.966560</td>\n",
       "      <td>3.233984</td>\n",
       "      <td>0.736825</td>\n",
       "      <td>2.674970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TF targets</td>\n",
       "      <td>0.235002</td>\n",
       "      <td>0.034926</td>\n",
       "      <td>1.011789</td>\n",
       "      <td>4.695149</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.828753</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   class       chr       nuc  nucexp_from_chr      cyto  \\\n",
       "0         Gene structure  0.418890  0.436481         0.527294  0.811643   \n",
       "1  Histone modifications  0.282092  0.429672         0.432815  0.279175   \n",
       "2          Gene location  0.000000  0.034203         0.000000  0.349090   \n",
       "3               Sequence  0.244348  0.179599         0.221814  0.365274   \n",
       "4      Codon frequencies  0.608907  0.903850         0.657559  0.665511   \n",
       "5    Poly(A) tail length  0.569194  0.558944         0.337198  0.548230   \n",
       "6      RNA modifications  0.000000  0.000000         0.213313  0.000000   \n",
       "7       microRNA targets  0.090218  0.000000         1.006080  6.605497   \n",
       "8            RBP targets  1.520364  2.736346         1.966560  3.233984   \n",
       "9             TF targets  0.235002  0.034926         1.011789  4.695149   \n",
       "\n",
       "   poly_entry  whole_cell  \n",
       "0    0.208134    0.555656  \n",
       "1    0.123018    0.354870  \n",
       "2    0.000000    0.186136  \n",
       "3    0.034131    0.263344  \n",
       "4    0.285600    0.699022  \n",
       "5    0.115533    0.300212  \n",
       "6    0.000000    0.000000  \n",
       "7    0.000000    1.373089  \n",
       "8    0.736825    2.674970  \n",
       "9    0.000000    0.828753  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feat_heat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>chr</th>\n",
       "      <th>nuc</th>\n",
       "      <th>nucexp_from_chr</th>\n",
       "      <th>cyto</th>\n",
       "      <th>poly_entry</th>\n",
       "      <th>whole_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gene structure</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Histone modifications</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gene location</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sequence</td>\n",
       "      <td>9.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Codon frequencies</td>\n",
       "      <td>36.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Poly(A) tail length</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RNA modifications</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>microRNA targets</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RBP targets</td>\n",
       "      <td>19.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TF targets</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   class   chr   nuc  nucexp_from_chr   cyto  poly_entry  \\\n",
       "0         Gene structure   6.0   9.0             10.0   10.0         6.0   \n",
       "1  Histone modifications   3.0   6.0              5.0   11.0         5.0   \n",
       "2          Gene location   0.0   2.0              0.0    4.0         0.0   \n",
       "3               Sequence   9.0  14.0             14.0   16.0         6.0   \n",
       "4      Codon frequencies  36.0  47.0             43.0   48.0        26.0   \n",
       "5    Poly(A) tail length  10.0   8.0              9.0   12.0         5.0   \n",
       "6      RNA modifications   0.0   0.0              1.0    0.0         0.0   \n",
       "7       microRNA targets   2.0   0.0             26.0  133.0         0.0   \n",
       "8            RBP targets  19.0  39.0             34.0   51.0        18.0   \n",
       "9             TF targets  13.0   2.0             41.0  110.0         0.0   \n",
       "\n",
       "   whole_cell  \n",
       "0         9.0  \n",
       "1         5.0  \n",
       "2         2.0  \n",
       "3        14.0  \n",
       "4        39.0  \n",
       "5         9.0  \n",
       "6         0.0  \n",
       "7        28.0  \n",
       "8        34.0  \n",
       "9        36.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feat_heat_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>24</th>\n",
       "      <th>25</th>\n",
       "      <th>26</th>\n",
       "      <th>27</th>\n",
       "      <th>28</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene structure</th>\n",
       "      <td>INTRONLEN</td>\n",
       "      <td>n_exons_max</td>\n",
       "      <td>CDSLEN</td>\n",
       "      <td>UTR5LEN</td>\n",
       "      <td>UTR3LEN</td>\n",
       "      <td>5UTR_proportion_max</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Histone modifications</th>\n",
       "      <td>Histone_mod_H3K36me3-human</td>\n",
       "      <td>Histone_mod_H3K79me2-human</td>\n",
       "      <td>Histone_mod_H4K20me1-human</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene location</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence</th>\n",
       "      <td>gene_GC_content</td>\n",
       "      <td>TOP_motif_strong</td>\n",
       "      <td>min_3SS</td>\n",
       "      <td>ARE_motif_max_intronic</td>\n",
       "      <td>min_5SS</td>\n",
       "      <td>median_3SS</td>\n",
       "      <td>UTR5GC</td>\n",
       "      <td>UTR3GC</td>\n",
       "      <td>ARE_motif_n_3UTR</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Codon frequencies</th>\n",
       "      <td>Codon.CTG</td>\n",
       "      <td>Codon.TTC</td>\n",
       "      <td>Codon.GGA</td>\n",
       "      <td>Codon.TCC</td>\n",
       "      <td>Codon.CAC</td>\n",
       "      <td>Codon.CGG</td>\n",
       "      <td>Codon.CGA</td>\n",
       "      <td>Codon.GGC</td>\n",
       "      <td>Codon.AGC</td>\n",
       "      <td>Codon.CAG</td>\n",
       "      <td>Codon.TTA</td>\n",
       "      <td>Codon.GTG</td>\n",
       "      <td>Codon.GTT</td>\n",
       "      <td>Codon.TGG</td>\n",
       "      <td>Codon.GGT</td>\n",
       "      <td>Codon.AGT</td>\n",
       "      <td>Codon.CCT</td>\n",
       "      <td>Codon.AAT</td>\n",
       "      <td>Codon.CAT</td>\n",
       "      <td>Codon.TTG</td>\n",
       "      <td>Codon.GAC</td>\n",
       "      <td>Codon.CGT</td>\n",
       "      <td>Codon.ACC</td>\n",
       "      <td>Codon.CCG</td>\n",
       "      <td>Codon.GTC</td>\n",
       "      <td>Codon.GTA</td>\n",
       "      <td>Codon.CTT</td>\n",
       "      <td>Codon.GAT</td>\n",
       "      <td>Codon.TAC</td>\n",
       "      <td>Codon.GGG</td>\n",
       "      <td>Codon.GAG</td>\n",
       "      <td>Codon.TAT</td>\n",
       "      <td>Codon.GCA</td>\n",
       "      <td>Codon.ATA</td>\n",
       "      <td>Codon.CCA</td>\n",
       "      <td>Codon.ATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Poly(A) tail length</th>\n",
       "      <td>tail_length_Q75_chromatin</td>\n",
       "      <td>tail_length_median_chromatin</td>\n",
       "      <td>tail_length_Q25_chromatin</td>\n",
       "      <td>tail_length_Q75_total</td>\n",
       "      <td>tail_length_Q25_cytoplasm</td>\n",
       "      <td>tail_length_Q75_polysome</td>\n",
       "      <td>tail_length_Q25_total</td>\n",
       "      <td>tail_length_Q75_cytoplasm</td>\n",
       "      <td>tail_length_median_total</td>\n",
       "      <td>tail_length_Q25_polysome</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA modifications</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microRNA targets</th>\n",
       "      <td>TF_MSigDB_MIR106B_5P</td>\n",
       "      <td>TF_MSigDB_MIR19B_3P</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBP targets</th>\n",
       "      <td>RBP_SND1</td>\n",
       "      <td>RBP_DDX24</td>\n",
       "      <td>RBP_ZC3H11A</td>\n",
       "      <td>RBP_SRSF1</td>\n",
       "      <td>RBP_BUD13</td>\n",
       "      <td>RBP_NIPBL</td>\n",
       "      <td>RBP_APOBEC3C</td>\n",
       "      <td>RBP_KHDRBS1</td>\n",
       "      <td>RBP_FAM120A</td>\n",
       "      <td>RBP_HNRNPK</td>\n",
       "      <td>RBP_NOLC1</td>\n",
       "      <td>RBP_AQR</td>\n",
       "      <td>RBP_IGF2BP1</td>\n",
       "      <td>RBP_EFTUD2</td>\n",
       "      <td>RBP_TRA2A</td>\n",
       "      <td>RBP_GTF2F1</td>\n",
       "      <td>RBP_HNRNPC</td>\n",
       "      <td>RBP_RBFOX2</td>\n",
       "      <td>RBP_PUM2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TF targets</th>\n",
       "      <td>TF_MSigDB_DBP_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_AHRR_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_BARX2_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_ZNF843_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_HOXB4_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_ELF2_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_ZNF30_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_SETX_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_DIDO1_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_NFE2L1_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_ZNF592_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_WRNIP1_TARGET_GENES</td>\n",
       "      <td>TF_MSigDB_KDM5D_TARGET_GENES</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                0  \\\n",
       "Gene structure                          INTRONLEN   \n",
       "Histone modifications  Histone_mod_H3K36me3-human   \n",
       "Gene location                                None   \n",
       "Sequence                          gene_GC_content   \n",
       "Codon frequencies                       Codon.CTG   \n",
       "Poly(A) tail length     tail_length_Q75_chromatin   \n",
       "RNA modifications                            None   \n",
       "microRNA targets             TF_MSigDB_MIR106B_5P   \n",
       "RBP targets                              RBP_SND1   \n",
       "TF targets             TF_MSigDB_DBP_TARGET_GENES   \n",
       "\n",
       "                                                  1  \\\n",
       "Gene structure                          n_exons_max   \n",
       "Histone modifications    Histone_mod_H3K79me2-human   \n",
       "Gene location                                  None   \n",
       "Sequence                           TOP_motif_strong   \n",
       "Codon frequencies                         Codon.TTC   \n",
       "Poly(A) tail length    tail_length_median_chromatin   \n",
       "RNA modifications                              None   \n",
       "microRNA targets                TF_MSigDB_MIR19B_3P   \n",
       "RBP targets                               RBP_DDX24   \n",
       "TF targets              TF_MSigDB_AHRR_TARGET_GENES   \n",
       "\n",
       "                                                  2  \\\n",
       "Gene structure                               CDSLEN   \n",
       "Histone modifications    Histone_mod_H4K20me1-human   \n",
       "Gene location                                  None   \n",
       "Sequence                                    min_3SS   \n",
       "Codon frequencies                         Codon.GGA   \n",
       "Poly(A) tail length       tail_length_Q25_chromatin   \n",
       "RNA modifications                              None   \n",
       "microRNA targets                               None   \n",
       "RBP targets                             RBP_ZC3H11A   \n",
       "TF targets             TF_MSigDB_BARX2_TARGET_GENES   \n",
       "\n",
       "                                                   3  \\\n",
       "Gene structure                               UTR5LEN   \n",
       "Histone modifications                           None   \n",
       "Gene location                                   None   \n",
       "Sequence                      ARE_motif_max_intronic   \n",
       "Codon frequencies                          Codon.TCC   \n",
       "Poly(A) tail length            tail_length_Q75_total   \n",
       "RNA modifications                               None   \n",
       "microRNA targets                                None   \n",
       "RBP targets                                RBP_SRSF1   \n",
       "TF targets             TF_MSigDB_ZNF843_TARGET_GENES   \n",
       "\n",
       "                                                  4  \\\n",
       "Gene structure                              UTR3LEN   \n",
       "Histone modifications                          None   \n",
       "Gene location                                  None   \n",
       "Sequence                                    min_5SS   \n",
       "Codon frequencies                         Codon.CAC   \n",
       "Poly(A) tail length       tail_length_Q25_cytoplasm   \n",
       "RNA modifications                              None   \n",
       "microRNA targets                               None   \n",
       "RBP targets                               RBP_BUD13   \n",
       "TF targets             TF_MSigDB_HOXB4_TARGET_GENES   \n",
       "\n",
       "                                                 5  \\\n",
       "Gene structure                 5UTR_proportion_max   \n",
       "Histone modifications                         None   \n",
       "Gene location                                 None   \n",
       "Sequence                                median_3SS   \n",
       "Codon frequencies                        Codon.CGG   \n",
       "Poly(A) tail length       tail_length_Q75_polysome   \n",
       "RNA modifications                             None   \n",
       "microRNA targets                              None   \n",
       "RBP targets                              RBP_NIPBL   \n",
       "TF targets             TF_MSigDB_ELF2_TARGET_GENES   \n",
       "\n",
       "                                                  6  \\\n",
       "Gene structure                                 None   \n",
       "Histone modifications                          None   \n",
       "Gene location                                  None   \n",
       "Sequence                                     UTR5GC   \n",
       "Codon frequencies                         Codon.CGA   \n",
       "Poly(A) tail length           tail_length_Q25_total   \n",
       "RNA modifications                              None   \n",
       "microRNA targets                               None   \n",
       "RBP targets                            RBP_APOBEC3C   \n",
       "TF targets             TF_MSigDB_ZNF30_TARGET_GENES   \n",
       "\n",
       "                                                 7  \\\n",
       "Gene structure                                None   \n",
       "Histone modifications                         None   \n",
       "Gene location                                 None   \n",
       "Sequence                                    UTR3GC   \n",
       "Codon frequencies                        Codon.GGC   \n",
       "Poly(A) tail length      tail_length_Q75_cytoplasm   \n",
       "RNA modifications                             None   \n",
       "microRNA targets                              None   \n",
       "RBP targets                            RBP_KHDRBS1   \n",
       "TF targets             TF_MSigDB_SETX_TARGET_GENES   \n",
       "\n",
       "                                                  8  \\\n",
       "Gene structure                                 None   \n",
       "Histone modifications                          None   \n",
       "Gene location                                  None   \n",
       "Sequence                           ARE_motif_n_3UTR   \n",
       "Codon frequencies                         Codon.AGC   \n",
       "Poly(A) tail length        tail_length_median_total   \n",
       "RNA modifications                              None   \n",
       "microRNA targets                               None   \n",
       "RBP targets                             RBP_FAM120A   \n",
       "TF targets             TF_MSigDB_DIDO1_TARGET_GENES   \n",
       "\n",
       "                                                   9  \\\n",
       "Gene structure                                  None   \n",
       "Histone modifications                           None   \n",
       "Gene location                                   None   \n",
       "Sequence                                        None   \n",
       "Codon frequencies                          Codon.CAG   \n",
       "Poly(A) tail length         tail_length_Q25_polysome   \n",
       "RNA modifications                               None   \n",
       "microRNA targets                                None   \n",
       "RBP targets                               RBP_HNRNPK   \n",
       "TF targets             TF_MSigDB_NFE2L1_TARGET_GENES   \n",
       "\n",
       "                                                  10  \\\n",
       "Gene structure                                  None   \n",
       "Histone modifications                           None   \n",
       "Gene location                                   None   \n",
       "Sequence                                        None   \n",
       "Codon frequencies                          Codon.TTA   \n",
       "Poly(A) tail length                             None   \n",
       "RNA modifications                               None   \n",
       "microRNA targets                                None   \n",
       "RBP targets                                RBP_NOLC1   \n",
       "TF targets             TF_MSigDB_ZNF592_TARGET_GENES   \n",
       "\n",
       "                                                  11  \\\n",
       "Gene structure                                  None   \n",
       "Histone modifications                           None   \n",
       "Gene location                                   None   \n",
       "Sequence                                        None   \n",
       "Codon frequencies                          Codon.GTG   \n",
       "Poly(A) tail length                             None   \n",
       "RNA modifications                               None   \n",
       "microRNA targets                                None   \n",
       "RBP targets                                  RBP_AQR   \n",
       "TF targets             TF_MSigDB_WRNIP1_TARGET_GENES   \n",
       "\n",
       "                                                 12          13         14  \\\n",
       "Gene structure                                 None        None       None   \n",
       "Histone modifications                          None        None       None   \n",
       "Gene location                                  None        None       None   \n",
       "Sequence                                       None        None       None   \n",
       "Codon frequencies                         Codon.GTT   Codon.TGG  Codon.GGT   \n",
       "Poly(A) tail length                            None        None       None   \n",
       "RNA modifications                              None        None       None   \n",
       "microRNA targets                               None        None       None   \n",
       "RBP targets                             RBP_IGF2BP1  RBP_EFTUD2  RBP_TRA2A   \n",
       "TF targets             TF_MSigDB_KDM5D_TARGET_GENES        None       None   \n",
       "\n",
       "                               15          16          17         18  \\\n",
       "Gene structure               None        None        None       None   \n",
       "Histone modifications        None        None        None       None   \n",
       "Gene location                None        None        None       None   \n",
       "Sequence                     None        None        None       None   \n",
       "Codon frequencies       Codon.AGT   Codon.CCT   Codon.AAT  Codon.CAT   \n",
       "Poly(A) tail length          None        None        None       None   \n",
       "RNA modifications            None        None        None       None   \n",
       "microRNA targets             None        None        None       None   \n",
       "RBP targets            RBP_GTF2F1  RBP_HNRNPC  RBP_RBFOX2   RBP_PUM2   \n",
       "TF targets                   None        None        None       None   \n",
       "\n",
       "                              19         20         21         22         23  \\\n",
       "Gene structure              None       None       None       None       None   \n",
       "Histone modifications       None       None       None       None       None   \n",
       "Gene location               None       None       None       None       None   \n",
       "Sequence                    None       None       None       None       None   \n",
       "Codon frequencies      Codon.TTG  Codon.GAC  Codon.CGT  Codon.ACC  Codon.CCG   \n",
       "Poly(A) tail length         None       None       None       None       None   \n",
       "RNA modifications           None       None       None       None       None   \n",
       "microRNA targets            None       None       None       None       None   \n",
       "RBP targets                 None       None       None       None       None   \n",
       "TF targets                  None       None       None       None       None   \n",
       "\n",
       "                              24         25         26         27         28  \\\n",
       "Gene structure              None       None       None       None       None   \n",
       "Histone modifications       None       None       None       None       None   \n",
       "Gene location               None       None       None       None       None   \n",
       "Sequence                    None       None       None       None       None   \n",
       "Codon frequencies      Codon.GTC  Codon.GTA  Codon.CTT  Codon.GAT  Codon.TAC   \n",
       "Poly(A) tail length         None       None       None       None       None   \n",
       "RNA modifications           None       None       None       None       None   \n",
       "microRNA targets            None       None       None       None       None   \n",
       "RBP targets                 None       None       None       None       None   \n",
       "TF targets                  None       None       None       None       None   \n",
       "\n",
       "                              29         30         31         32         33  \\\n",
       "Gene structure              None       None       None       None       None   \n",
       "Histone modifications       None       None       None       None       None   \n",
       "Gene location               None       None       None       None       None   \n",
       "Sequence                    None       None       None       None       None   \n",
       "Codon frequencies      Codon.GGG  Codon.GAG  Codon.TAT  Codon.GCA  Codon.ATA   \n",
       "Poly(A) tail length         None       None       None       None       None   \n",
       "RNA modifications           None       None       None       None       None   \n",
       "microRNA targets            None       None       None       None       None   \n",
       "RBP targets                 None       None       None       None       None   \n",
       "TF targets                  None       None       None       None       None   \n",
       "\n",
       "                              34         35  \n",
       "Gene structure              None       None  \n",
       "Histone modifications       None       None  \n",
       "Gene location               None       None  \n",
       "Sequence                    None       None  \n",
       "Codon frequencies      Codon.CCA  Codon.ATG  \n",
       "Poly(A) tail length         None       None  \n",
       "RNA modifications           None       None  \n",
       "microRNA targets            None       None  \n",
       "RBP targets                 None       None  \n",
       "TF targets                  None       None  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = 'chr'\n",
    "feat_heat_names[ts]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table S6: add feature class to relevant features\n",
    "\n",
    "Run after the above as stand alone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "logger.info('Read df with relevant features')\n",
    "rel_feat = dict()\n",
    "for idx in params_opt[['rate', 'alpha']].drop_duplicates().index:\n",
    "    ts = params_opt['rate'][idx]\n",
    "    a_ = params_opt['alpha'][idx]\n",
    "    rel_feat[ts] = RND2[rt+ts+a_]   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-07-20 16:48:08] ML - Load files with feature name to class map (generated above)\n"
     ]
    }
   ],
   "source": [
    "logger.info('Load files with feature name to class map (generated above)')\n",
    "feat2cl = dict()\n",
    "for ts in rel_feat.keys():\n",
    "    filename = 'LASSOfinal_feat_class_elements_'+ts+'.tsv'\n",
    "    feat2cl[ts] = pd.read_csv(os.path.join(path,subfolder,filename), \n",
    "                                      index_col=0,header=None,sep='\\t') \n",
    "    feat2cl[ts] = feat2cl[ts].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene structure</th>\n",
       "      <th>Histone modifications</th>\n",
       "      <th>Gene location</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Codon frequencies</th>\n",
       "      <th>Poly(A) tail length</th>\n",
       "      <th>RNA modifications</th>\n",
       "      <th>microRNA targets</th>\n",
       "      <th>RBP targets</th>\n",
       "      <th>TF targets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>INTRONLEN</td>\n",
       "      <td>Histone_mod_H3K79me2-human</td>\n",
       "      <td>chr_band_16p13.3</td>\n",
       "      <td>TOP_motif_strong</td>\n",
       "      <td>Codon.CAC</td>\n",
       "      <td>tail_length_Q25_chromatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TF_MSigDB_MIR106B_5P</td>\n",
       "      <td>RBP_YBX3</td>\n",
       "      <td>TF_scRNAseq_ZBTB43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORFEXONDENSITY</td>\n",
       "      <td>Histone_mod_H3K36me3-human</td>\n",
       "      <td>chr_band_19q13.2</td>\n",
       "      <td>ORF.GAT</td>\n",
       "      <td>Codon.CGG</td>\n",
       "      <td>tail_length_Q75_chromatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TF_MSigDB_MIR32_5P</td>\n",
       "      <td>RBP_IGF2BP1</td>\n",
       "      <td>TF_MSigDB_RAG1_TARGET_GENES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mRNA_length_max</td>\n",
       "      <td>Histone_mod_H4K20me1-human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ORF.GTGGGAA</td>\n",
       "      <td>Codon.GAA</td>\n",
       "      <td>tail_length_Q75_total</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TF_MSigDB_MIR19B_3P</td>\n",
       "      <td>RBP_PABPC4</td>\n",
       "      <td>TF_MSigDB_SETX_TARGET_GENES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>n_exons_max</td>\n",
       "      <td>Histone_mod_H3K9me1-human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ARE_motif_max_intronic</td>\n",
       "      <td>Codon.TCC</td>\n",
       "      <td>tail_length_median_polysome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miRNA_family_GUGCAAA</td>\n",
       "      <td>RBP_FAM120A</td>\n",
       "      <td>TF_MSigDB_AHRR_TARGET_GENES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UTR3LEN</td>\n",
       "      <td>Histone_mod_H2AFZ-human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>min_5SS</td>\n",
       "      <td>Codon.TGT</td>\n",
       "      <td>tail_length_Q25_total</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TF_MSigDB_MIR1277_5P</td>\n",
       "      <td>RBP_UPF1</td>\n",
       "      <td>TF_MSigDB_HMGB1_TARGET_GENES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "0   Gene structure       Histone modifications     Gene location  \\\n",
       "1        INTRONLEN  Histone_mod_H3K79me2-human  chr_band_16p13.3   \n",
       "2   ORFEXONDENSITY  Histone_mod_H3K36me3-human  chr_band_19q13.2   \n",
       "3  mRNA_length_max  Histone_mod_H4K20me1-human               NaN   \n",
       "4      n_exons_max   Histone_mod_H3K9me1-human               NaN   \n",
       "5          UTR3LEN     Histone_mod_H2AFZ-human               NaN   \n",
       "\n",
       "0                Sequence Codon frequencies          Poly(A) tail length  \\\n",
       "1        TOP_motif_strong         Codon.CAC    tail_length_Q25_chromatin   \n",
       "2                 ORF.GAT         Codon.CGG    tail_length_Q75_chromatin   \n",
       "3             ORF.GTGGGAA         Codon.GAA        tail_length_Q75_total   \n",
       "4  ARE_motif_max_intronic         Codon.TCC  tail_length_median_polysome   \n",
       "5                 min_5SS         Codon.TGT        tail_length_Q25_total   \n",
       "\n",
       "0 RNA modifications      microRNA targets  RBP targets  \\\n",
       "1               NaN  TF_MSigDB_MIR106B_5P     RBP_YBX3   \n",
       "2               NaN    TF_MSigDB_MIR32_5P  RBP_IGF2BP1   \n",
       "3               NaN   TF_MSigDB_MIR19B_3P   RBP_PABPC4   \n",
       "4               NaN  miRNA_family_GUGCAAA  RBP_FAM120A   \n",
       "5               NaN  TF_MSigDB_MIR1277_5P     RBP_UPF1   \n",
       "\n",
       "0                    TF targets  \n",
       "1            TF_scRNAseq_ZBTB43  \n",
       "2   TF_MSigDB_RAG1_TARGET_GENES  \n",
       "3   TF_MSigDB_SETX_TARGET_GENES  \n",
       "4   TF_MSigDB_AHRR_TARGET_GENES  \n",
       "5  TF_MSigDB_HMGB1_TARGET_GENES  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feat2cl[ts].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: [2022-07-20 16:55:59] ML - Re-write nonzero features to file, now with classes LASSOfinal_chr_a_opt_nonzero_feats_w_class.tsv\n",
      "INFO: [2022-07-20 16:55:59] ML - Re-write nonzero features to file, now with classes LASSOfinal_nuc_a_opt_nonzero_feats_w_class.tsv\n",
      "INFO: [2022-07-20 16:55:59] ML - Re-write nonzero features to file, now with classes LASSOfinal_nucexp_from_chr_a_opt_nonzero_feats_w_class.tsv\n",
      "INFO: [2022-07-20 16:55:59] ML - Re-write nonzero features to file, now with classes LASSOfinal_cyto_a_opt_nonzero_feats_w_class.tsv\n",
      "INFO: [2022-07-20 16:55:59] ML - Re-write nonzero features to file, now with classes LASSOfinal_poly_entry_a_opt_nonzero_feats_w_class.tsv\n",
      "INFO: [2022-07-20 16:55:59] ML - Re-write nonzero features to file, now with classes LASSOfinal_whole_cell_a_opt_nonzero_feats_w_class.tsv\n"
     ]
    }
   ],
   "source": [
    "for ts in rel_feat.keys():\n",
    "    feature_class = []\n",
    "    for ft in rel_feat[ts]['features']:\n",
    "        for cl in feat2cl[ts].columns:\n",
    "            if ft in feat2cl[ts][cl].values:\n",
    "                feature_class.append(cl)\n",
    "                break\n",
    "    rel_feat[ts]['feature_class'] =  feature_class\n",
    "    filename = 'LASSOfinal_'+ts+'_a_opt_nonzero_feats_w_class.tsv'\n",
    "    rel_feat[ts].to_csv(os.path.join(path,'select_features',filename), \n",
    "                                                            sep='\\t', index=False) \n",
    "    logger.info('Re-write nonzero features to file, now with classes %s' % filename)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>features</th>\n",
       "      <th>coef</th>\n",
       "      <th>importance</th>\n",
       "      <th>feature_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBP_YBX3</td>\n",
       "      <td>-0.261762</td>\n",
       "      <td>0.261762</td>\n",
       "      <td>RBP targets</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBP_IGF2BP1</td>\n",
       "      <td>-0.214239</td>\n",
       "      <td>0.214239</td>\n",
       "      <td>RBP targets</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TF_MSigDB_MIR106B_5P</td>\n",
       "      <td>0.198486</td>\n",
       "      <td>0.198486</td>\n",
       "      <td>microRNA targets</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBP_PABPC4</td>\n",
       "      <td>-0.178002</td>\n",
       "      <td>0.178002</td>\n",
       "      <td>RBP targets</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TF_MSigDB_MIR32_5P</td>\n",
       "      <td>0.169654</td>\n",
       "      <td>0.169654</td>\n",
       "      <td>microRNA targets</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               features      coef  importance     feature_class\n",
       "0              RBP_YBX3 -0.261762    0.261762       RBP targets\n",
       "1           RBP_IGF2BP1 -0.214239    0.214239       RBP targets\n",
       "2  TF_MSigDB_MIR106B_5P  0.198486    0.198486  microRNA targets\n",
       "3            RBP_PABPC4 -0.178002    0.178002       RBP targets\n",
       "4    TF_MSigDB_MIR32_5P  0.169654    0.169654  microRNA targets"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rel_feat[ts].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
