---
title: "Fractionation-sequencing-analysis"
output: html_notebook
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyr)
library(ggplot2)
library(dplyr)
```

```{r}
gene_structure <- read.table("/Users/albert/Dropbox/Floor_Lab/Analysis/AXSNF10_Churchman_collab/gene_structure_v1.txt", sep='\t', header=TRUE, fill=TRUE)

test <- read.table("/Users/albert/Dropbox/Floor_Lab/Analysis/AXSNF10_Churchman_collab/ReadsPerGene_All/K562_DDX3-G325E_Cyto_rep1_ReadsPerGene.out.tab", sep='\t', header=FALSE, fill=TRUE, skip = 4)
# RPKM = numReads / ( geneLength/1000 * totalNumReads/1,000,000 )

# With --quantMode GeneCounts option STAR will count number reads per gene while mapping. A read is counted if it overlaps (1nt or more) one and only one gene. Both ends of the paired-end read are checked for overlaps. The counts coincide with those produced by htseq-count with default parameters. This option requires annotations in GTF format (i.e. gene id tag for each exon) specified in --sjdbGTFfile at the genome generation step or at the mapping step provided in option. STAR outputs read counts per gene into ReadsPerGene.out.tab file with 4 columns which correspond to different strandedness options:
# column 1: gene ID
# column 2: counts for unstranded RNA-seq
# 18
# column 3: counts for the 1st read strand aligned with RNA (htseq-count option -s yes)
# column 4: counts for the 2nd read strand aligned with RNA (htseq-count option -s reverse)

# https://chipster.csc.fi/manual/library-type-summary.html
# Directional, first strand:
# The second read (read 2) is from the original RNA strand/template, first read (read 1) is from the opposite strand. The information of the strand is preserved as the original RNA strand is degradated due to the dUTPs incorporated in the second synthesis step.
# Kits:
# All dUTP methods, NSR, NNSR
# TruSeq Stranded Total RNA Sample Prep Kit
# TruSeq Stranded mRNA Sample Prep Kit
# NEB Ultra Directional RNA Library Prep Kit <- ***
# Agilent SureSelect Strand-Specific
# Parameters:
# TopHat / Cufflinks / Cuffdiff: library-type fr-firststrand
# HISAT2: --rna-strandedness R (for SE) / RF (for PE)
# HTSeq: stranded -- reverse



file_list <- c("K562_DDX3-G325E_Cyto_rep1_ReadsPerGene.out.tab",
"K562_DDX3-G325E_Cyto_rep2_ReadsPerGene.out.tab",
"K562_DDX3-G325E_Nuc_rep1_ReadsPerGene.out.tab",
"K562_DDX3-G325E_Nuc_rep2_ReadsPerGene.out.tab",
"K562_DDX3-R326H_Cyto_rep1_ReadsPerGene.out.tab",
"K562_DDX3-R326H_Cyto_rep2_ReadsPerGene.out.tab",
"K562_DDX3-R326H_Nuc_rep1_ReadsPerGene.out.tab",
"K562_DDX3-R326H_Nuc_rep2_ReadsPerGene.out.tab",
"K562_DDX3-R351W_Cyto_rep1_ReadsPerGene.out.tab",
"K562_DDX3-R351W_Cyto_rep2_ReadsPerGene.out.tab",
"K562_DDX3-R351W_Nuc_rep1_ReadsPerGene.out.tab",
"K562_DDX3-R351W_Nuc_rep2_ReadsPerGene.out.tab",
"K562_DDX3-R376C_Cyto_rep1_ReadsPerGene.out.tab",
"K562_DDX3-R376C_Nuc_rep1_ReadsPerGene.out.tab",
"K562_DDX3-R376C_Nuc_rep2_ReadsPerGene.out.tab",
"K562_DDX3-R534S_Cyto_rep1_ReadsPerGene.out.tab",
"K562_DDX3-R534S_Cyto_rep2_ReadsPerGene.out.tab",
"K562_DDX3-R534S_Nuc_rep1_ReadsPerGene.out.tab",
"K562_DDX3-R534S_Nuc_rep2_ReadsPerGene.out.tab",
"K562_DDX3-T532M_Cyto_rep1_ReadsPerGene.out.tab",
"K562_DDX3-T532M_Cyto_rep2_ReadsPerGene.out.tab",
"K562_DDX3-T532M_Nuc_rep1_ReadsPerGene.out.tab",
"K562_DDX3-T532M_Nuc_rep2_ReadsPerGene.out.tab",
"K562_DDX3-wt_Cyto_rep1_ReadsPerGene.out.tab",
"K562_DDX3-wt_Cyto_rep2_ReadsPerGene.out.tab",
"K562_DDX3-wt_Nuc_rep1_ReadsPerGene.out.tab",
"K562_DDX3-wt_Nuc_rep2_ReadsPerGene.out.tab",
"K562_sgCNOT1_Cyto_rep1_ReadsPerGene.out.tab",
"K562_sgCNOT1_Cyto_rep2_ReadsPerGene.out.tab",
"K562_sgCNOT1_Nuc_rep1_ReadsPerGene.out.tab",
"K562_sgCNOT1_Nuc_rep2_ReadsPerGene.out.tab",
"K562_sgDDX3_Cyto_rep1_ReadsPerGene.out.tab",
"K562_sgDDX3_Cyto_rep2_ReadsPerGene.out.tab",
"K562_sgDDX3_Nuc_rep1_ReadsPerGene.out.tab",
"K562_sgDDX3_Nuc_rep2_ReadsPerGene.out.tab",
"K562_sgGal4_Cyto_rep1_ReadsPerGene.out.tab",
"K562_sgGal4_Cyto_rep2_ReadsPerGene.out.tab",
"K562_sgGal4_Nuc_rep1_ReadsPerGene.out.tab",
"K562_sgGal4_Nuc_rep2_ReadsPerGene.out.tab",
"K562_sgNUP160_Cyto_rep1_ReadsPerGene.out.tab",
"K562_sgNUP160_Cyto_rep2_ReadsPerGene.out.tab",
"K562_sgNUP160_Nuc_rep1_ReadsPerGene.out.tab",
"K562_sgNUP160_Nuc_rep2_ReadsPerGene.out.tab",
"K562_uninf_Cyto_rep1_ReadsPerGene.out.tab",
"K562_uninf_Cyto_rep2_ReadsPerGene.out.tab",
"K562_uninf_Nuc_rep1_ReadsPerGene.out.tab",
"K562_uninf_Nuc_rep2_ReadsPerGene.out.tab",
"K562_wt_Cyto_rep1_ReadsPerGene.out.tab",
"K562_wt_Cyto_rep2_ReadsPerGene.out.tab",
"K562_wt_Nuc_rep1_ReadsPerGene.out.tab",
"K562_wt_Nuc_rep2_ReadsPerGene.out.tab")

sample_list <- c("K562_DDX3-G325E_Cyto_rep1",
"K562_DDX3-G325E_Cyto_rep2",
"K562_DDX3-G325E_Nuc_rep1",
"K562_DDX3-G325E_Nuc_rep2",
"K562_DDX3-R326H_Cyto_rep1",
"K562_DDX3-R326H_Cyto_rep2",
"K562_DDX3-R326H_Nuc_rep1",
"K562_DDX3-R326H_Nuc_rep2",
"K562_DDX3-R351W_Cyto_rep1",
"K562_DDX3-R351W_Cyto_rep2",
"K562_DDX3-R351W_Nuc_rep1",
"K562_DDX3-R351W_Nuc_rep2",
"K562_DDX3-R376C_Cyto_rep1",
"K562_DDX3-R376C_Nuc_rep1",
"K562_DDX3-R376C_Nuc_rep2",
"K562_DDX3-R534S_Cyto_rep1",
"K562_DDX3-R534S_Cyto_rep2",
"K562_DDX3-R534S_Nuc_rep1",
"K562_DDX3-R534S_Nuc_rep2",
"K562_DDX3-T532M_Cyto_rep1",
"K562_DDX3-T532M_Cyto_rep2",
"K562_DDX3-T532M_Nuc_rep1",
"K562_DDX3-T532M_Nuc_rep2",
"K562_DDX3-wt_Cyto_rep1",
"K562_DDX3-wt_Cyto_rep2",
"K562_DDX3-wt_Nuc_rep1",
"K562_DDX3-wt_Nuc_rep2",
"K562_sgCNOT1_Cyto_rep1",
"K562_sgCNOT1_Cyto_rep2",
"K562_sgCNOT1_Nuc_rep1",
"K562_sgCNOT1_Nuc_rep2",
"K562_sgDDX3_Cyto_rep1",
"K562_sgDDX3_Cyto_rep2",
"K562_sgDDX3_Nuc_rep1",
"K562_sgDDX3_Nuc_rep2",
"K562_sgGal4_Cyto_rep1",
"K562_sgGal4_Cyto_rep2",
"K562_sgGal4_Nuc_rep1",
"K562_sgGal4_Nuc_rep2",
"K562_sgNUP160_Cyto_rep1",
"K562_sgNUP160_Cyto_rep2",
"K562_sgNUP160_Nuc_rep1",
"K562_sgNUP160_Nuc_rep2",
"K562_uninf_Cyto_rep1",
"K562_uninf_Cyto_rep2",
"K562_uninf_Nuc_rep1",
"K562_uninf_Nuc_rep2",
"K562_wt_Cyto_rep1",
"K562_wt_Cyto_rep2",
"K562_wt_Nuc_rep1",
"K562_wt_Nuc_rep2")
```

```{r}
# add all STAR output read files and use column 4 (second strand/read 2)

all_reads <- vector("list", length(file_list))
for (i in 1:length(file_list)) {
  sample <- read.table(paste0("/Users/albert/Dropbox/Floor_Lab/Analysis/AXSNF10_Churchman_collab/ReadsPerGene_All/", file_list[i]), sep='\t', header=FALSE, fill=TRUE, skip = 4)
  all_reads[[i]] <- sample$V4
}

df <- data.frame(do.call("cbind",all_reads))
rownames(df) <- test$V1
colnames(df) <- sample_list

indices <- match(gene_structure$Gene, rownames(df))
indices <- indices[!is.na(indices)]
df_coding_genes <- df[indices,]

indices2 <- match(rownames(df_coding_genes), gene_structure$Gene)

df_coding_genes$gene_names <- gene_structure$Gene[indices2]
df_coding_genes$Symbol <- gene_structure$Symbol[indices2]
df_coding_genes$median_mRNA <- gene_structure$mRNA_length_median[indices2]

# RPKM = numReads / ( geneLength/1000 * totalNumReads/1,000,000 )
df_coding_genes_matrix <- df_coding_genes[,1:51]
lengths <- df_coding_genes$median_mRNA
cS <- colSums(df_coding_genes_matrix, na.rm=TRUE)


rpkm <- t(t(df_coding_genes_matrix)/(cS/1000000))
cS[1:4]/1000000

# df_coding_genes_matrix[1:4,1:4]
# rpkm[1:4,1:4]

rpkm <- rpkm/(lengths/1000)

# lengths[1:4]/1000
# rpkm[1:4,1:4]
summary(rpkm)
# summary(lengths)
# summary(cS)
colSums(rpkm)

```

```{r}
#Encode Targets
encode_df <- read.table("/Users/albert/Dropbox/Floor_Lab/Analysis/AXSNF10_Churchman_collab/ENCODE_DDX3X_targets.txt", header=TRUE)
colnames(encode_df) <- c("chr", "start", "stop", "pval", "fc", "strand", "annotation_all", "annotation_primary", "genename", "region|bases_overlap", "RBP")

DDX3_targets <- encode_df[which(encode_df$RBP == "DDX3X"),]
head(DDX3_targets)
# lapply(DDX3_targets$annotation_primary, strsplit(x)[1])

# take first ENSG #, there are some miRNAs / lncRNAs that will be excluded  
t <- strsplit(DDX3_targets$annotation_primary, '\\|\\|')
t <- sapply(t,"[[",1)
t <- strsplit(t, '\\.')
t <- sapply(t,"[[",1)
DDX3_targets$annotation_primary_firstelement <- t
DDX3_targets$indices <- match(DDX3_targets$annotation_primary_firstelement,rownames(rpkm))
sum(is.na(DDX3_targets$indices))
#206 (uncollapsed ENSGs or lncRNAs that are not in the STAR/gene_structure dataset out of 6,987 entries (will have duplicates b/c multiple binding sites. not sure how many are unique or not))

rpkm <- data.frame(rpkm)
rpkm$encode_target <- "all-other-genes"
rpkm$encode_target[match(rownames(rpkm), DDX3_targets$annotation_primary_firstelement)] <- "DDX3-binder"
summary(match(rownames(rpkm), DDX3_targets$annotation_primary_firstelement))
#16,353 genes not included as DDX3 target out of 19,921 genes. 3,568 genes categorized as DDX3-binder

```


```{r}
# filtered analysis of RPKMs

library(dplyr)
library(ggplot2)

rpkm_filtered<-rpkm
# set RPKM filter threshold in x < ___
RPKM_cutoff = 20
rpkm_filtered[] <- lapply(rpkm, function(x) ifelse(x<RPKM_cutoff,0,x))


DDX3_KD_ratios_KDvsGal4 <- data.frame(cbind(rpkm_filtered$K562_sgDDX3_Nuc_rep1/rpkm_filtered$K562_sgGal4_Nuc_rep1,
                                            rpkm_filtered$K562_sgDDX3_Nuc_rep2/rpkm_filtered$K562_sgGal4_Nuc_rep2,
                                            rpkm_filtered$K562_sgDDX3_Cyto_rep1/rpkm_filtered$K562_sgGal4_Cyto_rep1,
                                            rpkm_filtered$K562_sgDDX3_Cyto_rep2/rpkm_filtered$K562_sgGal4_Cyto_rep2,
                                            rpkm_filtered$K562_sgCNOT1_Nuc_rep1/rpkm_filtered$K562_sgGal4_Nuc_rep1,
                                            rpkm_filtered$K562_sgCNOT1_Nuc_rep2/rpkm_filtered$K562_sgGal4_Nuc_rep2,
                                            rpkm_filtered$K562_sgCNOT1_Cyto_rep1/rpkm_filtered$K562_sgGal4_Cyto_rep1,
                                            rpkm_filtered$K562_sgCNOT1_Cyto_rep2/rpkm_filtered$K562_sgGal4_Cyto_rep2,
                                            rpkm_filtered$K562_sgNUP160_Nuc_rep1/rpkm_filtered$K562_sgGal4_Nuc_rep1,
                                            rpkm_filtered$K562_sgNUP160_Nuc_rep2/rpkm_filtered$K562_sgGal4_Nuc_rep2,
                                            rpkm_filtered$K562_sgNUP160_Cyto_rep1/rpkm_filtered$K562_sgGal4_Cyto_rep1,
                                            rpkm_filtered$K562_sgNUP160_Cyto_rep2/rpkm_filtered$K562_sgGal4_Cyto_rep2))

DDX3_KD_ratios_KDvsGal4 <- data.frame(lapply(DDX3_KD_ratios_KDvsGal4, as.numeric))
DDX3_KD_ratios_KDvsGal4 <- data.frame(lapply(DDX3_KD_ratios_KDvsGal4, log2))
DDX3_KD_ratios_KDvsGal4 <- cbind(DDX3_KD_ratios_KDvsGal4, rpkm_filtered$encode_target)
rownames(DDX3_KD_ratios_KDvsGal4) <- rownames(rpkm_filtered)
colnames(DDX3_KD_ratios_KDvsGal4) <- c("sgDDX3_Nuc_rep1",
                                       "sgDDX3_Nuc_rep2",
                                       "sgDDX3_Cyto_rep1",
                                       "sgDDX3_Cyto_rep2",
                                       "sgCNOT1_Nuc_rep1",
                                       "sgCNOT1_Nuc_rep2",
                                       "sgCNOT1_Cyto_rep1",
                                       "sgCNOT1_Cyto_rep2",
                                       "sgNUP160_Nuc_rep1",
                                       "sgNUP160_Nuc_rep2",
                                       "sgNUP160_Cyto_rep1",
                                       "sgNUP160_Cyto_rep2",
                                       "encode_target")

DDX3_KD_ratios_KDvsGal4$gene_names <- gene_structure$Symbol[match(rownames(DDX3_KD_ratios_KDvsGal4), gene_structure$Gene)]
DDX3_KD_ratios_KDvsGal4$ENSG <- rownames(DDX3_KD_ratios_KDvsGal4)

DDX3_KD_ratios_KDvsGal4_long <- DDX3_KD_ratios_KDvsGal4 %>% pivot_longer(cols=c("sgDDX3_Nuc_rep1",
                                       "sgDDX3_Nuc_rep2",
                                       "sgDDX3_Cyto_rep1",
                                       "sgDDX3_Cyto_rep2",
                                       "sgCNOT1_Nuc_rep1",
                                       "sgCNOT1_Nuc_rep2",
                                       "sgCNOT1_Cyto_rep1",
                                       "sgCNOT1_Cyto_rep2",
                                       "sgNUP160_Nuc_rep1",
                                       "sgNUP160_Nuc_rep2",
                                       "sgNUP160_Cyto_rep1",
                                       "sgNUP160_Cyto_rep2"),
                             names_to = "sample",
                             values_to ="rpkm_filtered_ratio")

DDX3_KD_ratios_KDvsGal4_long$sample <- factor(DDX3_KD_ratios_KDvsGal4_long$sample, levels = c("sgDDX3_Nuc_rep1",
                                       "sgDDX3_Nuc_rep2",
                                       "sgDDX3_Cyto_rep1",
                                       "sgDDX3_Cyto_rep2",
                                       "sgCNOT1_Nuc_rep1",
                                       "sgCNOT1_Nuc_rep2",
                                       "sgCNOT1_Cyto_rep1",
                                       "sgCNOT1_Cyto_rep2",
                                       "sgNUP160_Nuc_rep1",
                                       "sgNUP160_Nuc_rep2",
                                       "sgNUP160_Cyto_rep1",
                                       "sgNUP160_Cyto_rep2"))

DDX3_KD_ratios_KDvsGal4_long$encode_target <- as.factor(DDX3_KD_ratios_KDvsGal4_long$encode_target)

DDX3_KD_ratios_long_KDvsGal4_omit <- DDX3_KD_ratios_KDvsGal4_long[complete.cases(DDX3_KD_ratios_KDvsGal4_long),]
DDX3_KD_ratios_long_KDvsGal4_omit <- DDX3_KD_ratios_long_KDvsGal4_omit[is.finite(DDX3_KD_ratios_long_KDvsGal4_omit$rpkm_filtered_ratio),]
ggplot(DDX3_KD_ratios_long_KDvsGal4_omit, 
       aes(x=sample, y=rpkm_filtered_ratio, fill=encode_target)) + 
  geom_boxplot() + 
  ylab("log2 RPKM sgGene/ RPKM sgGal4") + ggtitle(paste0("RPKM filter < ", RPKM_cutoff))+
  theme_bw() + 
  theme(text = element_text(size = 30)) +
  scale_x_discrete(guide = guide_axis(angle = 90)) 
  



ggplot(DDX3_KD_ratios_long_KDvsGal4_omit[grep("Nuc", DDX3_KD_ratios_long_KDvsGal4_omit$sample),], 
       aes(x=sample, y=rpkm_filtered_ratio, fill=encode_target)) + 
  geom_boxplot() + 
  ylab("log2 RPKM sgGene/ RPKM sgGal4") + ggtitle(paste0("RPKM filter < ", RPKM_cutoff))+
  theme_bw() +
  scale_x_discrete(guide = guide_axis(angle = 90))   +
  theme(text = element_text(size = 30))



ggplot(DDX3_KD_ratios_long_KDvsGal4_omit[grep("sgDDX3_Nuc", DDX3_KD_ratios_long_KDvsGal4_omit$sample),], 
       aes(x=sample, y=rpkm_filtered_ratio, fill=encode_target)) + 
  geom_boxplot() + 
  ylab("log2 RPKM sgGene/ RPKM sgGal4") + ggtitle(paste0("RPKM filter < ", RPKM_cutoff))+
  theme_bw() + 
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme(text = element_text(size = 30))
```
```{r}
#wilcox test on replicates after averaging

DDX3_KD_ratios_KDvsGal4_ave  <- data.frame(cbind(rowMeans(cbind(DDX3_KD_ratios_KDvsGal4$sgDDX3_Nuc_rep1, DDX3_KD_ratios_KDvsGal4$sgDDX3_Nuc_rep2), na.rm = TRUE),
                                      rowMeans(cbind(DDX3_KD_ratios_KDvsGal4$sgCNOT1_Nuc_rep1_Nuc_rep1, DDX3_KD_ratios_KDvsGal4$sgCNOT1_Nuc_rep2), na.rm = TRUE),
                                      rowMeans(cbind(DDX3_KD_ratios_KDvsGal4$sgNUP160_Nuc_rep1, DDX3_KD_ratios_KDvsGal4$sgNUP160_Nuc_rep2), na.rm = TRUE),
                                      DDX3_KD_ratios_KDvsGal4$encode_target))
colnames(DDX3_KD_ratios_KDvsGal4_ave) <- c("sgDDX3_nuc_avg", "sgCNOT1_nuc_avg", "sgNUP160_nuc_avg", "encode_target")
DDX3_KD_ratios_KDvsGal4_ave$sgDDX3_nuc_avg <- as.numeric(DDX3_KD_ratios_KDvsGal4_ave$sgDDX3_nuc_avg)
DDX3_KD_ratios_KDvsGal4_ave$sgCNOT1_nuc_avg <- as.numeric(DDX3_KD_ratios_KDvsGal4_ave$sgCNOT1_nuc_avg)
DDX3_KD_ratios_KDvsGal4_ave$sgNUP160_nuc_avg <- as.numeric(DDX3_KD_ratios_KDvsGal4_ave$sgNUP160_nuc_avg)

wilcox.test(sgDDX3_nuc_avg~encode_target, data=DDX3_KD_ratios_KDvsGal4_ave)
wilcox.test(sgCNOT1_nuc_avg~encode_target, data=DDX3_KD_ratios_KDvsGal4_ave)
wilcox.test(sgNUP160_nuc_avg~encode_target, data=DDX3_KD_ratios_KDvsGal4_ave)
```

```{r}
### analysis of just medians DDX3 KD only ### 
DDX3_KD_subset_ratios_nolog2 <- data.frame(cbind(rpkm_filtered$K562_sgDDX3_Nuc_rep1/rpkm_filtered$K562_sgGal4_Nuc_rep1,
                                            rpkm_filtered$K562_sgDDX3_Nuc_rep2/rpkm_filtered$K562_sgGal4_Nuc_rep2,
                                            rpkm_filtered$K562_sgDDX3_Cyto_rep1/rpkm_filtered$K562_sgGal4_Cyto_rep1,
                                            rpkm_filtered$K562_sgDDX3_Cyto_rep2/rpkm_filtered$K562_sgGal4_Cyto_rep2))

DDX3_KD_subset_ratios_nolog2 <- data.frame(lapply(DDX3_KD_subset_ratios_nolog2, as.numeric))
DDX3_KD_subset_ratios_nolog2 <- cbind(DDX3_KD_subset_ratios_nolog2, rpkm_filtered$encode_target)
rownames(DDX3_KD_subset_ratios_nolog2) <- rownames(rpkm_filtered)
colnames(DDX3_KD_subset_ratios_nolog2) <- c("sgDDX3_Nuc_rep1",
                                                    "sgDDX3_Nuc_rep2",
                                                    "sgDDX3_Cyto_rep1",
                                                    "sgDDX3_Cyto_rep2",
                                             "encode_target")

# need to remove 0s , skewing the median
# a 0 in the ratio of RPKM is counted as Inf once you take log2(0) and excluded in the box and whisker plot so have to remove them here too.

filtered_df <- DDX3_KD_subset_ratios_nolog2
filtered_df[sapply(filtered_df, is.nan)] <- NA
filtered_df[sapply(filtered_df, is.infinite)] <- NA
filtered_df <- replace(filtered_df, filtered_df==0, NA)

DDX3_binder_median <- unlist(lapply(filtered_df[which(filtered_df$encode_target == "DDX3-binder"),1:4], 
                                        median, 
                                        na.rm=TRUE))
all_genes_median <- unlist(lapply(filtered_df[which(filtered_df$encode_target == "all-other-genes"),1:4], 
                                        median, 
                                        na.rm=TRUE))
median_ratios <- DDX3_binder_median/all_genes_median
median_ratios_normalized <- (DDX3_binder_median-all_genes_median)/all_genes_median


# working directly w/ median of the ratios
df_medians <- data.frame(cbind(DDX3_binder_median, 
                       all_genes_median,
                       names(DDX3_binder_median)))
colnames(df_medians) <- c("DDX3_binder_median",
                          "all_genes_median",
                          "sample_name")
df_medians$sample_name <- factor(df_medians$sample_name, levels = c("sgDDX3_Nuc_rep1",
                                                    "sgDDX3_Nuc_rep2",
                                                    "sgDDX3_Cyto_rep1",
                                                    "sgDDX3_Cyto_rep2"))

df_medians$DDX3_binder_median <- as.numeric(df_medians$DDX3_binder_median)
df_medians$all_genes_median <- as.numeric(df_medians$all_genes_median)
library(tidyr)
df_medians_long <- df_medians %>% pivot_longer(cols=c('DDX3_binder_median', 'all_genes_median'),
                               names_to="group",
                               values_to="medians")
ggplot(df_medians_long, aes(x=sample_name, y=medians, fill=group)) + 
  geom_bar(stat="identity", position="dodge") + theme_bw()


# ratio of median of ratios

df_ratios <- data.frame(cbind(median_ratios, 
                       median_ratios_normalized, 
                       names(DDX3_binder_median)))

colnames(df_ratios) <- c("median_ratios",
                          "median_ratios_normalized",
                          "sample_name")
df_ratios$sample_name <- factor(df_ratios$sample_name, levels = c("sgDDX3_Nuc_rep1",
                                                    "sgDDX3_Nuc_rep2",
                                                    "sgDDX3_Cyto_rep1",
                                                    "sgDDX3_Cyto_rep2"))
df_ratios$median_ratios <- as.numeric(df_ratios$median_ratios)
df_ratios$median_ratios_normalized <- as.numeric(df_ratios$median_ratios_normalized)


ggplot(df_ratios, aes(x=sample_name, y=median_ratios)) + 
  geom_bar(stat="identity") + theme_bw()

ggplot(df_ratios, aes(x=sample_name, y=median_ratios_normalized)) + 
  geom_bar(stat="identity") + theme_bw() + geom_hline(yintercept = 0, color="black")
```
```{r}
# median analysis for all KD # 

DDX3_KD_ratios_KDvsGal4_nolog2 <- data.frame(cbind(rpkm_filtered$K562_sgDDX3_Nuc_rep1/rpkm_filtered$K562_sgGal4_Nuc_rep1,
                                            rpkm_filtered$K562_sgDDX3_Nuc_rep2/rpkm_filtered$K562_sgGal4_Nuc_rep2,
                                            rpkm_filtered$K562_sgDDX3_Cyto_rep1/rpkm_filtered$K562_sgGal4_Cyto_rep1,
                                            rpkm_filtered$K562_sgDDX3_Cyto_rep2/rpkm_filtered$K562_sgGal4_Cyto_rep2,
                                            rpkm_filtered$K562_sgCNOT1_Nuc_rep1/rpkm_filtered$K562_sgGal4_Nuc_rep1,
                                            rpkm_filtered$K562_sgCNOT1_Nuc_rep2/rpkm_filtered$K562_sgGal4_Nuc_rep2,
                                            rpkm_filtered$K562_sgCNOT1_Cyto_rep1/rpkm_filtered$K562_sgGal4_Cyto_rep1,
                                            rpkm_filtered$K562_sgCNOT1_Cyto_rep2/rpkm_filtered$K562_sgGal4_Cyto_rep2,
                                            rpkm_filtered$K562_sgNUP160_Nuc_rep1/rpkm_filtered$K562_sgGal4_Nuc_rep1,
                                            rpkm_filtered$K562_sgNUP160_Nuc_rep2/rpkm_filtered$K562_sgGal4_Nuc_rep2,
                                            rpkm_filtered$K562_sgNUP160_Cyto_rep1/rpkm_filtered$K562_sgGal4_Cyto_rep1,
                                            rpkm_filtered$K562_sgNUP160_Cyto_rep2/rpkm_filtered$K562_sgGal4_Cyto_rep2))

DDX3_KD_ratios_KDvsGal4_nolog2 <- data.frame(lapply(DDX3_KD_ratios_KDvsGal4_nolog2, as.numeric))
DDX3_KD_ratios_KDvsGal4_nolog2 <- cbind(DDX3_KD_ratios_KDvsGal4_nolog2, rpkm_filtered$encode_target)
rownames(DDX3_KD_ratios_KDvsGal4_nolog2) <- rownames(rpkm_filtered)
colnames(DDX3_KD_ratios_KDvsGal4_nolog2) <- c("sgDDX3_Nuc_rep1",
                                       "sgDDX3_Nuc_rep2",
                                       "sgDDX3_Cyto_rep1",
                                       "sgDDX3_Cyto_rep2",
                                       "sgCNOT1_Nuc_rep1",
                                       "sgCNOT1_Nuc_rep2",
                                       "sgCNOT1_Cyto_rep1",
                                       "sgCNOT1_Cyto_rep2",
                                       "sgNUP160_Nuc_rep1",
                                       "sgNUP160_Nuc_rep2",
                                       "sgNUP160_Cyto_rep1",
                                       "sgNUP160_Cyto_rep2",
                                       "encode_target")

# need to remove 0s , skewing the median
# a 0 in the ratio of RPKM is counted as Inf once you take log2(0) and excluded in the box and whisker plot so have to remove them here too.

filtered_df <- DDX3_KD_ratios_KDvsGal4_nolog2
filtered_df[sapply(filtered_df, is.nan)] <- NA
filtered_df[sapply(filtered_df, is.infinite)] <- NA
filtered_df <- replace(filtered_df, filtered_df==0, NA)

DDX3_binder_median <- unlist(lapply(filtered_df[which(filtered_df$encode_target == "DDX3-binder"),1:12], 
                                        median, 
                                        na.rm=TRUE))
all_genes_median <- unlist(lapply(filtered_df[which(filtered_df$encode_target == "all-other-genes"),1:12], 
                                        median, 
                                        na.rm=TRUE))
median_ratios <- DDX3_binder_median/all_genes_median
median_ratios_normalized <- (DDX3_binder_median-all_genes_median)/all_genes_median
# working directly w/ median of the ratios
df_medians <- data.frame(cbind(DDX3_binder_median, 
                       all_genes_median,
                       names(DDX3_binder_median)))
colnames(df_medians) <- c("DDX3_binder_median",
                          "all_genes_median",
                          "sample_name")
df_medians$sample_name <- factor(df_medians$sample_name, levels = c("sgDDX3_Nuc_rep1",
                                       "sgDDX3_Nuc_rep2",
                                       "sgDDX3_Cyto_rep1",
                                       "sgDDX3_Cyto_rep2",
                                       "sgCNOT1_Nuc_rep1",
                                       "sgCNOT1_Nuc_rep2",
                                       "sgCNOT1_Cyto_rep1",
                                       "sgCNOT1_Cyto_rep2",
                                       "sgNUP160_Nuc_rep1",
                                       "sgNUP160_Nuc_rep2",
                                       "sgNUP160_Cyto_rep1",
                                       "sgNUP160_Cyto_rep2",
                                       "sgMCM3AP_Nuc_rep1"))

df_medians$DDX3_binder_median <- as.numeric(df_medians$DDX3_binder_median)
df_medians$all_genes_median <- as.numeric(df_medians$all_genes_median)
library(tidyr)
df_medians_long <- df_medians %>% pivot_longer(cols=c('DDX3_binder_median', 'all_genes_median'),
                               names_to="group",
                               values_to="medians")
ggplot(df_medians_long, aes(x=sample_name, y=medians, fill=group)) + 
  geom_bar(stat="identity", position="dodge")  + theme_bw()

ggplot(df_medians_long[1:4,], aes(x=sample_name, y=medians, fill=group)) + 
  geom_bar(stat="identity", position="dodge") +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) 

# ratio of median of ratios
df_ratios <- data.frame(cbind(median_ratios, 
                       median_ratios_normalized, 
                       names(DDX3_binder_median)))

colnames(df_ratios) <- c("median_ratios",
                          "median_ratios_normalized",
                          "sample_name")
df_ratios$sample_name <- factor(df_ratios$sample_name, levels = c("sgDDX3_Nuc_rep1",
                                       "sgDDX3_Nuc_rep2",
                                       "sgDDX3_Cyto_rep1",
                                       "sgDDX3_Cyto_rep2",
                                       "sgCNOT1_Nuc_rep1",
                                       "sgCNOT1_Nuc_rep2",
                                       "sgCNOT1_Cyto_rep1",
                                       "sgCNOT1_Cyto_rep2",
                                       "sgNUP160_Nuc_rep1",
                                       "sgNUP160_Nuc_rep2",
                                       "sgNUP160_Cyto_rep1",
                                       "sgNUP160_Cyto_rep2"))
df_ratios$median_ratios <- as.numeric(df_ratios$median_ratios)
df_ratios$median_ratios_normalized <- as.numeric(df_ratios$median_ratios_normalized)


ggplot(df_ratios, aes(x=sample_name, y=median_ratios)) + 
  geom_bar(stat="identity") + theme_bw()

ggplot(df_ratios[1:2,], aes(x=sample_name, y=median_ratios)) + 
  geom_bar(stat="identity") + theme_bw() +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) 

ggplot(df_ratios, aes(x=sample_name, y=median_ratios_normalized)) + 
  geom_bar(stat="identity") + geom_hline(yintercept = 0, color="black") + theme_bw()
ggplot(df_ratios[1:2,], aes(x=sample_name, y=median_ratios_normalized)) + 
  geom_bar(stat="identity") + geom_hline(yintercept = 0, color="black") +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) 


# collapsing replicates of medians analysis
df_medians_rep_means <- df_medians[,1:2] %>% mutate(grp = 1+ (row_number()-1) %/% 2) %>% 
  group_by(grp) %>% 
  summarise(across(everything(), mean, na.rm = TRUE)) %>% 
  select(-grp)
df_medians_rep_means$sample_name <- c("sgDDX3_Nuc",
                                       "sgDDX3_Cyto",
                                       "sgCNOT1_Nuc",
                                       "sgCNOT1_Cyto",
                                       "sgNUP160_Nuc",
                                       "sgNUP160_Cyto")

df_medians_rep_sd <- df_medians[,1:2] %>% mutate(grp = 1+ (row_number()-1) %/% 2) %>% 
  group_by(grp) %>% 
  summarise(across(everything(), sd, na.rm = TRUE)) %>% 
  select(-grp)
df_medians_rep_sd$sample_name <- c("sgDDX3_Nuc",
                                       "sgDDX3_Cyto",
                                       "sgCNOT1_Nuc",
                                       "sgCNOT1_Cyto",
                                       "sgNUP160_Nuc",
                                       "sgNUP160_Cyto")

df_long_medians_rep_means <- df_medians_rep_means %>% pivot_longer(cols=c('DDX3_binder_median', 'all_genes_median'),
                               names_to="group",
                               values_to="medians")

df_long_medians_rep_sd <- df_medians_rep_sd %>% pivot_longer(cols=c('DDX3_binder_median', 'all_genes_median'),
                               names_to="group",
                               values_to="sd")
df_long_medians_rep_means$sd <- df_long_medians_rep_sd$sd
df_long_medians_rep_means$sample_name <- factor(df_long_medians_rep_means$sample_name, levels = c("sgDDX3_Nuc",
                                       "sgDDX3_Cyto",
                                       "sgCNOT1_Nuc",
                                       "sgCNOT1_Cyto",
                                       "sgNUP160_Nuc",
                                       "sgNUP160_Cyto"))
# graphing median replicate collapse + error bars
ggplot(df_long_medians_rep_means, aes(x=sample_name, y=medians, fill=group)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=medians-sd, ymax=medians+sd), width=.2, position=position_dodge(.9)) +
  theme_bw() +
  theme(text = element_text(size = 15)) 

# ratio analysis replicate collapse and graphing
df_ratios_long_mean <- data.frame(df_ratios[,2]) %>% mutate(grp = 1+ (row_number()-1) %/% 2) %>% 
  group_by(grp) %>% 
  summarise(across(everything(), mean, na.rm = TRUE)) %>% 
  select(-grp)

df_ratios_long_mean$sample_name <- c("sgDDX3",
                                       "sgDDX3_Cyto",
                                       "sgCNOT1",
                                       "sgCNOT1_Cyto",
                                       "sgNUP160",
                                       "sgNUP160_Cyto")

df_ratios_long_mean$sample_name <- factor(df_ratios_long_mean$sample_name, levels = c("sgDDX3",
                                       "sgDDX3_Cyto",
                                       "sgCNOT1",
                                       "sgCNOT1_Cyto",
                                       "sgNUP160",
                                       "sgNUP160_Cyto"))

df_ratios_long_sd <- data.frame(df_ratios[,2]) %>% mutate(grp = 1+ (row_number()-1) %/% 2) %>% 
  group_by(grp) %>% 
  summarise(across(everything(), sd, na.rm = TRUE)) %>% 
  select(-grp)

df_ratios_long_sd$sample_name <- c("sgDDX3",
                                       "sgDDX3_Cyto",
                                       "sgCNOT1",
                                       "sgCNOT1_Cyto",
                                       "sgNUP160",
                                       "sgNUP160_Cyto")


df_ratios_mean_and_sd <- data.frame(cbind(df_ratios_long_mean, df_ratios_long_sd$df_ratios...2.))
colnames(df_ratios_mean_and_sd) <- c("ratio_normalized", "sample_name", "sd")
df_ratios_mean_and_sd$SEM <- df_ratios_mean_and_sd$sd / sqrt(2)

# with SD, not SEM (SEM = SD/sqrt sample size)

ggplot(df_ratios_mean_and_sd, aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-sd, ymax=ratio_normalized+sd), width=.2, position=position_dodge(.9)) +
  theme_bw() + 
  geom_hline(yintercept = 0) +
  theme(text = element_text(size = 15)) 

ggplot(df_ratios_mean_and_sd[1:6,], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-sd, ymax=ratio_normalized+sd), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")

#final fig 
ggplot(df_ratios_mean_and_sd[c(1,3,5),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-sd, ymax=ratio_normalized+sd), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")

#final fig no error bar
ggplot(df_ratios_mean_and_sd[c(1,3,5),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_hline(yintercept = 0) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")

#ex
ggplot(df_ratios_mean_and_sd[1,], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-sd, ymax=ratio_normalized+sd), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")

#ex no error bar
ggplot(df_ratios_mean_and_sd[1,], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_hline(yintercept = 0) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")









# with SEM not SD
ggplot(df_ratios_mean_and_sd, aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-SEM, ymax=ratio_normalized+SEM), width=.2, position=position_dodge(.9)) +
  theme_bw() + 
  geom_hline(yintercept = 0) +
  theme(text = element_text(size = 15)) 

ggplot(df_ratios_mean_and_sd[1:6,], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-SEM, ymax=ratio_normalized+SEM), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")

#final fig 
# ggplot(df_ratios_mean_and_sd[c(1,3,5),], aes(x=sample_name, y=ratio_normalized)) + 
#   geom_bar(stat="identity", color="black", position=position_dodge()) +
#   geom_errorbar(aes(ymin=ratio_normalized-SEM, ymax=ratio_normalized+SEM), width=.2, position=position_dodge(.9)) +
#   geom_hline(yintercept = 0) +
#   scale_x_discrete(guide = guide_axis(angle = 45)) +
#   theme_bw() +
#   theme(text = element_text(size = 30)) +
#   ylab("Ratio of DDX3X targets to all other mRNAs") + 
#   xlab("averaged replicates")

ratios <- df_ratios[c(1:2, 5:6, 9:10),]
ratios$sample_name <- c("sgDDX3", "sgDDX3", "sgCNOT1", "sgCNOT1", "sgNUP160", "sgNUP160")

ggplot(df_ratios_mean_and_sd[c(1,3,5),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_point(aes(x=sample_name, y=median_ratios_normalized), ratios, size=3) +
  geom_hline(yintercept = 0) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")


#ex
ggplot(df_ratios_mean_and_sd[1,], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-SEM, ymax=ratio_normalized+SEM), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")



```


```{r}
# filtered RPKM analysis on the rescue expts taking ratio of KD + Rescue to Uninfected

rpkm_filtered<-rpkm
# set RPKM filter threshold in x < ___
RPKM_cutoff = 20
rpkm_filtered[] <- lapply(rpkm, function(x) ifelse(x<RPKM_cutoff,0,x))


hist(log(rpkm$K562_DDX3.G325E_Nuc_rep1),xlim = c(-5,10))
hist(log(rpkm_filtered$K562_DDX3.G325E_Nuc_rep1),xlim = c(-5,10))


DDX3_rescue_ratios_filtered <- data.frame(cbind(rpkm_filtered$K562_DDX3.wt_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.wt_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.wt_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.wt_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.G325E_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.G325E_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.G325E_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.G325E_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.R326H_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.R326H_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.R326H_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.R326H_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.R351W_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.R351W_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.R351W_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.R351W_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.R376C_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
#                                                rpkm_filtered$K562_DDX3.R376C_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.R376C_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
#                                                rpkm_filtered$K562_DDX3.R376C_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.R534S_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.R534S_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.R534S_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.R534S_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.T532M_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.T532M_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.T532M_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.T532M_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2))

DDX3_rescue_ratios_filtered <- data.frame(lapply(DDX3_rescue_ratios_filtered, as.numeric))
DDX3_rescue_ratios_filtered <- data.frame(lapply(DDX3_rescue_ratios_filtered, log2))
DDX3_rescue_ratios_filtered <- cbind(DDX3_rescue_ratios_filtered, rpkm_filtered$encode_target)
rownames(DDX3_rescue_ratios_filtered) <- rownames(rpkm_filtered)
colnames(DDX3_rescue_ratios_filtered) <- c("DDX3-wt_Nuc_rep1",
                                           "DDX3-wt_Nuc_rep2",
                                           "DDX3-wt_Cyto_rep1",
                                           "DDX3-wt_Cyto_rep2",
                                           "DDX3-G325E_Nuc_rep1",
                                           "DDX3-G325E_Nuc_rep2",
                                           "DDX3-G325E_Cyto_rep1",
                                           "DDX3-G325E_Cyto_rep2",
                                           "DDX3-R326H_Nuc_rep1",
                                           "DDX3-R326H_Nuc_rep2",
                                           "DDX3-R326H_Cyto_rep1",
                                           "DDX3-R326H_Cyto_rep2",
                                           "DDX3-R351W_Nuc_rep1",
                                           "DDX3-R351W_Nuc_rep2",
                                           "DDX3-R351W_Cyto_rep1",
                                           "DDX3-R351W_Cyto_rep2",
                                           "DDX3-R376C_Nuc_rep1",
                                           "DDX3-R376C_Cyto_rep1",
                                           "DDX3-R534S_Nuc_rep1",
                                           "DDX3-R534S_Nuc_rep2",
                                           "DDX3-R534S_Cyto_rep1",
                                           "DDX3-R534S_Cyto_rep2",
                                           "DDX3-T532M_Nuc_rep1",
                                           "DDX3-T532M_Nuc_rep2",
                                           "DDX3-T532M_Cyto_rep1",
                                           "DDX3-T532M_Cyto_rep2",
                                           "encode_target")

DDX3_rescue_ratios_filtered$gene_names <- gene_structure$Symbol[match(rownames(DDX3_rescue_ratios_filtered), gene_structure$Gene)]
DDX3_rescue_ratios_filtered$ENSG <- rownames(DDX3_rescue_ratios_filtered)


DDX3_rescue_ratios_filtered_long <- DDX3_rescue_ratios_filtered %>% pivot_longer(cols=c("DDX3-wt_Nuc_rep1",
                                                                                        "DDX3-wt_Nuc_rep2",
                                                                                        "DDX3-wt_Cyto_rep1",
                                                                                        "DDX3-wt_Cyto_rep2",
                                                                                        "DDX3-G325E_Nuc_rep1",
                                                                                        "DDX3-G325E_Nuc_rep2",
                                                                                        "DDX3-G325E_Cyto_rep1",
                                                                                        "DDX3-G325E_Cyto_rep2",
                                                                                        "DDX3-R326H_Nuc_rep1",
                                                                                        "DDX3-R326H_Nuc_rep2",
                                                                                        "DDX3-R326H_Cyto_rep1",
                                                                                        "DDX3-R326H_Cyto_rep2",
                                                                                        "DDX3-R351W_Nuc_rep1",
                                                                                        "DDX3-R351W_Nuc_rep2",
                                                                                        "DDX3-R351W_Cyto_rep1",
                                                                                        "DDX3-R351W_Cyto_rep2",
                                                                                        "DDX3-R376C_Nuc_rep1",
                                                                                        "DDX3-R376C_Cyto_rep1",
                                                                                        "DDX3-R534S_Nuc_rep1",
                                                                                        "DDX3-R534S_Nuc_rep2",
                                                                                        "DDX3-R534S_Cyto_rep1",
                                                                                        "DDX3-R534S_Cyto_rep2",
                                                                                        "DDX3-T532M_Nuc_rep1",
                                                                                        "DDX3-T532M_Nuc_rep2",
                                                                                        "DDX3-T532M_Cyto_rep1",
                                                                                        "DDX3-T532M_Cyto_rep2"),
                                                                                 names_to = "sample",
                                                                                 values_to ="RPKM_ratio")



DDX3_rescue_ratios_filtered_long$sample <- factor(DDX3_rescue_ratios_filtered_long$sample, 
                                                  ordered=TRUE,
                                                  levels=c("DDX3-wt_Nuc_rep1",
                                                           "DDX3-wt_Nuc_rep2",
                                                           "DDX3-wt_Cyto_rep1",
                                                           "DDX3-wt_Cyto_rep2",
                                                           "DDX3-G325E_Nuc_rep1",
                                                           "DDX3-G325E_Nuc_rep2",
                                                           "DDX3-G325E_Cyto_rep1",
                                                           "DDX3-G325E_Cyto_rep2",
                                                           "DDX3-R326H_Nuc_rep1",
                                                           "DDX3-R326H_Nuc_rep2",
                                                           "DDX3-R326H_Cyto_rep1",
                                                           "DDX3-R326H_Cyto_rep2",
                                                           "DDX3-R351W_Nuc_rep1",
                                                           "DDX3-R351W_Nuc_rep2",
                                                           "DDX3-R351W_Cyto_rep1",
                                                           "DDX3-R351W_Cyto_rep2",
                                                           "DDX3-R376C_Nuc_rep1",
                                                           "DDX3-R376C_Cyto_rep1",
                                                           "DDX3-R534S_Nuc_rep1",
                                                           "DDX3-R534S_Nuc_rep2",
                                                           "DDX3-R534S_Cyto_rep1",
                                                           "DDX3-R534S_Cyto_rep2",
                                                           "DDX3-T532M_Nuc_rep1",
                                                           "DDX3-T532M_Nuc_rep2",
                                                           "DDX3-T532M_Cyto_rep1",
                                                           "DDX3-T532M_Cyto_rep2"))

DDX3_rescue_ratios_filtered_long$encode_target <- as.factor(DDX3_rescue_ratios_filtered_long$encode_target)

DDX3_rescue_ratios_filtered_long_omit <- DDX3_rescue_ratios_filtered_long[complete.cases(DDX3_rescue_ratios_filtered_long),]
DDX3_rescue_ratios_filtered_long_omit <- DDX3_rescue_ratios_filtered_long_omit[is.finite(DDX3_rescue_ratios_filtered_long_omit$RPKM_ratio),]
ggplot(DDX3_rescue_ratios_filtered_long_omit,
       aes(x=sample, y=RPKM_ratio, fill=encode_target)) +
  geom_boxplot() +
  ylab("log2 RPKM sgGene/ RPKM sgGal4")  +
  ggtitle(paste0("RPKM filter < ", RPKM_cutoff)) +
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw() +
  theme(text = element_text(size = 20))

ggplot(DDX3_rescue_ratios_filtered_long_omit[grep("Nuc", DDX3_rescue_ratios_filtered_long_omit$sample),],
       aes(x=sample, y=RPKM_ratio, fill=encode_target)) +
  geom_boxplot() +
  ylab("log2 RPKM sgGene/ RPKM sgGal4")  +
  ggtitle(paste0("RPKM filter < ", RPKM_cutoff)) +
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw() +
  theme(text = element_text(size = 20))



#wilcox test on averaged replicates

DDX3_rescue_ratios_ave  <- data.frame(cbind(rowMeans(cbind(DDX3_rescue_ratios_filtered$`DDX3-wt_Nuc_rep1`, DDX3_rescue_ratios_filtered$`DDX3-wt_Nuc_rep2`), na.rm = TRUE),
                                            rowMeans(cbind(DDX3_rescue_ratios_filtered$`DDX3-G325E_Nuc_rep1`, DDX3_rescue_ratios_filtered$`DDX3-G325E_Nuc_rep2`), na.rm = TRUE),
                                            rowMeans(cbind(DDX3_rescue_ratios_filtered$`DDX3-R326H_Nuc_rep1`, DDX3_rescue_ratios_filtered$`DDX3-R326H_Nuc_rep2`), na.rm = TRUE),
                                            rowMeans(cbind(DDX3_rescue_ratios_filtered$`DDX3-R351W_Nuc_rep1`,DDX3_rescue_ratios_filtered$`DDX3-R351W_Nuc_rep2`), na.rm = TRUE),
                                            rowMeans(cbind(DDX3_rescue_ratios_filtered$`DDX3-R534S_Nuc_rep1`,DDX3_rescue_ratios_filtered$`DDX3-R534S_Nuc_rep2`), na.rm = TRUE),
                                            rowMeans(cbind(DDX3_rescue_ratios_filtered$`DDX3-T532M_Nuc_rep1`, DDX3_rescue_ratios_filtered$`DDX3-T532M_Nuc_rep2` ), na.rm = TRUE),
                                            DDX3_rescue_ratios_filtered$encode_target))
colnames(DDX3_rescue_ratios_ave) <- c("DDX3_wt_nuc_avg", 
                                           "G325E_nuc_avg",
                                           "R326H_nuc_avg",
                                           "R351W_nuc_avg",
                                           "R534S_nuc_avg",
                                           "T532M_nuc_avg",
                                           "encode_target")
DDX3_rescue_ratios_ave$DDX3_wt_nuc_avg <- as.numeric(DDX3_rescue_ratios_ave$DDX3_wt_nuc_avg)
DDX3_rescue_ratios_ave$G325E_nuc_avg <- as.numeric(DDX3_rescue_ratios_ave$G325E_nuc_avg)
DDX3_rescue_ratios_ave$R326H_nuc_avg <- as.numeric(DDX3_rescue_ratios_ave$R326H_nuc_avg)
DDX3_rescue_ratios_ave$R351W_nuc_avg <- as.numeric(DDX3_rescue_ratios_ave$R351W_nuc_avg)
DDX3_rescue_ratios_ave$R534S_nuc_avg <- as.numeric(DDX3_rescue_ratios_ave$R534S_nuc_avg)
DDX3_rescue_ratios_ave$T532M_nuc_avg <- as.numeric(DDX3_rescue_ratios_ave$T532M_nuc_avg)

wilcox.test(DDX3_wt_nuc_avg~encode_target, data=DDX3_rescue_ratios_ave)
wilcox.test(G325E_nuc_avg~encode_target, data=DDX3_rescue_ratios_ave)
wilcox.test(R326H_nuc_avg~encode_target, data=DDX3_rescue_ratios_ave)
wilcox.test(R351W_nuc_avg~encode_target, data=DDX3_rescue_ratios_ave)
wilcox.test(R534S_nuc_avg~encode_target, data=DDX3_rescue_ratios_ave)
wilcox.test(T532M_nuc_avg~encode_target, data=DDX3_rescue_ratios_ave)
```
```{r}
# median analysis for all rescue with averaged replicates # 

DDX3_rescue_ratios_filtered_nolog2 <- data.frame(cbind(rpkm_filtered$K562_DDX3.wt_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.wt_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.wt_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.wt_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.G325E_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.G325E_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.G325E_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.G325E_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.R326H_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.R326H_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.R326H_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.R326H_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.R351W_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.R351W_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.R351W_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.R351W_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.R376C_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
#                                                rpkm_filtered$K562_DDX3.R376C_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.R376C_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
#                                                rpkm_filtered$K562_DDX3.R376C_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.R534S_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.R534S_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.R534S_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.R534S_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2,
                                                rpkm_filtered$K562_DDX3.T532M_Nuc_rep1/rpkm_filtered$K562_uninf_Nuc_rep1,
                                                rpkm_filtered$K562_DDX3.T532M_Nuc_rep2/rpkm_filtered$K562_uninf_Nuc_rep2,
                                                rpkm_filtered$K562_DDX3.T532M_Cyto_rep1/rpkm_filtered$K562_uninf_Cyto_rep1,
                                                rpkm_filtered$K562_DDX3.T532M_Cyto_rep2/rpkm_filtered$K562_uninf_Cyto_rep2))

DDX3_rescue_ratios_filtered_nolog2 <- data.frame(lapply(DDX3_rescue_ratios_filtered_nolog2, as.numeric))
DDX3_rescue_ratios_filtered_nolog2 <- cbind(DDX3_rescue_ratios_filtered_nolog2, rpkm_filtered$encode_target)
rownames(DDX3_rescue_ratios_filtered_nolog2) <- rownames(rpkm_filtered)
colnames(DDX3_rescue_ratios_filtered_nolog2) <- c("DDX3-wt_Nuc_rep1",
                                                  "DDX3-wt_Nuc_rep2",
                                                  "DDX3-wt_Cyto_rep1",
                                                  "DDX3-wt_Cyto_rep2",
                                                  "DDX3-G325E_Nuc_rep1",
                                                  "DDX3-G325E_Nuc_rep2",
                                                  "DDX3-G325E_Cyto_rep1",
                                                  "DDX3-G325E_Cyto_rep2",
                                                  "DDX3-R326H_Nuc_rep1",
                                                  "DDX3-R326H_Nuc_rep2",
                                                  "DDX3-R326H_Cyto_rep1",
                                                  "DDX3-R326H_Cyto_rep2",
                                                  "DDX3-R351W_Nuc_rep1",
                                                  "DDX3-R351W_Nuc_rep2",
                                                  "DDX3-R351W_Cyto_rep1",
                                                  "DDX3-R351W_Cyto_rep2",
                                                  "DDX3-R376C_Nuc_rep1",
                                                  "DDX3-R376C_Cyto_rep1",
                                                  "DDX3-R534S_Nuc_rep1",
                                                  "DDX3-R534S_Nuc_rep2",
                                                  "DDX3-R534S_Cyto_rep1",
                                                  "DDX3-R534S_Cyto_rep2",
                                                  "DDX3-T532M_Nuc_rep1",
                                                  "DDX3-T532M_Nuc_rep2",
                                                  "DDX3-T532M_Cyto_rep1",
                                                  "DDX3-T532M_Cyto_rep2",
                                                  "encode_target")

# need to remove 0s , skewing the median
# a 0 in the ratio of RPKM is counted as Inf once you take log2(0) and excluded in the box and whisker plot so have to remove them here too.

filtered_df <- DDX3_rescue_ratios_filtered_nolog2
filtered_df[sapply(filtered_df, is.nan)] <- NA
filtered_df[sapply(filtered_df, is.infinite)] <- NA
filtered_df <- replace(filtered_df, filtered_df==0, NA)

DDX3_binder_median <- unlist(lapply(filtered_df[which(filtered_df$encode_target == "DDX3-binder"),1:26], 
                                    median, 
                                    na.rm=TRUE))
all_genes_median <- unlist(lapply(filtered_df[which(filtered_df$encode_target == "all-other-genes"),1:26], 
                                  median, 
                                  na.rm=TRUE))
median_ratios <- DDX3_binder_median/all_genes_median
median_ratios_normalized <- (DDX3_binder_median-all_genes_median)/all_genes_median

# working directly w/ median of the ratios
df_medians <- data.frame(cbind(DDX3_binder_median, 
                               all_genes_median,
                               names(DDX3_binder_median)))
colnames(df_medians) <- c("DDX3_binder_median",
                          "all_genes_median",
                          "sample_name")
df_medians$sample_name <- factor(df_medians$sample_name,
                                 ordered=TRUE,
                                 levels = c("DDX3-wt_Nuc_rep1",
                                            "DDX3-wt_Nuc_rep2",
                                            "DDX3-wt_Cyto_rep1",
                                            "DDX3-wt_Cyto_rep2",
                                            "DDX3-G325E_Nuc_rep1",
                                            "DDX3-G325E_Nuc_rep2",
                                            "DDX3-G325E_Cyto_rep1",
                                            "DDX3-G325E_Cyto_rep2",
                                            "DDX3-R326H_Nuc_rep1",
                                            "DDX3-R326H_Nuc_rep2",
                                            "DDX3-R326H_Cyto_rep1",
                                            "DDX3-R326H_Cyto_rep2",
                                            "DDX3-R351W_Nuc_rep1",
                                            "DDX3-R351W_Nuc_rep2",
                                            "DDX3-R351W_Cyto_rep1",
                                            "DDX3-R351W_Cyto_rep2",
                                            "DDX3-R376C_Nuc_rep1",
                                            "DDX3-R376C_Cyto_rep1",
                                            "DDX3-R534S_Nuc_rep1",
                                            "DDX3-R534S_Nuc_rep2",
                                            "DDX3-R534S_Cyto_rep1",
                                            "DDX3-R534S_Cyto_rep2",
                                            "DDX3-T532M_Nuc_rep1",
                                            "DDX3-T532M_Nuc_rep2",
                                            "DDX3-T532M_Cyto_rep1",
                                            "DDX3-T532M_Cyto_rep2"))

df_medians$DDX3_binder_median <- as.numeric(df_medians$DDX3_binder_median)
df_medians$all_genes_median <- as.numeric(df_medians$all_genes_median)
library(tidyr)

df_medians_long <- df_medians %>% pivot_longer(cols=c('DDX3_binder_median', 'all_genes_median'),
                                               names_to="group",
                                               values_to="medians")

ggplot(df_medians_long, aes(x=sample_name, y=medians, fill=group)) + 
  geom_bar(stat="identity", position="dodge")  + 
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw() +
  theme(axis.text = element_text(size = 20))


# ratio of median of ratios

df_ratios <- data.frame(cbind(median_ratios, 
                              median_ratios_normalized, 
                              names(DDX3_binder_median)))

colnames(df_ratios) <- c("median_ratios",
                         "median_ratios_normalized",
                         "sample_name")
df_ratios$sample_name <- factor(df_ratios$sample_name, ordered=TRUE,
                                levels = c("DDX3-wt_Nuc_rep1",
                                           "DDX3-wt_Nuc_rep2",
                                           "DDX3-wt_Cyto_rep1",
                                           "DDX3-wt_Cyto_rep2",
                                           "DDX3-G325E_Nuc_rep1",
                                           "DDX3-G325E_Nuc_rep2",
                                           "DDX3-G325E_Cyto_rep1",
                                           "DDX3-G325E_Cyto_rep2",
                                           "DDX3-R326H_Nuc_rep1",
                                           "DDX3-R326H_Nuc_rep2",
                                           "DDX3-R326H_Cyto_rep1",
                                           "DDX3-R326H_Cyto_rep2",
                                           "DDX3-R351W_Nuc_rep1",
                                           "DDX3-R351W_Nuc_rep2",
                                           "DDX3-R351W_Cyto_rep1",
                                           "DDX3-R351W_Cyto_rep2",
                                           "DDX3-R376C_Nuc_rep1",
                                           "DDX3-R376C_Cyto_rep1",
                                           "DDX3-R534S_Nuc_rep1",
                                           "DDX3-R534S_Nuc_rep2",
                                           "DDX3-R534S_Cyto_rep1",
                                           "DDX3-R534S_Cyto_rep2",
                                           "DDX3-T532M_Nuc_rep1",
                                           "DDX3-T532M_Nuc_rep2",
                                           "DDX3-T532M_Cyto_rep1",
                                           "DDX3-T532M_Cyto_rep2"))
df_ratios$median_ratios <- as.numeric(df_ratios$median_ratios)
df_ratios$median_ratios_normalized <- as.numeric(df_ratios$median_ratios_normalized)


ggplot(df_ratios, aes(x=sample_name, y=median_ratios)) + 
  geom_bar(stat="identity") + 
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw() +
  theme(axis.text = element_text(size = 20))

ggplot(df_ratios, aes(x=sample_name, y=median_ratios_normalized)) + 
  geom_bar(stat="identity") + geom_hline(yintercept = 0, color="black") + 
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw() +
  theme(axis.text = element_text(size = 20))



# collapsing replicates of medians analysis
# duplicate R376C_Nuc_rep1 and DDX3-R376C_Cyto_rep1	 b/c there is no second replicate, but for the average replicates code to work there needs to be 2 values. 
df_medians_added_dup <- rbind(df_medians[1:17,], 
                              df_medians[17,], 
                              df_medians[18,], 
                              df_medians[18:26,])
df_medians_rep_means <- df_medians_added_dup[,1:2] %>% mutate(grp = 1+ (row_number()-1) %/% 2) %>% 
  group_by(grp) %>% 
  summarise(across(everything(), mean, na.rm = TRUE)) %>% 
  select(-grp)
df_medians_rep_means$sample_name <- c("DDX3-wt_Nuc",
                                      "DDX3-wt_Cyto",
                                      "DDX3-G325E_Nuc",
                                      "DDX3-G325E_Cyto",
                                      "DDX3-R326H_Nuc",
                                      "DDX3-R326H_Cyto",
                                      "DDX3-R351W_Nuc",
                                      "DDX3-R351W_Cyto",
                                      "DDX3-R376C_Nuc",
                                      "DDX3-R376C_Cyto",
                                      "DDX3-R534S_Nuc",
                                      "DDX3-R534S_Cyto",
                                      "DDX3-T532M_Nuc",
                                      "DDX3-T532M_Cyto")

df_medians_rep_sd <- df_medians_added_dup[,1:2] %>% mutate(grp = 1+ (row_number()-1) %/% 2) %>% 
  group_by(grp) %>% 
  summarise(across(everything(), sd, na.rm = TRUE)) %>% 
  select(-grp)
df_medians_rep_sd$sample_name <- c("DDX3-wt_Nuc",
                                   "DDX3-wt_Cyto",
                                   "DDX3-G325E_Nuc",
                                   "DDX3-G325E_Cyto",
                                   "DDX3-R326H_Nuc",
                                   "DDX3-R326H_Cyto",
                                   "DDX3-R351W_Nuc",
                                   "DDX3-R351W_Cyto",
                                   "DDX3-R376C_Nuc",
                                   "DDX3-R376C_Cyto",
                                   "DDX3-R534S_Nuc",
                                   "DDX3-R534S_Cyto",
                                   "DDX3-T532M_Nuc",
                                   "DDX3-T532M_Cyto")
df_long_medians_rep_means <- df_medians_rep_means %>% pivot_longer(cols=c('DDX3_binder_median', 'all_genes_median'),
                                                                   names_to="group",
                                                                   values_to="medians")

df_long_medians_rep_sd <- df_medians_rep_sd %>% pivot_longer(cols=c('DDX3_binder_median', 'all_genes_median'),
                                                             names_to="group",
                                                             values_to="sd")
df_long_medians_rep_means$sd <- df_long_medians_rep_sd$sd
df_long_medians_rep_means$sample_name <- factor(df_long_medians_rep_means$sample_name, ordered=TRUE, levels = c("DDX3-wt_Nuc",
                                                                                                                "DDX3-wt_Cyto",
                                                                                                                "DDX3-G325E_Nuc",
                                                                                                                "DDX3-G325E_Cyto",
                                                                                                                "DDX3-R326H_Nuc",
                                                                                                                "DDX3-R326H_Cyto",
                                                                                                                "DDX3-R351W_Nuc",
                                                                                                                "DDX3-R351W_Cyto",
                                                                                                                "DDX3-R376C_Nuc",
                                                                                                                "DDX3-R376C_Cyto",
                                                                                                                "DDX3-R534S_Nuc",
                                                                                                                "DDX3-R534S_Cyto",
                                                                                                                "DDX3-T532M_Nuc",
                                                                                                                "DDX3-T532M_Cyto"))
# graphing median replicate collapse + error bars
ggplot(df_long_medians_rep_means, aes(x=sample_name, y=medians, fill=group)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=medians-sd, ymax=medians+sd), width=.2, position=position_dodge(.9)) +
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw() +
  theme(axis.text = element_text(size = 20))

# ratio analysis replicate collapse and graphing
df_ratios_added_dup <- rbind(df_ratios[1:17,], 
                              df_ratios[17,], 
                              df_ratios[18,], 
                              df_ratios[18:26,])
df_ratios_long_mean <- data.frame(df_ratios_added_dup[,2]) %>% mutate(grp = 1+ (row_number()-1) %/% 2) %>% 
  group_by(grp) %>% 
  summarise(across(everything(), mean, na.rm = TRUE)) %>% 
  select(-grp)

df_ratios_long_mean$sample_name <- c("DDX3-wt",
                                   "DDX3-wt_Cyto",
                                   "DDX3-G325E",
                                   "DDX3-G325E_Cyto",
                                   "DDX3-R326H",
                                   "DDX3-R326H_Cyto",
                                   "DDX3-R351W",
                                   "DDX3-R351W_Cyto",
                                   "DDX3-R376C",
                                   "DDX3-R376C_Cyto",
                                   "DDX3-R534S",
                                   "DDX3-R534S_Cyto",
                                   "DDX3-T532M",
                                   "DDX3-T532M_Cyto")
df_ratios_long_mean$sample_name <- factor(df_ratios_long_mean$sample_name, ordered=TRUE, levels = c("DDX3-wt",
                                                                                                    "DDX3-wt_Cyto",
                                                                                                    "DDX3-G325E",
                                                                                                    "DDX3-G325E_Cyto",
                                                                                                    "DDX3-R326H",
                                                                                                    "DDX3-R326H_Cyto",
                                                                                                    "DDX3-R351W",
                                                                                                    "DDX3-R351W_Cyto",
                                                                                                    "DDX3-R376C",
                                                                                                    "DDX3-R376C_Cyto",
                                                                                                    "DDX3-R534S",
                                                                                                    "DDX3-R534S_Cyto",
                                                                                                    "DDX3-T532M",
                                                                                                    "DDX3-T532M_Cyto"))
df_ratios_long_sd <- data.frame(df_ratios_added_dup[,2]) %>% mutate(grp = 1+ (row_number()-1) %/% 2) %>% 
  group_by(grp) %>% 
  summarise(across(everything(), sd, na.rm = TRUE)) %>% 
  select(-grp)
df_ratios_long_sd$sample_name <-c("DDX3-wt_Nuc",
                                   "DDX3-wt_Cyto",
                                   "DDX3-G325E_Nuc",
                                   "DDX3-G325E_Cyto",
                                   "DDX3-R326H_Nuc",
                                   "DDX3-R326H_Cyto",
                                   "DDX3-R351W_Nuc",
                                   "DDX3-R351W_Cyto",
                                   "DDX3-R376C_Nuc",
                                   "DDX3-R376C_Cyto",
                                   "DDX3-R534S_Nuc",
                                   "DDX3-R534S_Cyto",
                                   "DDX3-T532M_Nuc",
                                   "DDX3-T532M_Cyto")
df_ratios_mean_and_sd <- data.frame(cbind(df_ratios_long_mean, df_ratios_long_sd$df_ratios_added_dup...2.))
colnames(df_ratios_mean_and_sd) <- c("ratio_normalized", "sample_name", "sd")
df_ratios_mean_and_sd$SEM <- df_ratios_mean_and_sd$sd / sqrt(2)

# using SD
ggplot(df_ratios_mean_and_sd, aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-sd, ymax=ratio_normalized+sd), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) + 
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw() +
  theme(axis.text = element_text(size = 20))

#exclude R376C missing a replicate for that sample
ggplot(df_ratios_mean_and_sd[c(1:8,11:14),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-sd, ymax=ratio_normalized+sd), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) + 
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")

# exclude R376C - no replicate
ggplot(df_ratios_mean_and_sd[c(1, 3, 5, 7, 11, 13),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-sd, ymax=ratio_normalized+sd), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) + 
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")

# exclude R376C - no replicate
ggplot(df_ratios_mean_and_sd[c(1, 3, 5, 7, 11, 13),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_hline(yintercept = 0) + 
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")




# using SEM, not SD
ggplot(df_ratios_mean_and_sd, aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-SEM, ymax=ratio_normalized+SEM), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) + 
  scale_x_discrete(guide = guide_axis(angle = 90)) +
  theme_bw() +
  theme(axis.text = element_text(size = 20))

# exclude R376C - no replicate
ggplot(df_ratios_mean_and_sd[c(1:8,11:14),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-SEM, ymax=ratio_normalized+SEM), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) + 
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")

# exclude R376C - no replicate
ggplot(df_ratios_mean_and_sd[c(1, 3, 5, 7, 11, 13),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_errorbar(aes(ymin=ratio_normalized-SEM, ymax=ratio_normalized+SEM), width=.2, position=position_dodge(.9)) +
  geom_hline(yintercept = 0) + 
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")


# take all ratios of each replicate before averaging of nuclear samples. exclude R376C samples (missing a replicate)
ratios <- df_ratios[c(1:2, 5:6,  9:10, 13:14, 19:20, 23:24),]
ratios$sample_name <- c("DDX3-wt", "DDX3-wt", "DDX3-G325E",  "DDX3-G325E", "DDX3-R326H", "DDX3-R326H","DDX3-R351W", "DDX3-R351W","DDX3-R534S", "DDX3-R534S","DDX3-T532M","DDX3-T532M")
# exclude R376C - no replicate
ggplot(df_ratios_mean_and_sd[c(1, 3, 5, 7, 11, 13),], aes(x=sample_name, y=ratio_normalized)) + 
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  geom_point(aes(x=sample_name, y=median_ratios_normalized), ratios, size=3)+
  geom_hline(yintercept = 0) + 
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size = 30)) +
  ylab("Ratio of DDX3X targets to all other mRNAs") + 
  xlab("averaged replicates")
```













