---
title: "Nuc / Cyto ratios: spike in normalized scRNAseq vs rates"
author: "Robert Ietswaart"
date: "1/29/2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/n/groups/churchman/ri23/code')
```

## R Markdown
Source: `Nuc_cyto_ratios_NucWash_vs_rates_20231031.Rmd`
```{r}
# BiocManager::install(c("dplyr", "ggrepel"))#Only once required to install
library(dplyr)
library(ggplot2)
library(ggrepel)

'%!in%' <- function(x,y)!('%in%'(x,y))
```

## Load data
```{r}
path <- '/n/groups/churchman/ri23/bseq/NC20240129/'

#load DEseq2 results of Cyto over Nuc spike in normalized RNA level ratios: so inverse of Nuc / Cyto
file_de = '/n/groups/churchman/ri23/bseq/CG20230908/DEseq2/DEseq2_contrast_condition_C_vs_N.csv'
de = read.csv(file_de, sep=',', header = TRUE, row.names = 1)
# Threshold_padj = 0.1
# de[which((!is.na(de$padj)) & (de$padj >= Threshold_padj)), "log2FoldChange"] <- 0

file_de_nw_n = '/n/groups/churchman/ri23/bseq/CG20230908/DEseq2/DEseq2_contrast_condition_NW_vs_N.csv'
de_nw_n = read.csv(file_de_nw_n, sep=',', header = TRUE, row.names = 1)

file_de_nw_c = '/n/groups/churchman/ri23/bseq/CG20230908/DEseq2/DEseq2_contrast_condition_NW_vs_C.csv'
de_nw_c = read.csv(file_de_nw_c, sep=',', header = TRUE, row.names = 1)

#normalized counts
file_nc = '/n/groups/churchman/ri23/bseq/CG20230908/DEseq2/DEseq2_spikein_normed_counts.csv'
nc = read.csv(file_nc, sep=',', header = TRUE, row.names = 1)

# #Theshold for total RNA expression
# Threshold_RNA <- 10 #10000 #1000 #100 #0
# de <- de[de$baseMean >= Threshold_RNA,]

#load PUNDs
file_punds = '/n/groups/churchman/ri23/bseq/BayesFactor20240112/Nucdeg_20240112_human_final.tsv'
punds = read.csv(file_punds, sep='\t', header = TRUE, row.names = 1)

#Load raw rates: for PUNDs
path_rates <- '/n/groups/churchman/ri23/bseq/Bayes20240120_K562/'
filename <- 'Bayes_Rates_20240120_human_final.tsv'
rates <- read.csv(paste(path_rates,filename,sep=""), header=T, sep="\t", row.names = 1)
#Filter for genes with DE data
rates <- rates[rownames(rates) %in% rownames(de),]

#Append DE data: note the - to go from Log2(Cyto/Nuc) -> Nuc/Cyto
rates$nuc_cyto_DE_ratio <- 2^(-de$log2FoldChange[match(rownames(rates), rownames(de))])

#Append DE data: NucWash/Nuc
rates$nw_nuc_DE_ratio <- 2^(de_nw_n$log2FoldChange[match(rownames(rates), rownames(de_nw_n))])

#Append DE data: NucWash/Cyto
rates$nw_cyto_DE_ratio <- 2^(de_nw_c$log2FoldChange[match(rownames(rates), rownames(de_nw_c))])
```

#Calculate the normed counts Nuc / Cyto ratio 
as alternative to the one derived from DEseq2 log2FoldChange
```{r}
eps <- 1e-1
# Compartments = c('N', 'NW', 'NW_NRRD', 'C', 'WC')
nc$WC <- exp(( log(nc$X1A_WC_S1 + eps) + log(nc$X1B_WC_S5 + eps) + log(nc$X2B_WC_S9 + eps) + log(nc$X2C_WC_S13 + eps) ) / 4)
nc$C <-  exp(( log(nc$X1A_C_S2 + eps) + log(nc$X1B_C_S6 + eps) + log(nc$X2B_C_S10 + eps) + log(nc$X2C_C_S14 + eps) ) / 4)
nc$NW <-  exp(( log(nc$X1A_NW_S3 + eps) + log(nc$X1B_NW_S7 + eps) + log(nc$X2B_NW_S11 + eps) + log(nc$X2C_NW_S15 + eps) ) / 4)
nc$N <-  exp(( log(nc$X1A_N_S4 + eps) + log(nc$X1B_N_S8 + eps) + log(nc$X2B_N_S12 + eps) + log(nc$X2C_N_S16 + eps) ) / 4)
nc$NW_NRRD <-  exp(( log(nc$X1B_NW_NRRD_S17 + eps) + log(nc$X2B_NW_NRRD_S18 + eps) + log(nc$X2C_NW_NRRD_S19 + eps) ) / 3)

nc$nuc_cyto_normed_counts_ratio <- nc$N / nc$C
rates$nuc_cyto_normed_counts_ratio <- nc$nuc_cyto_normed_counts_ratio[match(rownames(rates), rownames(nc))]
rates$NW <- nc$NW[match(rownames(rates), rownames(nc))]
```


#Calculate predicted Nuc/Cyto ratio from rate for all and PUNDs
geometric average of rates.
N / CY = k_CY / k_N = t1/2_N / t1/2_CY
N / CY = k_CY * (kE + kR + kND) / kE*kR
```{r}
Timescales = c('nuc', 'cyto', 'chr_release', 'nucexp', 'nucdeg')
ot = '.Mean' #'.MAP' #alternative: 
reps <- c('T.', 'U.')
rt = 'k_'#rate_type, alternative: 'half_life_'

for(ts in Timescales){
  if(ts == 'chr_release'){
      rates[[ts]] <- sqrt(rates[[paste(reps[1],rt,'chr',ot,sep='')]] * rates[[paste(reps[2],rt,'chr',ot,sep='')]])
      #PUNDs:
      rates[[ts]][rownames(rates) %in% rownames(punds)] <- sqrt(rates[[paste(reps[1],rt,ts,'_from_nucdeg.MAP',sep='')]][rownames(rates) %in% rownames(punds)] * rates[[paste(reps[2],rt,ts,'_from_nucdeg.MAP',sep='')]][rownames(rates) %in% rownames(punds)])
  }#if
  else if(ts == 'nucexp'){
      rates[[ts]] <- sqrt(rates[[paste(reps[1],rt,ts,'_from_nucres',ot,sep='')]] * rates[[paste(reps[2],rt,ts,'_from_nucres',ot,sep='')]])
      #PUNDs:
      rates[[ts]][rownames(rates) %in% rownames(punds)] <- sqrt(rates[[paste(reps[1],rt,ts,'_from_nucdeg.MAP',sep='')]][rownames(rates) %in% rownames(punds)] * rates[[paste(reps[2],rt,ts,'_from_nucdeg.MAP',sep='')]][rownames(rates) %in% rownames(punds)]) 
  }#else if 
  else if(ts == 'nucdeg'){
      #Default
      rates[[ts]] <- 0
      #PUNDs:
      rates[[ts]][rownames(rates) %in% rownames(punds)] <- sqrt(rates[[paste(reps[1],rt,ts,'.MAP',sep='')]][rownames(rates) %in% rownames(punds)] * rates[[paste(reps[2],rt,ts,'.MAP',sep='')]][rownames(rates) %in% rownames(punds)])

      # # Only for #Scatterplot: NW levels and enrichment vs half life nucdeg
      # rates[[ts]] <- sqrt(rates[[paste(reps[1],rt,ts,ot,sep='')]] * rates[[paste(reps[2],rt,ts,ot,sep='')]])  
  }#else if
  else{
      rates[[ts]] <- sqrt(rates[[paste(reps[1],rt,ts,ot,sep='')]] * rates[[paste(reps[2],rt,ts,ot,sep='')]])  
  }#else
}#for
```

#Calculate rate Nuc / Cyto ratio
```{r}
rates$rates_nuc_cyto_ratio <- rates$cyto * (rates$chr_release + rates$nucexp + rates$nucdeg) / (rates$chr_release * rates$nucexp)
rates$rates_nuc_cyto_ratio_from_nucres <- rates$cyto / rates$nuc
```

#Binned boxplots DEseq: Figure S3G
```{r}
df <- rates
df <- df[!is.na(df$nuc_cyto_DE_ratio),]
df <- df[!is.na(df$rates_nuc_cyto_ratio),]
df$PUND <- rownames(df) %in% rownames(punds)
breaks <- c(0, 1, 2, 3, 4, Inf)#c(0.1, 0.13, 0.17, 0.22, 0.28, 0.36, 0.47)#, Inf) #c(0.1, 0.2, 0.3, 0.4, Inf)#    # For example, bins from 20-30, 30-40, etc.
df$x_bin <- cut(df$nuc_cyto_DE_ratio, breaks = breaks, include.lowest = TRUE, labels = FALSE, right = TRUE)

p <- ggplot(df, aes(x=factor(x_bin), y=rates_nuc_cyto_ratio)) +
  geom_violin(width=0.8, aes(fill = PUND)) +
  geom_boxplot(aes(group = interaction(PUND, x_bin)), fill=NA,width=0.8, outlier.shape=NA, position = position_dodge2(preserve = "single")) +
  scale_x_discrete(labels=paste0(breaks[-length(breaks)], "-", breaks[-1])) +
  scale_fill_manual(values = c("grey", "#B80000")) +
  labs(x="Differential expression nuclear to cytoplasmic fold change", y="RNA flow nuclear to cytoplasmic RNA ratio \n kCY(kE+kR+kND) / kRkE") +
  ylim(0, 3) +
  theme_minimal() #+
  # theme(axis.title.x=element_blank(),
  #       axis.title.y=element_blank())
p

table(df$x_bin)
dim(df)
sum(df$PUND)
sum(df$PUND==FALSE)

fname=paste(path,'Nuc_cyto_ratio_DEseq_RNAflow.',sep="")
ggsave(paste(fname,'.pdf',sep=''),p,width = 8,height=6)#inch

```